% Generated by Sphinx.
\def\sphinxdocclass{report}
\documentclass[a4paper,11pt,english]{sphinxmanual}
\usepackage[utf8]{inputenc}
\DeclareUnicodeCharacter{00A0}{\nobreakspace}
\usepackage{cmap}
\usepackage[T1]{fontenc}
\usepackage{babel}
\usepackage{times}
\usepackage[Bjarne]{fncychap}
\usepackage{longtable}
\usepackage{sphinx}
\usepackage{multirow}

\addto\captionsenglish{\renewcommand{\figurename}{Fig. }}
\addto\captionsenglish{\renewcommand{\tablename}{Table }}
\floatname{literal-block}{Listing }



\title{CBMPy User Guide}
\date{February 25, 2016}
\release{0.7.4}
\author{Brett G. Olivier}
\newcommand{\sphinxlogo}{\includegraphics{pysces_cbm1_head.pdf}\par}
\renewcommand{\releasename}{Release}
\makeindex

\makeatletter
\def\PYG@reset{\let\PYG@it=\relax \let\PYG@bf=\relax%
    \let\PYG@ul=\relax \let\PYG@tc=\relax%
    \let\PYG@bc=\relax \let\PYG@ff=\relax}
\def\PYG@tok#1{\csname PYG@tok@#1\endcsname}
\def\PYG@toks#1+{\ifx\relax#1\empty\else%
    \PYG@tok{#1}\expandafter\PYG@toks\fi}
\def\PYG@do#1{\PYG@bc{\PYG@tc{\PYG@ul{%
    \PYG@it{\PYG@bf{\PYG@ff{#1}}}}}}}
\def\PYG#1#2{\PYG@reset\PYG@toks#1+\relax+\PYG@do{#2}}

\expandafter\def\csname PYG@tok@gd\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.63,0.00,0.00}{##1}}}
\expandafter\def\csname PYG@tok@gu\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.50,0.00,0.50}{##1}}}
\expandafter\def\csname PYG@tok@gt\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.27,0.87}{##1}}}
\expandafter\def\csname PYG@tok@gs\endcsname{\let\PYG@bf=\textbf}
\expandafter\def\csname PYG@tok@gr\endcsname{\def\PYG@tc##1{\textcolor[rgb]{1.00,0.00,0.00}{##1}}}
\expandafter\def\csname PYG@tok@cm\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.25,0.50,0.56}{##1}}}
\expandafter\def\csname PYG@tok@vg\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.38,0.84}{##1}}}
\expandafter\def\csname PYG@tok@m\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@mh\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@cs\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.50,0.56}{##1}}\def\PYG@bc##1{\setlength{\fboxsep}{0pt}\colorbox[rgb]{1.00,0.94,0.94}{\strut ##1}}}
\expandafter\def\csname PYG@tok@ge\endcsname{\let\PYG@it=\textit}
\expandafter\def\csname PYG@tok@vc\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.38,0.84}{##1}}}
\expandafter\def\csname PYG@tok@il\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@go\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.20,0.20,0.20}{##1}}}
\expandafter\def\csname PYG@tok@cp\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@gi\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.63,0.00}{##1}}}
\expandafter\def\csname PYG@tok@gh\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.00,0.50}{##1}}}
\expandafter\def\csname PYG@tok@ni\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.84,0.33,0.22}{##1}}}
\expandafter\def\csname PYG@tok@nl\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.13,0.44}{##1}}}
\expandafter\def\csname PYG@tok@nn\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.05,0.52,0.71}{##1}}}
\expandafter\def\csname PYG@tok@no\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.38,0.68,0.84}{##1}}}
\expandafter\def\csname PYG@tok@na\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@nb\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@nc\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.05,0.52,0.71}{##1}}}
\expandafter\def\csname PYG@tok@nd\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.33,0.33,0.33}{##1}}}
\expandafter\def\csname PYG@tok@ne\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@nf\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.02,0.16,0.49}{##1}}}
\expandafter\def\csname PYG@tok@si\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.44,0.63,0.82}{##1}}}
\expandafter\def\csname PYG@tok@s2\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@vi\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.38,0.84}{##1}}}
\expandafter\def\csname PYG@tok@nt\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.02,0.16,0.45}{##1}}}
\expandafter\def\csname PYG@tok@nv\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.38,0.84}{##1}}}
\expandafter\def\csname PYG@tok@s1\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@gp\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.78,0.36,0.04}{##1}}}
\expandafter\def\csname PYG@tok@sh\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@ow\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@sx\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.78,0.36,0.04}{##1}}}
\expandafter\def\csname PYG@tok@bp\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@c1\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.25,0.50,0.56}{##1}}}
\expandafter\def\csname PYG@tok@kc\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@c\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.25,0.50,0.56}{##1}}}
\expandafter\def\csname PYG@tok@mf\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@err\endcsname{\def\PYG@bc##1{\setlength{\fboxsep}{0pt}\fcolorbox[rgb]{1.00,0.00,0.00}{1,1,1}{\strut ##1}}}
\expandafter\def\csname PYG@tok@mb\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@ss\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.32,0.47,0.09}{##1}}}
\expandafter\def\csname PYG@tok@sr\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.14,0.33,0.53}{##1}}}
\expandafter\def\csname PYG@tok@mo\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@kd\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@mi\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@kn\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@o\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PYG@tok@kr\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@s\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@kp\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@w\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.73,0.73}{##1}}}
\expandafter\def\csname PYG@tok@kt\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.56,0.13,0.00}{##1}}}
\expandafter\def\csname PYG@tok@sc\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@sb\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@k\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@se\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@sd\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}

\def\PYGZbs{\char`\\}
\def\PYGZus{\char`\_}
\def\PYGZob{\char`\{}
\def\PYGZcb{\char`\}}
\def\PYGZca{\char`\^}
\def\PYGZam{\char`\&}
\def\PYGZlt{\char`\<}
\def\PYGZgt{\char`\>}
\def\PYGZsh{\char`\#}
\def\PYGZpc{\char`\%}
\def\PYGZdl{\char`\$}
\def\PYGZhy{\char`\-}
\def\PYGZsq{\char`\'}
\def\PYGZdq{\char`\"}
\def\PYGZti{\char`\~}
% for compatibility with earlier versions
\def\PYGZat{@}
\def\PYGZlb{[}
\def\PYGZrb{]}
\makeatother

\renewcommand\PYGZsq{\textquotesingle}

\begin{document}

\maketitle
\tableofcontents
\phantomsection\label{cbmpy::doc}


Contents:


\chapter{CBMPy: Installation Guide}
\label{install_doc:cbmpy-reference-guide}\label{install_doc::doc}\label{install_doc:cbmpy-installation-guide}

\section{Support}
\label{install_doc:support}
CBMPy is Open Source software released under the GNU GPL 3 licence (included with the source code)
and is in constant development. All the latest downloads, documentation and development information
is available at \textbf{CBMPy on SourceForge}: \href{http://cbmpy.sourceforge.net}{http://cbmpy.sourceforge.net}.


\section{Python standard library modules}
\label{install_doc:python-standard-library-modules}
CBMPy is developed and tested against Python 2.7.x. The following Python Standard
Library modules are used in CBMPy and should be available as part of any CPython
distribution and not require additional installation:

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{l+s}{\PYGZsq{}}\PYG{l+s}{cPickle}\PYG{l+s}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s}{\PYGZsq{}}\PYG{l+s}{cStringIO}\PYG{l+s}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s}{\PYGZsq{}}\PYG{l+s}{cgi}\PYG{l+s}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s}{\PYGZsq{}}\PYG{l+s}{copy}\PYG{l+s}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s}{\PYGZsq{}}\PYG{l+s}{gc}\PYG{l+s}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s}{\PYGZsq{}}\PYG{l+s}{itertools}\PYG{l+s}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s}{\PYGZsq{}}\PYG{l+s}{locale}\PYG{l+s}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s}{\PYGZsq{}}\PYG{l+s}{math}\PYG{l+s}{\PYGZsq{}}\PYG{p}{,}
\PYG{l+s}{\PYGZsq{}}\PYG{l+s}{multiprocessing}\PYG{l+s}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s}{\PYGZsq{}}\PYG{l+s}{os}\PYG{l+s}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s}{\PYGZsq{}}\PYG{l+s}{pprint}\PYG{l+s}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s}{\PYGZsq{}}\PYG{l+s}{random}\PYG{l+s}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s}{\PYGZsq{}}\PYG{l+s}{re}\PYG{l+s}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s}{\PYGZsq{}}\PYG{l+s}{shutil}\PYG{l+s}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s}{\PYGZsq{}}\PYG{l+s}{subprocess}\PYG{l+s}{\PYGZsq{}}\PYG{p}{,}
\PYG{l+s}{\PYGZsq{}}\PYG{l+s}{time}\PYG{l+s}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s}{\PYGZsq{}}\PYG{l+s}{urllib2}\PYG{l+s}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s}{\PYGZsq{}}\PYG{l+s}{webbrowser}\PYG{l+s}{\PYGZsq{}}\PYG{p}{,} \PYG{l+s}{\PYGZsq{}}\PYG{l+s}{xml}\PYG{l+s}{\PYGZsq{}}
\end{Verbatim}


\section{Required libraries (Python bindings)}
\label{install_doc:required-libraries-python-bindings}
Besides those mentioned above, the following packages are required for CBMPy's
core functionality. Note that it is possible to install CBMPy using only
\emph{numpy} but that only very limited subset of functionality is then available.
CBMPy is primarily developed on Microsoft Windows and Ubuntu Linux and where possible the package
name is provided such that can be used with the software center or package manager
\code{sudo apt-get install \textless{}package\textgreater{}} (please see the man pages for sudo and apt-get
if you don't know what this). A comprehensive list of modules are listed at the end of this
document however I consider these to be the minimum requirements. In the case of
external C/C++ libraries the Python bindings should be installed as well (e.g. libSBML). Many
of these are available in \emph{batteries included} Python distributions.


\section{Installation types: quick reference}
\label{install_doc:installation-types-quick-reference}

\subsection{Minimal}
\label{install_doc:minimal}\begin{itemize}
\item {} 
\textbf{numpy} \href{http://numpy.scipy.org}{http://numpy.scipy.org}

\item {} 
\textbf{libsbml} (+ Python bindings) \href{http://sbml.org/download}{http://sbml.org/download}

\item {} 
\textbf{PyQT4} \href{http://www.riverbankcomputing.com/software/pyqt/download}{http://www.riverbankcomputing.com/software/pyqt/download}

\item {} 
Optimization libraries (one or more of):
- CPLEX (LP, MILP): \href{http:://www.ibm.com}{http:://www.ibm.com}
- GLPK  (LP): \href{http://tfinley.net/software/pyglpk/}{http://tfinley.net/software/pyglpk/}

\end{itemize}


\subsection{Full (highly recommended)}
\label{install_doc:full-highly-recommended}\begin{itemize}
\item {} 
\textbf{xlwt} \href{http://pypi.python.org/pypi/xlwt}{http://pypi.python.org/pypi/xlwt}

\item {} 
\textbf{xlrd} \href{http://pypi.python.org/pypi/xlrd}{http://pypi.python.org/pypi/xlrd}

\item {} 
wxPython

\item {} 
Matplotlib

\item {} 
Sympy

\end{itemize}


\subsection{Complete}
\label{install_doc:complete}
Web services and database:
\begin{itemize}
\item {} 
pysqlite2

\item {} 
suds

\end{itemize}

Advanced functionality:
\begin{itemize}
\item {} 
SciPy

\item {} 
H5Py

\item {} 
NetworkX

\end{itemize}

User tools:
\begin{itemize}
\item {} 
iPython

\item {} 
iPython-notebook

\item {} 
SCiTE

\end{itemize}


\section{Generic installation Windows (XP, 7, 8.1)}
\label{install_doc:generic-installation-windows-xp-7-8-1}
For the modeller that does not want to customize his installation and install
all of the above packages by hand there are some \emph{batteries included} Python
distributions which have many (if not most) of the packages listed above. An
Open Source distributions is \emph{Python(x,y)} available from \href{http://code.google.com/p/pythonxy}{http://code.google.com/p/pythonxy}
Alternatives include commercial distribution such as Anaconda \href{http://continuum.io}{http://continuum.io} and the
Enthought Python Distribution (EPD) \href{http://www.enthought.com}{http://www.enthought.com}

Python(x,y) has a huge number of additional packages in addition to the base
Python distribution, best of all it is Open Source and free for use. First of
all download Python(x,y), I would recommend the
latest \emph{Python 2.7.x} distribution. In addition to the default packages
automatically selected by the installer it is highly recommended to install
either all the additional packages. If not at least select the following
packages from the \emph{Python} branch of the Python(x,y) installation directory:
\begin{itemize}
\item {} 
WxPython

\item {} 
Sympy

\item {} 
NetworkX

\item {} 
xlrd

\item {} 
xlwt

\item {} 
h5py

\item {} 
wxPython

\item {} 
PyQT4

\end{itemize}

You should now have a working Python 2.7.x distribution. Try firing up an
advanced shell like \emph{iPython} and play around and get to grips with the
fantastic, free text editor \emph{SciTE}.


\subsection{Installing CBMPy}
\label{install_doc:installing-cbmpy}
There are two ways to install CBMPy either download the latest release as
source bundle or binary from \href{http://cbmpy.sourceforge.net}{http://cbmpy.sourceforge.net} and unzip or execute from a
a temporary directory (recommended). Or, if you want the latest
(greatest and potentially broken) version grab the latest revision from the
the CBMPy Subversion repository:

\begin{Verbatim}[commandchars=\\\{\}]
svn co http://sourceforge.net/p/cbmpy/code/HEAD/tree/trunk/cbmpy cbmpy
\end{Verbatim}

In both cases you should should now have a directory that contains a file
\emph{setup.py} which can install by simply typing the following into a Windows shell
(command line):

\begin{Verbatim}[commandchars=\\\{\}]
python setup.py build
python setup.py install
\end{Verbatim}


\subsection{Installing libSBML with Python bindings}
\label{install_doc:installing-libsbml-with-python-bindings}
It is highly recommended to install libSBML which CBMPy uses to provide support
for the Systems Biology Markup Language (SBML). First go to the libSBML download
page \href{http://sbml.org/Software/libSBML}{http://sbml.org/Software/libSBML} page follow the \emph{Download libSBML} --\textgreater{} \emph{Stable} --\textgreater{}
\emph{Windows} --\textgreater{} \emph{32bit} path and download libSBML (e.g. libSBML-5.10.0-win-x86.exe). The latest
stable version can be found at \href{http://sbml.org/Software/libSBML}{http://sbml.org/Software/libSBML}
\begin{quote}

\href{http://sourceforge.net/projects/sbml/files/libsbml/5.10.0/stable/Windows/32-bit/libSBML-5.10.0-win-x86.exe/download}{http://sourceforge.net/projects/sbml/files/libsbml/5.10.0/stable/Windows/32-bit/libSBML-5.10.0-win-x86.exe/download}
\end{quote}

Run the installer and make sure you select the Python Bindings during installation
or install the appropriate Python bindings that match your Python(x,y) version directly e.g.
(libSBML-5.10.0-win-py2.7-x86.exe)


\subsection{Optmization (1): IBM cplex optimization studio (Academic)}
\label{install_doc:optmization-1-ibm-cplex-optimization-studio-academic}
If you have access to the the IBM CPLEX solver. It is a a good idea to use the latest available version.
Again choose the appropriate 32 or 64 bit version and an installation path that suites your setup.
\begin{itemize}
\item {} 
Run \textbf{cplex\_studio126.win-x86-32.exe}

\item {} 
Select English language and accept licence

\item {} 
Set ``Program'' install directory to C:\textbackslash{}ILOG\textbackslash{}CPLEX\_Studio126

\item {} 
Allow default associations to be set and PATH update

\end{itemize}

Once installation is complete we need to install the Python bindings
\begin{itemize}
\item {} 
Open a terminal

\item {} 
Execute \code{cd c:\textbackslash{}\textbackslash{}ILOG\textbackslash{}\textbackslash{}CPLEX\_Studio126\textbackslash{}\textbackslash{}cplex\textbackslash{}\textbackslash{}python\textbackslash{}\textbackslash{}x86\_win32}

\item {} 
Execute \code{python setup.py install}

\end{itemize}


\subsection{Optmization (2): GLPK}
\label{install_doc:optmization-2-glpk}
CBMPy 0.7.4 includes support for the free, Open Source GLPK solver. This allows access
to CBMPy's LP functionality (MILP's requires CPLEX). A port of PyGLPK 0.3
is maintained by the OpenCOBRA project which is mirrored here:
\begin{quote}

\href{https://sourceforge.net/projects/cbmpy/files/tools/glpk/}{https://sourceforge.net/projects/cbmpy/files/tools/glpk/}
\end{quote}

Select the binary or source distribution you require and either execute the binary:
\begin{itemize}
\item {} 
Execute \code{glpk-0.3.win32-py2.7.exe}

\end{itemize}


\subsection{Testing your new installation}
\label{install_doc:testing-your-new-installation}
If everything has gone according to plan you can test your installation:
\begin{itemize}
\item {} 
Open a terminal

\item {} 
Execute \code{ipython}

\item {} 
In ipython shell, execute \code{import numpy, h5py, xlrd, xlwt}

\end{itemize}

No import errors should occur.
\begin{itemize}
\item {} 
Execute \code{import libsbml}

\item {} 
Execute \code{libsbml.LIBSBML\_VERSION\_STRING}

\end{itemize}

A successful test should return (for example):

\begin{Verbatim}[commandchars=\\\{\}]
In : libsbml.LIBSBML\PYGZus{}VERSION\PYGZus{}STRING
Out: \PYGZsq{}51000\PYGZsq{}
\end{Verbatim}
\begin{itemize}
\item {} 
Execute \code{import cbmpy as cbm}

\end{itemize}

This should return:

\begin{Verbatim}[commandchars=\\\{\}]
In [1]: import cbmpy as cbm

*******************************************************************
* Welcome to CBMPy (0.7.4) \PYGZhy{} PySCeS Constraint Based Modelling    *
*                http://cbmpy.sourceforge.net                     *
* Copyright(C) Brett G. Olivier 2010 \PYGZhy{} 2015                       *
* Dept. of Systems Bioinformatics                                 *
* Vrije Universiteit Amsterdam, Amsterdam, The Netherlands        *
* CBMPy is distributed under the GNU GPL v 3.0 licence, see       *
* LICENCE (supplied with this release) for details                *
*******************************************************************
\end{Verbatim}

Exit ipython with CTRL-D

If you installed CPLEX then try:
\begin{itemize}
\item {} 
Open a terminal

\item {} 
Execute \code{ipython}

\item {} 
Execute \code{import cplex}

\item {} 
Execute \code{lp = cplex.Cplex()}

\item {} 
Execute \code{lp.solve()}

\end{itemize}

A succesful test should return:

\begin{Verbatim}[commandchars=\\\{\}]
In : lp.solve()
Tried aggregator 1 time.
No LP presolve or aggregator reductions.
Presolve time =    0.00 sec.
\end{Verbatim}

Exit ipython with CTRL-D

If you installed GLPK then try:
\begin{itemize}
\item {} 
Open a terminal

\item {} 
Execute \code{ipython}

\item {} 
Execute \code{import glpk}

\item {} 
Execute \code{lp = glpk.LPX()}

\end{itemize}

A succesful test should return:

\begin{Verbatim}[commandchars=\\\{\}]
In : glpk.LPX()
\PYGZlt{}glpk.LPX 0\PYGZhy{}by\PYGZhy{}0 at 0x036C24C8\PYGZgt{}
\end{Verbatim}

Exit ipython with CTRL-D


\subsection{Install CBMPy (http://cbmpy.sourceforge.net)}
\label{install_doc:install-cbmpy-http-cbmpy-sourceforge-net}
Download the latest version of CBMPy
\begin{itemize}
\item {} 
Run \textbf{cbmpy-0.7.x.win32.exe} (or newer for 32 bit Windows)

\item {} 
Run \textbf{cbmpy-0.7.x.amd64.exe} (or newer for 64 bit Windows)

\end{itemize}

Test installation:
\begin{itemize}
\item {} 
Open a terminal

\item {} 
Execute \code{ipython}

\item {} 
Execute \code{import cbmpy as cbm}

\end{itemize}

This should return:

\begin{Verbatim}[commandchars=\\\{\}]
In [1]: import cbmpy as cbm

*****
Using GLPK
*****

WX GUI tools available.
Qt4 GUI tools available

CBMPy environment
******************
Revision: r346


*******************************************************************
* Welcome to CBMPy (0.7.4) \PYGZhy{} PySCeS Constraint Based Modelling    *
*                http://cbmpy.sourceforge.net                     *
* Copyright(C) Brett G. Olivier 2010 \PYGZhy{} 2015                       *
* Dept. of Systems Bioinformatics                                 *
* Vrije Universiteit Amsterdam, Amsterdam, The Netherlands        *
* CBMPy is distributed under the GNU GPL v 3.0 licence, see       *
* LICENCE (supplied with this release) for details                *
*******************************************************************
\end{Verbatim}

Exit ipython with CTRL-D


\section{Linux: Ubuntu}
\label{install_doc:linux-ubuntu}
On Linux many of the base dependencies are available as packages or from the Python Cheeseshop (\href{http://pypi.python.org/pypi}{http://pypi.python.org/pypi}).
For \textbf{libSBML}, \textbf{CPLEX} and/or \textbf{GLPK} please see the \emph{Generic installation on Microsoft Windows (XP, 7, 2008)} for more details.
For example using \textbf{Ubuntu} the base dependencies can be easily installed (depending on what functionality is required).
If you don't know what these packages are please look them up before installing.

Required:

\begin{Verbatim}[commandchars=\\\{\}]
sudo apt\PYGZhy{}get install python\PYGZhy{}dev python\PYGZhy{}numpy

\PYGZhy{} libSBML for SBML support.
\end{Verbatim}

Please see \href{http://sbml.org/Software/libSBML}{http://sbml.org/Software/libSBML} or try the following. Depending on your configurationyou need to install libxml2, bzip2 and their associated ``dev'' packages:

\begin{Verbatim}[commandchars=\\\{\}]
apt\PYGZhy{}get install libxml2 libxml2\PYGZhy{}dev
apt\PYGZhy{}get install zlib1g zlib1g\PYGZhy{}dev
apt\PYGZhy{}get install bzip2 libbz2\PYGZhy{}dev

easy\PYGZus{}install pip

\PYGZsh{} for standard libSBML
pip install python\PYGZhy{}libsbml

\PYGZsh{} for \PYGZdq{}experimental\PYGZdq{} libSBML (for FBC V2 and Groups support)
pip install python\PYGZhy{}libsbml\PYGZhy{}experimental
\end{Verbatim}
\begin{itemize}
\item {} 
Optimization (at least one of):
\begin{itemize}
\item {} 
IBM CPLEX: \href{http://www.ibm.com}{http://www.ibm.com}

\item {} 
PyGLPK: \href{https://sourceforge.net/projects/cbmpy/files/tools/glpk/}{https://sourceforge.net/projects/cbmpy/files/tools/glpk/}

\end{itemize}

\end{itemize}

Please note that due to changes in the GLPK API the current version of PyGLPK (0.3) \textbf{only supports GLPK up
until version 4.47}. If your system has a newer version of GLPK then the current workaround is to uninstall the newer
version and compile 4.47 from source (also available from the above directory). Dependencies are standard Linux build tools
and GMP etc:

\begin{Verbatim}[commandchars=\\\{\}]
tar xzf glpk\PYGZhy{}4.47.tar.gz
cd glpk\PYGZhy{}4.47
./configure \PYGZhy{}\PYGZhy{}with\PYGZhy{}gmp
make
make check
sudo make install
\end{Verbatim}

Graphical interfaces (highly recommended):

\begin{Verbatim}[commandchars=\\\{\}]
sudo apt\PYGZhy{}get install python\PYGZhy{}wxgtk2.8 python\PYGZhy{}qt4 python\PYGZhy{}matplotlib
\end{Verbatim}

Extended IO (highly recommended):

\begin{Verbatim}[commandchars=\\\{\}]
sudo apt\PYGZhy{}get install python\PYGZhy{}xlrd python\PYGZhy{}xlwt python\PYGZhy{}sympy
\end{Verbatim}

Web services and database:

\begin{Verbatim}[commandchars=\\\{\}]
sudo apt\PYGZhy{}get install python\PYGZhy{}suds python\PYGZhy{}pysqlite2
\end{Verbatim}

Advanced functionality:

\begin{Verbatim}[commandchars=\\\{\}]
sudo apt\PYGZhy{}get install python\PYGZhy{}scipy python\PYGZhy{}h5py python\PYGZhy{}networkx
\end{Verbatim}

User tools (highly recommended):

\begin{Verbatim}[commandchars=\\\{\}]
sudo apt\PYGZhy{}get install ipython ipython\PYGZhy{}notebook scite
\end{Verbatim}


\section{Linux: Ubuntu 14.04}
\label{install_doc:linux-ubuntu-14-04}

\subsection{Python2}
\label{install_doc:python2}
First we create a scientific Python workbench:

\begin{Verbatim}[commandchars=\\\{\}]
sudo apt\PYGZhy{}get install python\PYGZhy{}dev python\PYGZhy{}numpy python\PYGZhy{}scipy
sudo apt\PYGZhy{}get install python\PYGZhy{}matplotlib  python\PYGZhy{}pip
sudo apt\PYGZhy{}get install python\PYGZhy{}sympy python\PYGZhy{}suds python\PYGZhy{}xlrd
sudo apt\PYGZhy{}get install python\PYGZhy{}xlwt python\PYGZhy{}h5py
sudo apt\PYGZhy{}get install python\PYGZhy{}wxgtk2.8 python\PYGZhy{}qt4
sudo apt\PYGZhy{}get install ipython ipython\PYGZhy{}notebook
\end{Verbatim}


\subsection{libSBML}
\label{install_doc:libsbml}
Installing libSBML is now easy using Pip:

\begin{Verbatim}[commandchars=\\\{\}]
sudo apt\PYGZhy{}get install libxml2 libxml2\PYGZhy{}dev
sudo apt\PYGZhy{}get install zlib1g zlib1g\PYGZhy{}dev
sudo apt\PYGZhy{}get install bzip2 libbz2\PYGZhy{}dev

sudo pip install python\PYGZhy{}libsbml
\end{Verbatim}


\subsection{glpk/python-glpk}
\label{install_doc:glpk-python-glpk}
GLPK needs to be version 4.47 to work with glpk-0.3:

\begin{Verbatim}[commandchars=\\\{\}]
sudo apt\PYGZhy{}get install libgmp\PYGZhy{}dev
\end{Verbatim}

cd GLPK source (e.g. glpk-4.47):

\begin{Verbatim}[commandchars=\\\{\}]
./configure \PYGZhy{}\PYGZhy{}with\PYGZhy{}gmp
make
make check
sudo make install
sudo ldconfig
\end{Verbatim}

cd to python-glpk source (glpk-0.3):

\begin{Verbatim}[commandchars=\\\{\}]
make
sudo make install
\end{Verbatim}


\subsection{CBMPy}
\label{install_doc:cbmpy}
Finally, install CBMPy:

\begin{Verbatim}[commandchars=\\\{\}]
python setup.py build sdist
sudo python setup.py install
\end{Verbatim}


\subsection{Installing PyscesMarinerCBM}
\label{install_doc:installing-pyscesmarinercbm}
This will install PySCeS Mariner that adds SOAP web-services
capability to CBMPy. First unpack pyscesmariner-0.7.7.zip and install
the cherrypy webserver:

\begin{Verbatim}[commandchars=\\\{\}]
sudo apt\PYGZhy{}get install python\PYGZhy{}cherrypy
\end{Verbatim}


\subsection{Install soaplib}
\label{install_doc:install-soaplib}
cd \textless{}pysces\_cbm\_mariner\textgreater{}/misc:

\begin{Verbatim}[commandchars=\\\{\}]
tar \PYGZhy{}xf soaplib\PYGZhy{}0.8.1.tar.gz
cd soaplib\PYGZhy{}0.8.1
python setup.py build sdist
sudo python setup.py install
\end{Verbatim}


\subsection{Install Mariner}
\label{install_doc:install-mariner}
cd \textless{}pysces\_cbm\_mariner\textgreater{} and set mariner configuration (not needed for Ubuntu, Windows or if the server does not read SBML):

\begin{Verbatim}[commandchars=\\\{\}]
sudo nano /usr/local/lib/python2.7/dist\PYGZhy{}packages/pyscesmariner/MarinerConfig.py
PATH\PYGZus{}LIBSBMLTHREAD = \PYGZsq{}/usr/local/lib/python2.7/dist\PYGZhy{}packages/pyscesmariner/libSBMLthread.pyc\PYGZsq{}
PATH\PYGZus{}LIBSBML\PYGZus{}CONVERTTHREAD = \PYGZsq{}/usr/local/lib/python2.7/dist\PYGZhy{}packages/pyscesmariner/libSBMLConvertThread.py\PYGZsq{}
\end{Verbatim}

cd to \textless{}pysces\_cbm\_mariner\textgreater{}:

\begin{Verbatim}[commandchars=\\\{\}]
python setup.py build sdist
sudo python setup.py install
\end{Verbatim}


\subsection{Test installation}
\label{install_doc:test-installation}
Open a new terminal window:

\begin{Verbatim}[commandchars=\\\{\}]
\PYGZsh{} cd \PYGZlt{}pysces\PYGZus{}cbm\PYGZus{}mariner\PYGZgt{}/demo
python cbm\PYGZus{}server\PYGZus{}demo.py
\end{Verbatim}

Open another terminal and run the client demo:

\begin{Verbatim}[commandchars=\\\{\}]
python cbm\PYGZus{}client\PYGZus{}demo.py
\end{Verbatim}

Kill the server by closing the terminal window.


\subsection{Python3}
\label{install_doc:python3}
Not all dependencies are available for Python3:

\begin{Verbatim}[commandchars=\\\{\}]
sudo apt\PYGZhy{}get install python3\PYGZhy{}dev python3\PYGZhy{}numpy python3\PYGZhy{}scipy
sudo apt\PYGZhy{}get install python3\PYGZhy{}matplotlib  python3\PYGZhy{}pip
sudo apt\PYGZhy{}get install python3\PYGZhy{}xlrd python3\PYGZhy{}h5py

\PYGZsh{} need to find out what is going on with Python3 and xlwt suds
\PYGZsh{} easy\PYGZus{}install3 sympy ???
\PYGZsh{} wxPython and PyQt4 not in Ubuntu P3 builds yet

sudo apt\PYGZhy{}get install ipython3 ipython3\PYGZhy{}notebook

sudo apt\PYGZhy{}get install libxml2 libxml2\PYGZhy{}dev
sudo apt\PYGZhy{}get install zlib1g zlib1g\PYGZhy{}dev
sudo apt\PYGZhy{}get install bzip2 libbz2\PYGZhy{}dev

sudo pip3 install python\PYGZhy{}libsbml

sudo apt\PYGZhy{}get install python\PYGZhy{}qt4 python\PYGZhy{}qt4\PYGZhy{}dev python\PYGZhy{}sip
sudo apt\PYGZhy{}get install python\PYGZhy{}sip\PYGZhy{}dev build\PYGZhy{}essential
\end{Verbatim}


\section{Apple Macintosh: OS X}
\label{install_doc:apple-macintosh-os-x}
Installation is similar to Linux except packages are installed using distutils and pip. The first step is to install the Mac development tools \code{xcode}

Install \code{Python} packages:

\begin{Verbatim}[commandchars=\\\{\}]
sudo easy\PYGZus{}install numpy ipython scipy matplotlib
sudo easy\PYGZus{}install xlrd xlwt sympy suds pyparsing pip
\end{Verbatim}

Use pip to install advanced Ipython and libsbml:

\begin{Verbatim}[commandchars=\\\{\}]
sudo pip install ipython[notebook]
ARCHFLAGS=\PYGZhy{}Wno\PYGZhy{}error=unused\PYGZhy{}command\PYGZhy{}line\PYGZhy{}argument\PYGZhy{}hard\PYGZhy{}error\PYGZhy{}in\PYGZhy{}future  pip install python\PYGZhy{}libsbml
\end{Verbatim}

For \code{solvers}, either install your own copy of CPLEX or build PyGLPK which requires building both the GMP and GLPK libraries.

\code{GMP} (\href{https://gmplib.org/}{https://gmplib.org/}):

\begin{Verbatim}[commandchars=\\\{\}]
download gmp
./configure \PYGZhy{}\PYGZhy{}prefix=/usr/local
make
make check
sudo make install
\end{Verbatim}

\code{GLPK}  (\href{http://sourceforge.net/projects/cbmpy/files/tools/glpk}{http://sourceforge.net/projects/cbmpy/files/tools/glpk}):

\begin{Verbatim}[commandchars=\\\{\}]
download glpk\PYGZhy{}4.47.tar.gz
./configure \PYGZhy{}\PYGZhy{}prefix=/usr/local \PYGZhy{}\PYGZhy{}with\PYGZhy{}gmp
make
sudo make install
\end{Verbatim}

\code{PyGLPK} (\href{http://sourceforge.net/projects/cbmpy/files/tools/glpk}{http://sourceforge.net/projects/cbmpy/files/tools/glpk}):

\begin{Verbatim}[commandchars=\\\{\}]
download python\PYGZhy{}glpk\PYGZhy{}0.3
python setup.py build
sudo python setup.py install
\end{Verbatim}


\section{Installing PySCeS-CBM Mariner (Microsoft Windows and Linux)}
\label{install_doc:installing-pysces-cbm-mariner-microsoft-windows-and-linux}
The PySCeS Mariner module exposes the CBMPy functionality as SOAP
web services (e.g. as a backend to FAME (\href{http://F-A-M-E.org}{http://F-A-M-E.org})). It is available for download from SourceForge:
\begin{itemize}
\item {} 
PySCeS-CBM Mariner: \href{http://sourceforge.net/projects/cbmpy/files/release/pysces\_mariner/}{http://sourceforge.net/projects/cbmpy/files/release/pysces\_mariner/}

\end{itemize}


\subsection{Dependencies: CherryPy, libXML and SOAPlib}
\label{install_doc:dependencies-cherrypy-libxml-and-soaplib}
PySCeS-CBM Mariner requires (pure python) soaplib 0.8.1 (supplied with it) or
downloadable from:

\begin{Verbatim}[commandchars=\\\{\}]
https//sourceforge.net/projects/cbmpy/files/tools/soaplib/
\end{Verbatim}

Soaplib itself has two dependencies which should be installed first:
\begin{itemize}
\item {} 
LXML (\href{http://lxml.de}{http://lxml.de})
\begin{itemize}
\item {} 
Windows: install with \code{easy\_install lxml}

\item {} 
Linux (Ubuntu) use \code{sudo apt-get install python-lxml}

\end{itemize}

\item {} 
CherryPy (\href{http://www.cherrypy.org}{http://www.cherrypy.org})
\begin{itemize}
\item {} 
Windows: install with \code{easy\_install cherrypy}

\item {} 
Linux (Ubuntu) use \code{sudo apt-get install python-cherrypy}

\end{itemize}

\item {} 
SOAPLIB 0.8.1:
\begin{itemize}
\item {} 
Windows: \code{Execute soaplib-0.8.1.win32.exe}

\item {} 
Linux: Unpack the zip archive and run \code{sudo python setup.py install}

\end{itemize}

\end{itemize}

Test installation:
\begin{itemize}
\item {} 
Open a terminal

\item {} 
Execute ``ipython''

\item {} 
Execute ``import cherrypy, lxml, soaplib'' no errors or warnings should be generated

\item {} 
Exit ipython with CTRL-D

\item {} 
change directory to supplied soaplib tests e.g. ``cd e:\textbackslash{}cbmpy\textbackslash{}tests\textbackslash{}soaplib''

\item {} 
Execute ``python binary\_test.py''

\item {} 
Execute ``python primitive\_test.py''

\end{itemize}

All tests should pass.


\subsection{PySCeS-CBM Mariner (http://cbmpy.sourceforge.net)}
\label{install_doc:pysces-cbm-mariner-http-cbmpy-sourceforge-net}
Download and install the latest version (0.7.4 or newer is required for CBMPy 0.7+):
\begin{itemize}
\item {} 
Windows: \code{Execute pyscesmariner-0.7.7.zip}

\item {} 
Linux: unpack the archive and run \code{sudo python setup.py install}

\end{itemize}

To test installation, on Linux execute the commands in \emph{run\_server.bat} from the terminal directly.
\begin{itemize}
\item {} 
Open two terminals and in both

\item {} 
Change directory to supplied PySCeS-CBM Mariner tests e.g. \code{cd e:\textbackslash{}\textbackslash{}cbmpy\textbackslash{}\textbackslash{}tests\textbackslash{}\textbackslash{}pyscesmariner}

\item {} 
In terminal one \code{Execute run\_server.bat}

\end{itemize}

Which should now display:

\begin{Verbatim}[commandchars=\\\{\}]
E:\PYGZbs{}\PYGZbs{}cbmpy\PYGZbs{}\PYGZbs{}tests\PYGZbs{}\PYGZbs{}pyscesmariner\PYGZgt{}python cbm\PYGZus{}server\PYGZus{}demo.py
Mariner using E:\PYGZbs{}\PYGZbs{}cbmpy\PYGZbs{}\PYGZbs{}tests\PYGZbs{}\PYGZbs{}pyscesmariner as a working directory
Mariner server name: 10.0.2.15
Mariner using port: 31313

Welcome to the PySCeS Constraint Based Modelling Toolkit (0.7.4)

\PYGZlt{}snipped\PYGZgt{}

Multiple Environment Module (0.6.2 [r1147])

PySCeSCBM/Mariner initialising ... this console is now blocked
\end{Verbatim}

In terminal two:
\begin{itemize}
\item {} 
Execute \code{python cbm\_client\_demo.py}

\end{itemize}

This should end without errors and display \code{done.} Congratulations
you have successfully installed CBMPy and PySCeS-CBM Mariner!


\chapter{Introduction}
\label{manual_doc:introduction}\label{manual_doc:introducing-cbmpy}\label{manual_doc::doc}
PySCeS CBMPy is a new platform for constraint based modelling and analysis. It has been designed using principles
developed in the PySCeS simulation software project: usability, flexibility and accessibility. Its architecture
is both extensible and flexible using data structures that are intuitive to the biologist (metabolites, reactions, compartments)
while transparently translating these into the underlying mathematical structures used in advanced analysis (LP's, MILP's).

PySCeS CBMPy implements popular analyses such as FBA, FVA, element/charge balancing, network analysis and model editing as
well as advanced methods developed specifically for the ecosystem modelling: minimal distance methods, flux minimization and input selection.

To cater for a diverse range of modelling needs PySCeS CBMPy supports user interaction via:
\begin{itemize}
\item {} 
interactive console, scripting for advanced use or as a library for software development

\item {} 
GUI, for quick access to a visual representation of the model, analysis methods and annotation tools

\item {} 
SOAP based web services: using the Mariner framework much high level functionality is exposed for integration into web tools

\end{itemize}

For more information on the development and use of PySCeS CBMPy visit the website (\href{http:cbmpy.sourceforge.net}{http:cbmpy.sourceforge.net}) for up to date information and
feel free to contact the development team (\href{mailto:bgoli@users.sourceforge.net}{bgoli@users.sourceforge.net}).


\chapter{CBMPy Module Reference}
\label{modules_doc:cbmpy-module-reference}\label{modules_doc:module-cbmpy.CBCommon}\label{modules_doc::doc}\index{cbmpy.CBCommon (module)}

\section{CBMPy: CBCommon module}
\label{modules_doc:cbmpy-cbcommon-module}
PySCeS Constraint Based Modelling (\href{http://cbmpy.sourceforge.net}{http://cbmpy.sourceforge.net})
Copyright (C) 2009-2015 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see \textless{}\href{http://www.gnu.org/licenses/}{http://www.gnu.org/licenses/}\textgreater{}

Author: Brett G. Olivier
Contact email: \href{mailto:bgoli@users.sourceforge.net}{bgoli@users.sourceforge.net}
Last edit: \$Author: bgoli \$ (\$Id: CBCommon.py 346 2015-08-03 14:09:32Z bgoli \$)
\index{ComboGen (class in cbmpy.CBCommon)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBCommon.ComboGen}\pysigline{\strong{class }\code{cbmpy.CBCommon.}\bfcode{ComboGen}}
Generate sets of unique combinations

\end{fulllineitems}

\index{checkChemFormula() (in module cbmpy.CBCommon)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBCommon.checkChemFormula}\pysiglinewithargsret{\code{cbmpy.CBCommon.}\bfcode{checkChemFormula}}{\emph{cf}, \emph{quiet=False}}{}
Checks whether a string conforms to a Chemical Formula C3Br5 etc, returns True/False. Please see the SBML
Level 3 specification and \href{http://wikipedia.org/wiki/Hill\_system}{http://wikipedia.org/wiki/Hill\_system} for more information.
\begin{itemize}
\item {} 
\emph{cf} a string that contains a formula to check

\item {} 
\emph{quiet} {[}default=False{]} do not print error messages

\end{itemize}

\end{fulllineitems}

\index{checkId() (in module cbmpy.CBCommon)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBCommon.checkId}\pysiglinewithargsret{\code{cbmpy.CBCommon.}\bfcode{checkId}}{\emph{s}}{}
Checks the validity of the string to see if it conforms to a C variable. Returns true/false
\begin{itemize}
\item {} 
\emph{s} a string

\end{itemize}

\end{fulllineitems}

\index{extractGeneIdsFromString() (in module cbmpy.CBCommon)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBCommon.extractGeneIdsFromString}\pysiglinewithargsret{\code{cbmpy.CBCommon.}\bfcode{extractGeneIdsFromString}}{\emph{g}}{}
Extract and return a list of gene names from a gene association string formulation
\begin{itemize}
\item {} 
\emph{g} a COBRA style gene association string

\end{itemize}

\end{fulllineitems}

\index{fixId() (in module cbmpy.CBCommon)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBCommon.fixId}\pysiglinewithargsret{\code{cbmpy.CBCommon.}\bfcode{fixId}}{\emph{s}, \emph{replace=None}}{}
Checks a string (Sid) to see if it is a valid C style variable. first letter must be an underscore or letter,
the rest should be alphanumeric or underscore.
\begin{itemize}
\item {} 
\emph{s} the string to test

\item {} 
\emph{replace} {[}None{]} default is to leave out offensive character, otherwise replace with this one

\end{itemize}

\end{fulllineitems}

\index{parseGeneAssociation() (in module cbmpy.CBCommon)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBCommon.parseGeneAssociation}\pysiglinewithargsret{\code{cbmpy.CBCommon.}\bfcode{parseGeneAssociation}}{\emph{gs}}{}
Parse a COBRA style gene association into a nested list.
\begin{itemize}
\item {} 
\emph{gs} a string containing a gene association

\end{itemize}

\end{fulllineitems}

\index{processSpeciesChargeChemFormulaAnnot() (in module cbmpy.CBCommon)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBCommon.processSpeciesChargeChemFormulaAnnot}\pysiglinewithargsret{\code{cbmpy.CBCommon.}\bfcode{processSpeciesChargeChemFormulaAnnot}}{\emph{s}, \emph{getFromName=False}, \emph{overwriteChemFormula=False}, \emph{overwriteCharge=False}}{}
Disambiguate the chemical formula from either the Notes or the overloaded name
\begin{itemize}
\item {} 
\emph{s} a species object

\item {} 
\emph{getFromName} {[}default=False{]} whether to try strip the chemical formula from the name (old COBRA style)

\item {} 
\emph{overwriteChemFormula} {[}default=False{]}

\item {} 
\emph{overwriteCharge} {[}default=False{]}

\end{itemize}

\end{fulllineitems}

\phantomsection\label{modules_doc:module-cbmpy.CBConfig}\index{cbmpy.CBConfig (module)}

\section{CBMPy: CBConfig module}
\label{modules_doc:cbmpy-cbconfig-module}
PySCeS Constraint Based Modelling (\href{http://cbmpy.sourceforge.net}{http://cbmpy.sourceforge.net})
Copyright (C) 2009-2015 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see \textless{}\href{http://www.gnu.org/licenses/}{http://www.gnu.org/licenses/}\textgreater{}

Author: Brett G. Olivier
Contact email: \href{mailto:bgoli@users.sourceforge.net}{bgoli@users.sourceforge.net}
Last edit: \$Author: bgoli \$ (\$Id: CBConfig.py 404 2016-01-05 15:24:35Z bgoli \$)
\index{current\_version() (in module cbmpy.CBConfig)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBConfig.current_version}\pysiglinewithargsret{\code{cbmpy.CBConfig.}\bfcode{current\_version}}{}{}
Return the current CBMPy version as a string

\end{fulllineitems}

\index{current\_version\_tuple() (in module cbmpy.CBConfig)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBConfig.current_version_tuple}\pysiglinewithargsret{\code{cbmpy.CBConfig.}\bfcode{current\_version\_tuple}}{}{}
Return the current CBMPy version as a tuple (x, y, z)

\end{fulllineitems}

\phantomsection\label{modules_doc:module-cbmpy.CBCPLEX}\index{cbmpy.CBCPLEX (module)}

\section{CBMPy: CBCPLEX module}
\label{modules_doc:cbmpy-cbcplex-module}
PySCeS Constraint Based Modelling (\href{http://cbmpy.sourceforge.net}{http://cbmpy.sourceforge.net})
Copyright (C) 2009-2015 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see \textless{}\href{http://www.gnu.org/licenses/}{http://www.gnu.org/licenses/}\textgreater{}

Author: Brett G. Olivier
Contact email: \href{mailto:bgoli@users.sourceforge.net}{bgoli@users.sourceforge.net}
Last edit: \$Author: bgoli \$ (\$Id: CBCPLEX.py 390 2015-10-05 13:44:45Z bgoli \$)
\index{cplx\_FluxVariabilityAnalysis() (in module cbmpy.CBCPLEX)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBCPLEX.cplx_FluxVariabilityAnalysis}\pysiglinewithargsret{\code{cbmpy.CBCPLEX.}\bfcode{cplx\_FluxVariabilityAnalysis}}{\emph{fba}, \emph{selected\_reactions=None}, \emph{pre\_opt=True}, \emph{tol=None}, \emph{objF2constr=True}, \emph{rhs\_sense='lower'}, \emph{optPercentage=100.0}, \emph{work\_dir=None}, \emph{quiet=True}, \emph{debug=False}, \emph{oldlpgen=False}, \emph{markupmodel=True}, \emph{default\_on\_fail=False}, \emph{roundoff\_span=10}, \emph{method='o'}}{}
Perform a flux variability analysis on an fba model:
\begin{itemize}
\item {} 
\emph{fba} an FBA model object

\item {} 
\emph{selected reactions} {[}default=None{]} means use all reactions otherwise use the reactions listed here

\item {} 
\emph{pre\_opt} {[}default=True{]} attempt to presolve the FBA and report its results in the ouput, if this is disabled and \emph{objF2constr} is True then the rid/value of the current active objective is used

\item {} 
\emph{tol}  {[}default=None{]} do not floor/ceiling the objective function constraint, otherwise round of to \emph{tol}

\item {} 
\emph{rhs\_sense} {[}default='lower'{]} means objC \textgreater{}= objVal the inequality to use for the objective constraint can also be \emph{upper} or \emph{equal}

\item {} 
\emph{optPercentage} {[}default=100.0{]} means the percentage optimal value to use for the RHS of the objective constraint: optimal\_value*(optPercentage/100.0)

\item {} 
\emph{work\_dir} {[}default=None{]} the FVA working directory for temporary files default = cwd+fva

\item {} 
\emph{debug} {[}default=False{]} if True write out all the intermediate FVA LP's into work\_dir

\item {} 
\emph{quiet} {[}default=False{]} if enabled, supress CPLEX output

\item {} 
\emph{objF2constr} {[}default=True{]} add the model objective function as a constraint using rhs\_sense etc. If
this is True with pre\_opt=False then the id/value of the active objective is used to form the constraint

\item {} 
\emph{markupmodel} {[}default=True{]} add the values returned by the fva to the reaction.fva\_min and reaction.fva\_max

\item {} 
\emph{default\_on\_fail} {[}default=False{]} if \emph{pre\_opt} is enabled replace a failed minimum/maximum with the solution value

\item {} 
\emph{roundoff\_span} {[}default=10{]} number of digits is round off (not individual min/max values)

\item {} 
\emph{method} {[}default='o'{]} choose the CPLEX method to use for solution, default is automatic. See CPLEX reference manual for details
\begin{itemize}
\item {} 
`o': auto

\item {} 
`p': primal

\item {} 
`d': dual

\item {} 
`b': barrier (no crossover)

\item {} 
`h': barrier

\item {} 
`s': sifting

\item {} 
`c': concurrent

\end{itemize}

\end{itemize}

Returns an array with columns: Reaction, Reduced Costs, Variability Min, Variability Max, abs(Max-Min), MinStatus, MaxStatus and a list containing the row names.

\end{fulllineitems}

\index{cplx\_MinimizeNumActiveFluxes() (in module cbmpy.CBCPLEX)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBCPLEX.cplx_MinimizeNumActiveFluxes}\pysiglinewithargsret{\code{cbmpy.CBCPLEX.}\bfcode{cplx\_MinimizeNumActiveFluxes}}{\emph{fba}, \emph{selected\_reactions=None}, \emph{pre\_opt=True}, \emph{tol=None}, \emph{objF2constr=True}, \emph{rhs\_sense='lower'}, \emph{optPercentage=100.0}, \emph{work\_dir=None}, \emph{quiet=False}, \emph{debug=False}, \emph{objective\_coefficients=None}, \emph{return\_lp\_obj=False}, \emph{populate=None}, \emph{oldlpgen=False}}{}
Minimize the sum of active fluxes, updates the model with the values of the solution and returns the value
of the MILP objective function (not the model objective function which remains unchanged). If population mode is activated
output is as described below:
\begin{quote}
\begin{description}
\item[{Min: sum(Bi)}] \leavevmode
Bi = 0 -\textgreater{} Ci Ji = 0

\item[{Such that:}] \leavevmode
NJi = 0
Jbio = opt

\item[{where:}] \leavevmode
Binary Bi

\end{description}
\end{quote}

Arguments:
\begin{itemize}
\item {} 
\emph{fba} an FBA model object

\item {} 
\emph{selected reactions} {[}default=None{]} means use all reactions otherwise use the reactions listed here

\item {} 
\emph{pre\_opt} {[}default=True{]} attempt to presolve the FBA and report its results in the ouput, if this is diabled and \emph{objF2constr} is True then the vid/value of the current active objective is used

\item {} 
\emph{tol}  {[}default=None{]} do not floor/ceiling the objective function constraint, otherwise round of to \emph{tol}

\item {} 
\emph{rhs\_sense} {[}default='lower'{]} means objC \textgreater{}= objVal the inequality to use for the objective constraint can also be \emph{upper} or \emph{equal}
Note this does not necessarily mean the upper or lower bound, although practically it will. If in doubt use \emph{equal}

\item {} 
\emph{optPercentage} {[}default=100.0{]} means the percentage optimal value to use for the RHS of the objective constraint: optimal\_value * (optPercentage/100.0)

\item {} 
\emph{work\_dir} {[}default=None{]} the MSAF working directory for temporary files default = cwd+fva

\item {} 
\emph{debug} {[}default=False{]} if True write out all the intermediate MSAF LP's into work\_dir

\item {} 
\emph{quiet} {[}default=False{]} if enabled supress CPLEX output

\item {} 
\emph{objF2constr} {[}default=True{]} add the model objective function as a constraint using rhs\_sense etc. If
this is True with pre\_opt=False then the id/value of the active objective is used to form the constraint

\item {} 
\emph{objective\_coefficients} {[}default=None{]} a dictionary of (reaction\_id : float) pairs that provide the are introduced as objective coefficients to the absolute flux value. Note that the default value of the coefficient (non-specified) is +1.

\item {} 
\emph{return\_lp\_obj} {[}default=False{]} off by default when enabled it returns the CPLEX LP object

\item {} 
\emph{populate} {[}default=None{]} enable search algorithm to find multiple (sub)optimal solutions. Set with a tuple of (RELGAP=0.0, POPULATE\_LIMIT=20, TIME\_LIMIT=300) suggested values only.
- \emph{RELGAP} {[}default=0.0{]} relative gap to optimal solution
- \emph{POPULATE\_LIMIT} {[}default=20{]} terminate when so many solutions have been found
- \emph{TIME\_LIMIT} {[}default=300{]} terminate search after so many seconds important with higher values of \emph{POPULATION\_LIMIT}

\item {} 
\emph{with\_reduced\_costs} {[}default='uncsaled'{]} can be `scaled', `unscaled' or anything else which is None

\end{itemize}

With outputs:
\begin{itemize}
\item {} 
\emph{mincnt} the objective function value OR

\item {} 
\emph{mincnt, cpx} the objective function and cplex model OR

\item {} 
\emph{populate\_data, mincnt} a population data set OR

\item {} 
\emph{populate\_data, mincnt, cpx} both the cps object and population data set

\end{itemize}

depending on selected flags.

\end{fulllineitems}

\index{cplx\_MinimizeSumOfAbsFluxes() (in module cbmpy.CBCPLEX)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBCPLEX.cplx_MinimizeSumOfAbsFluxes}\pysiglinewithargsret{\code{cbmpy.CBCPLEX.}\bfcode{cplx\_MinimizeSumOfAbsFluxes}}{\emph{fba}, \emph{selected\_reactions=None}, \emph{pre\_opt=True}, \emph{tol=None}, \emph{objF2constr=True}, \emph{rhs\_sense='lower'}, \emph{optPercentage=100.0}, \emph{work\_dir=None}, \emph{quiet=False}, \emph{debug=False}, \emph{objective\_coefficients=None}, \emph{return\_lp\_obj=False}, \emph{oldlpgen=False}, \emph{with\_reduced\_costs=None}, \emph{method='o'}}{}
Minimize the sum of absolute fluxes sum(abs(J1) + abs(J2) + abs(J3) ... abs(Jn)) by adding two constraints per flux
and a variable representing the absolute value:
\begin{quote}
\begin{description}
\item[{Min: Ci abs\_Ji}] \leavevmode
Ji - abs\_Ji \textless{}= 0
Ji + abs\_Ji \textgreater{}= 0

\item[{Such that:}] \leavevmode
NJi = 0
Jopt = opt

\end{description}
\end{quote}

returns the value of the flux minimization objective function (not the model objective function which remains unchanged from)

Arguments:
\begin{itemize}
\item {} 
\emph{fba} an FBA model object

\item {} 
\emph{selected reactions} {[}default=None{]} means use all reactions otherwise use the reactions listed here

\item {} 
\emph{pre\_opt} {[}default=True{]} attempt to presolve the FBA and report its results in the ouput, if this is disabled and \emph{objF2constr} is True then the vid/value of the current active objective is used

\item {} 
\emph{tol}  {[}default=None{]} do not floor/ceiling the objective function constraint, otherwise round of to \emph{tol}

\item {} 
\emph{rhs\_sense} {[}default='lower'{]} means objC \textgreater{}= objVal the inequality to use for the objective constraint can also be \emph{upper} or \emph{equal}

\item {} 
\emph{optPercentage} {[}default=100.0{]} means the percentage optimal value to use for the RHS of the objective constraint: optimal\_value*(optPercentage/100.0)

\item {} 
\emph{work\_dir} {[}default=None{]} the MSAF working directory for temporary files default = cwd+fva

\item {} 
\emph{debug} {[}default=False{]} if True write out all the intermediate MSAF LP's into work\_dir

\item {} 
\emph{quiet} {[}default=False{]} if enabled supress CPLEX output

\item {} 
\emph{objF2constr} {[}default=True{]} add the model objective function as a constraint using rhs\_sense etc. If
this is True with pre\_opt=False then the id/value of the active objective is used to form the constraint

\item {} 
\emph{objective\_coefficients} {[}default=None{]} a dictionary of (reaction\_id : float) pairs that provide the are introduced as objective coefficients to the absolute flux value. Note that the default value of the coefficient (non-specified) is +1.

\item {} 
\emph{return\_lp\_obj} {[}default=False{]} off by default when enabled it returns the CPLEX LP object

\item {} 
\emph{with\_reduced\_costs} {[}default=None{]} if not None should be `scaled' or `unscaled'

\item {} 
\emph{method} {[}default='o'{]} choose the CPLEX method to use for solution, default is automatic. See CPLEX reference manual for details
\begin{itemize}
\item {} 
`o': auto

\item {} 
`p': primal

\item {} 
`d': dual

\item {} 
`b': barrier (no crossover)

\item {} 
`h': barrier

\item {} 
`s': sifting

\item {} 
`c': concurrent

\end{itemize}

\end{itemize}

With outputs:
\begin{itemize}
\item {} 
\emph{fba} an update instance of a CBModel. Note that the FBA model objective function value is the original value set as a constraint

\end{itemize}

\end{fulllineitems}

\index{cplx\_MultiFluxVariabilityAnalysis() (in module cbmpy.CBCPLEX)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBCPLEX.cplx_MultiFluxVariabilityAnalysis}\pysiglinewithargsret{\code{cbmpy.CBCPLEX.}\bfcode{cplx\_MultiFluxVariabilityAnalysis}}{\emph{lp}, \emph{selected\_reactions=None}, \emph{tol=1e-10}, \emph{rhs\_sense='lower'}, \emph{optPercentage=100.0}, \emph{work\_dir=None}, \emph{debug=False}}{}
Perform a flux variability analysis on a multistate LP
\begin{itemize}
\item {} 
\emph{lp} a multistate LP

\item {} 
\emph{selected reactions} {[}default=None{]} means use all reactions otherwise use the reactions listed here

\item {} 
\emph{pre\_opt} {[}default=True{]} attempt to presolve the FBA and report its results in the ouput

\item {} 
\emph{tol}  {[}default=1e-10{]} do floor/ceiling the objective function constraint, otherwise floor/ceil to \emph{tol}

\item {} 
\emph{rhs\_sense} {[}default='lower'{]} means objC \textgreater{}= objVal the inequality to use for the objective constraint can also be \emph{upper} or \emph{equal}

\item {} 
\emph{optPercentage} {[}default=100.0{]} means the percentage optimal value to use for the RHS of the objective constraint: optimal\_value*(optPercentage/100.0)

\item {} 
\emph{work\_dir} {[}default=None{]} the FVA working directory for temporary files default = cwd+fva

\item {} 
\emph{debug} {[}default=False{]} if True write out all the intermediate FVA LP's into work\_dir

\item {} 
\emph{bypass} {[}default=False{]} bypass everything and only run the min/max on lp

\end{itemize}

and returns an array with columns:

\begin{Verbatim}[commandchars=\\\{\}]
Reaction, Reduced Costs, Variability Min, Variability Max, abs(Max\PYGZhy{}Min), MinStatus, MaxStatus
\end{Verbatim}

and a list containing the row names.

\end{fulllineitems}

\index{cplx\_SolveMILP() (in module cbmpy.CBCPLEX)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBCPLEX.cplx_SolveMILP}\pysiglinewithargsret{\code{cbmpy.CBCPLEX.}\bfcode{cplx\_SolveMILP}}{\emph{c}, \emph{auto\_mipgap=False}}{}
Solve and MILP
\begin{itemize}
\item {} 
\emph{auto\_mipgap} auto decrease mipgap until mipgap == absmipgap

\end{itemize}

\end{fulllineitems}

\index{cplx\_WriteFVAtoCSV() (in module cbmpy.CBCPLEX)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBCPLEX.cplx_WriteFVAtoCSV}\pysiglinewithargsret{\code{cbmpy.CBCPLEX.}\bfcode{cplx\_WriteFVAtoCSV}}{\emph{pid}, \emph{fva}, \emph{names}, \emph{Dir=None}, \emph{fbaObj=None}}{}
Takes the resuls of a FluxVariabilityAnalysis method and writes it to a nice
csv file. Note this method has been refactored to \emph{CBWrite.WriteFVAtoCSV()}.
\begin{itemize}
\item {} 
\emph{pid} filename\_base for the CSV output

\item {} 
\emph{fva} FluxVariabilityAnalysis() OUTPUT\_ARRAY

\item {} 
\emph{names} FluxVariabilityAnalysis() OUTPUT\_NAMES

\item {} 
\emph{Dir} {[}default=None{]} if set the output directory for the csv files

\item {} 
\emph{fbaObj} {[}default=None{]} if supplied adds extra model information into the output tables

\end{itemize}

\end{fulllineitems}

\index{cplx\_analyzeModel() (in module cbmpy.CBCPLEX)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBCPLEX.cplx_analyzeModel}\pysiglinewithargsret{\code{cbmpy.CBCPLEX.}\bfcode{cplx\_analyzeModel}}{\emph{f}, \emph{lpFname=None}, \emph{return\_lp\_obj=False}, \emph{with\_reduced\_costs='unscaled'}, \emph{with\_sensitivity=False}, \emph{del\_intermediate=False}, \emph{build\_n=True}, \emph{quiet=False}, \emph{oldlpgen=False}, \emph{method='o'}}{}
Optimize a model and add the result of the optimization to the model object
(e.g. \emph{reaction.value}, \emph{objectiveFunction.value}). The stoichiometric
matrix is automatically generated. This is a common function available
in all solver interfaces. By default returns the objective function value
\begin{itemize}
\item {} 
\emph{f} an instantiated PySCeSCBM model object

\item {} 
\emph{lpFname} {[}default=None{]} the name of the intermediate LP file. If not specified no LP file is produced

\item {} 
\emph{return\_lp\_obj} {[}default=False{]} off by default when enabled it returns the CPLEX LP object

\item {} 
\emph{with\_reduced\_costs} {[}default='unscaled'{]} calculate and add reduced cost information to mode this can be: `unscaled' or `scaled'
or anything else which is interpreted as `None'. Scaled means s\_rcost = (r.reduced\_cost*rval)/obj\_value

\item {} 
\emph{with\_sensitivity} {[}default=False{]} add solution sensitivity information (not yet implemented)

\item {} 
\emph{del\_intermediate} {[}default=False{]} redundant except if output file is produced and deleted (not useful)

\item {} 
\emph{build\_n} {[}default=True{]} generate stoichiometry from the reaction network (reactions/reagents/species)

\item {} 
\emph{quiet} {[}default=False{]} suppress cplex output

\item {} 
\emph{method} {[}default='o'{]} choose the CPLEX method to use for solution, default is automatic. See CPLEX reference manual for details
\begin{itemize}
\item {} 
`o': auto

\item {} 
`p': primal

\item {} 
`d': dual

\item {} 
`b': barrier (no crossover)

\item {} 
`h': barrier

\item {} 
`s': sifting

\item {} 
`c': concurrent

\end{itemize}

\end{itemize}

\end{fulllineitems}

\index{cplx\_constructLPfromFBA() (in module cbmpy.CBCPLEX)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBCPLEX.cplx_constructLPfromFBA}\pysiglinewithargsret{\code{cbmpy.CBCPLEX.}\bfcode{cplx\_constructLPfromFBA}}{\emph{fba}, \emph{fname=None}}{}
Create a CPLEX LP in memory.
- \emph{fba} an FBA object
- \emph{fname} optional filename if defined writes out the constructed lp

\end{fulllineitems}

\index{cplx\_fixConSense() (in module cbmpy.CBCPLEX)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBCPLEX.cplx_fixConSense}\pysiglinewithargsret{\code{cbmpy.CBCPLEX.}\bfcode{cplx\_fixConSense}}{\emph{operator}}{}
Fixes the sense of inequality operators, returns corrected sense symbol
\begin{itemize}
\item {} 
\emph{operator} the operator to check

\end{itemize}

\end{fulllineitems}

\index{cplx\_func\_GetCPXandPresolve() (in module cbmpy.CBCPLEX)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBCPLEX.cplx_func_GetCPXandPresolve}\pysiglinewithargsret{\code{cbmpy.CBCPLEX.}\bfcode{cplx\_func\_GetCPXandPresolve}}{\emph{fba}, \emph{pre\_opt}, \emph{objF2constr}, \emph{quiet=False}, \emph{oldlpgen=False}, \emph{with\_reduced\_costs='unscaled'}, \emph{method='o'}}{}
This is a utility function that does a presolve for FVA, MSAF etc. Generates properly formatted
empty objects if pre\_opt == False
\begin{itemize}
\item {} 
\emph{pre\_opt} a boolean

\item {} 
\emph{fba} a CBModel object

\item {} 
\emph{objF2constr} add objective function as constraint

\item {} 
\emph{quiet} {[}default=False{]} supress cplex output

\item {} 
\emph{with\_reduced\_costs} {[}default='unscaled'{]} can be `scaled' or `unscaled'

\item {} 
\emph{method} {[}default='o'{]} choose the CPLEX method to use for solution, default is automatic. See CPLEX reference manual for details
\begin{itemize}
\item {} 
`o': auto

\item {} 
`p': primal

\item {} 
`d': dual

\item {} 
`b': barrier (no crossover)

\item {} 
`h': barrier

\item {} 
`s': sifting

\item {} 
`c': concurrent

\end{itemize}

\end{itemize}

Returns: pre\_sol, pre\_oid, pre\_oval, OPTIMAL\_PRESOLUTION, REDUCED\_COSTS

\end{fulllineitems}

\index{cplx\_func\_SetObjectiveFunctionAsConstraint() (in module cbmpy.CBCPLEX)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBCPLEX.cplx_func_SetObjectiveFunctionAsConstraint}\pysiglinewithargsret{\code{cbmpy.CBCPLEX.}\bfcode{cplx\_func\_SetObjectiveFunctionAsConstraint}}{\emph{cpx}, \emph{rhs\_sense}, \emph{oval}, \emph{tol}, \emph{optPercentage}}{}~\begin{description}
\item[{Take the objective function and ``optimum'' value and add it as a constraint}] \leavevmode\begin{itemize}
\item {} 
\emph{cpx} a cplex object

\item {} 
\emph{oval} the objective value

\item {} 
\emph{tol}  {[}default=None{]} do not floor/ceiling the objective function constraint, otherwise round of to \emph{tol}

\item {} 
\emph{rhs\_sense} {[}default='lower'{]} means objC \textgreater{}= objVal the inequality to use for the objective constraint can also be \emph{upper} or \emph{equal}

\item {} 
\emph{optPercentage} {[}default=100.0{]} means the percentage optimal value to use for the RHS of the objective constraint: optimal\_value*(optPercentage/100.0)

\end{itemize}

\end{description}

\end{fulllineitems}

\index{cplx\_getCPLEXModelFromLP() (in module cbmpy.CBCPLEX)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBCPLEX.cplx_getCPLEXModelFromLP}\pysiglinewithargsret{\code{cbmpy.CBCPLEX.}\bfcode{cplx\_getCPLEXModelFromLP}}{\emph{lptFile}, \emph{Dir=None}}{}
Load a LPT (CPLEX format) file and return a CPLX LP model
\begin{itemize}
\item {} 
\emph{lptfile} an CPLEX LP format file

\item {} 
\emph{Dir} an optional directory

\end{itemize}

\end{fulllineitems}

\index{cplx\_getDualValues() (in module cbmpy.CBCPLEX)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBCPLEX.cplx_getDualValues}\pysiglinewithargsret{\code{cbmpy.CBCPLEX.}\bfcode{cplx\_getDualValues}}{\emph{c}}{}
Get the get the dual values of the solution
\begin{itemize}
\item {} 
\emph{c} a CPLEX LP

\end{itemize}

Output is a dictionary of \{name : value\} pairs

\end{fulllineitems}

\index{cplx\_getModelFromLP() (in module cbmpy.CBCPLEX)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBCPLEX.cplx_getModelFromLP}\pysiglinewithargsret{\code{cbmpy.CBCPLEX.}\bfcode{cplx\_getModelFromLP}}{\emph{lptFile}, \emph{Dir=None}}{}
Load a LPT (CPLEX format) file and return a CPLX LP model
\begin{itemize}
\item {} 
\emph{lptfile} an CPLEX LP format file

\item {} 
\emph{Dir} an optional directory

\end{itemize}

\end{fulllineitems}

\index{cplx\_getModelFromObj() (in module cbmpy.CBCPLEX)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBCPLEX.cplx_getModelFromObj}\pysiglinewithargsret{\code{cbmpy.CBCPLEX.}\bfcode{cplx\_getModelFromObj}}{\emph{fba}}{}
Return a CPLEX object from a FBA model object (via LP file)

\end{fulllineitems}

\index{cplx\_getOptimalSolution() (in module cbmpy.CBCPLEX)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBCPLEX.cplx_getOptimalSolution}\pysiglinewithargsret{\code{cbmpy.CBCPLEX.}\bfcode{cplx\_getOptimalSolution}}{\emph{c}}{}
From a CPLX model extract a tuple of solution, ObjFuncName and ObjFuncVal

\end{fulllineitems}

\index{cplx\_getOptimalSolution2() (in module cbmpy.CBCPLEX)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBCPLEX.cplx_getOptimalSolution2}\pysiglinewithargsret{\code{cbmpy.CBCPLEX.}\bfcode{cplx\_getOptimalSolution2}}{\emph{c}, \emph{names}}{}
From a CPLX model extract a tuple of solution, ObjFuncName and ObjFuncVal

\end{fulllineitems}

\index{cplx\_getReducedCosts() (in module cbmpy.CBCPLEX)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBCPLEX.cplx_getReducedCosts}\pysiglinewithargsret{\code{cbmpy.CBCPLEX.}\bfcode{cplx\_getReducedCosts}}{\emph{c}, \emph{scaled=False}}{}
Extract ReducedCosts from LP and return as a dictionary `Rid' : reduced cost
\begin{itemize}
\item {} 
\emph{c} a cplex LP object

\item {} 
\emph{scaled} scale the reduced cost by the optimal flux value

\end{itemize}

\end{fulllineitems}

\index{cplx\_getSensitivities() (in module cbmpy.CBCPLEX)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBCPLEX.cplx_getSensitivities}\pysiglinewithargsret{\code{cbmpy.CBCPLEX.}\bfcode{cplx\_getSensitivities}}{\emph{c}}{}
Get the sensitivities of each constraint on the objective function with inpt
\begin{itemize}
\item {} 
\emph{c} a CPLEX LP

\end{itemize}

Output is a tuple of bound and objective sensitivities where the objective
sensitivity is described in the CPLEX reference manual as:

\begin{Verbatim}[commandchars=\\\{\}]
... the objective sensitivity shows each variable, its reduced cost and the range over
which its objective function coefficient can vary without forcing a change
in the optimal basis. The current value of each objective coefficient is
also displayed for reference.

\PYGZhy{} *objective coefficient sensitivity* \PYGZob{}flux : (reduced\PYGZus{}cost, lower\PYGZus{}obj\PYGZus{}sensitivity, coeff\PYGZus{}value, upper\PYGZus{}obj\PYGZus{}sensitivity)\PYGZcb{}
\PYGZhy{} *rhs sensitivity* \PYGZob{}constraint : (low, value, high)\PYGZcb{}
\PYGZhy{} *bound sensitivity ranges* \PYGZob{}flux : (lb\PYGZus{}low, lb\PYGZus{}high, ub\PYGZus{}low, ub\PYGZus{}high)\PYGZcb{}
\end{Verbatim}

\end{fulllineitems}

\index{cplx\_getShadowPrices() (in module cbmpy.CBCPLEX)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBCPLEX.cplx_getShadowPrices}\pysiglinewithargsret{\code{cbmpy.CBCPLEX.}\bfcode{cplx\_getShadowPrices}}{\emph{c}}{}
Returns a dictionary of shadow prices containing `Rid' : (lb, rhs, ub)
\begin{itemize}
\item {} 
\emph{c} a cplex LP object

\end{itemize}

\end{fulllineitems}

\index{cplx\_getSolutionStatus() (in module cbmpy.CBCPLEX)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBCPLEX.cplx_getSolutionStatus}\pysiglinewithargsret{\code{cbmpy.CBCPLEX.}\bfcode{cplx\_getSolutionStatus}}{\emph{c}}{}
Returns one of:
\begin{itemize}
\item {} 
\emph{LPS\_OPT}: solution is optimal;

\item {} 
\emph{LPS\_FEAS}: solution is feasible;

\item {} 
\emph{LPS\_INFEAS}: solution is infeasible;

\item {} 
\emph{LPS\_NOFEAS}: problem has no feasible solution;

\item {} 
\emph{LPS\_UNBND}: problem has unbounded solution;

\item {} 
\emph{LPS\_UNDEF}: solution is undefined.

\item {} 
\emph{LPS\_NONE}: no solution

\end{itemize}

\end{fulllineitems}

\index{cplx\_runInputScan() (in module cbmpy.CBCPLEX)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBCPLEX.cplx_runInputScan}\pysiglinewithargsret{\code{cbmpy.CBCPLEX.}\bfcode{cplx\_runInputScan}}{\emph{fba}, \emph{exDict}, \emph{wDir}, \emph{input\_lb=-10.0}, \emph{input\_ub=0.0}, \emph{writeHformat=False}, \emph{rationalLPout=False}}{}
scans all inputs

\end{fulllineitems}

\index{cplx\_setFBAsolutionToModel() (in module cbmpy.CBCPLEX)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBCPLEX.cplx_setFBAsolutionToModel}\pysiglinewithargsret{\code{cbmpy.CBCPLEX.}\bfcode{cplx\_setFBAsolutionToModel}}{\emph{fba}, \emph{lp}, \emph{with\_reduced\_costs='unscaled'}}{}
Sets the FBA solution from a CPLEX solution to an FBA object
\begin{itemize}
\item {} 
\emph{fba} and fba object

\item {} 
\emph{lp} a CPLEX LP object

\item {} 
\emph{with\_reduced\_costs} {[}default='unscaled'{]} calculate and add reduced cost information to mode this can be: `unscaled' or `scaled'
or anything else which is interpreted as None. Scaled is: s\_rcost = (r.reduced\_cost*rval)/obj\_value

\end{itemize}

\end{fulllineitems}

\index{cplx\_setMIPGapTolerance() (in module cbmpy.CBCPLEX)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBCPLEX.cplx_setMIPGapTolerance}\pysiglinewithargsret{\code{cbmpy.CBCPLEX.}\bfcode{cplx\_setMIPGapTolerance}}{\emph{c}, \emph{tol}}{}
Sets the the relative MIP gap tolerance

\end{fulllineitems}

\index{cplx\_setObjective() (in module cbmpy.CBCPLEX)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBCPLEX.cplx_setObjective}\pysiglinewithargsret{\code{cbmpy.CBCPLEX.}\bfcode{cplx\_setObjective}}{\emph{c}, \emph{pid}, \emph{expr=None}, \emph{sense='maximize'}, \emph{reset=True}}{}
Set a new objective function note that there is a major memory leak in
\emph{c.variables.get\_names()} whch is used when reset=True. If this is a problem
use cplx\_setObjective2 which takes \emph{names} as an input:
\begin{itemize}
\item {} 
\emph{c} a CPLEX LP object

\item {} 
\emph{pid} the r\_id of the flux to be optimized

\item {} 
\emph{expr} a list of (coefficient, flux) pairs

\item {} 
\emph{sense} `maximize'/'minimize'

\item {} 
\emph{reset} {[}default=True{]} reset all objective function coefficients to zero

\end{itemize}

\end{fulllineitems}

\index{cplx\_setObjective2() (in module cbmpy.CBCPLEX)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBCPLEX.cplx_setObjective2}\pysiglinewithargsret{\code{cbmpy.CBCPLEX.}\bfcode{cplx\_setObjective2}}{\emph{c}, \emph{pid}, \emph{names}, \emph{expr=None}, \emph{sense='maximize'}, \emph{reset=True}}{}
Set a new objective function. This is a workaround function to avoid the
e is a major memory leak in \emph{c.variables.get\_names()} whch is used
in cplx\_setObjective()  when reset=True.

\end{fulllineitems}

\index{cplx\_setOutputStreams() (in module cbmpy.CBCPLEX)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBCPLEX.cplx_setOutputStreams}\pysiglinewithargsret{\code{cbmpy.CBCPLEX.}\bfcode{cplx\_setOutputStreams}}{\emph{lp}, \emph{mode='default'}}{}
Sets the noise level of the solver, mode can be one of:
\begin{itemize}
\item {} 
\emph{None} silent i.e. no output

\item {} 
\emph{`file'} set solver to silent and output logs to \emph{CPLX\_RESULT\_STREAM\_FILE} cplex\_output.log

\item {} 
\emph{`iostream'} set solver to silent and output logs to \emph{CPLX\_RESULT\_STREAM\_IO} csio

\item {} 
\emph{`default'} or anything else noisy with full output closes STREAM\_IO and STREAM\_FILE (default)

\end{itemize}

\end{fulllineitems}

\index{cplx\_singleGeneScan() (in module cbmpy.CBCPLEX)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBCPLEX.cplx_singleGeneScan}\pysiglinewithargsret{\code{cbmpy.CBCPLEX.}\bfcode{cplx\_singleGeneScan}}{\emph{fba}, \emph{r\_off\_low=0.0}, \emph{r\_off\_upp=0.0}, \emph{optrnd=8}, \emph{altout=False}}{}
Perform a single gene deletion scan
\begin{itemize}
\item {} 
\emph{fba} a model object

\item {} 
\emph{r\_off\_low} the lower bound of a deactivated reaction

\item {} 
\emph{r\_off\_upp} the upper bound of a deactivated reaction

\item {} 
\emph{optrnd} {[}default=8{]} round off the optimal value

\item {} 
\emph{altout} {[}default=False{]} by default return a list of gene:opt pairs, alternatively (True) return an extended result set including gene groups, optima and effect map

\end{itemize}

\end{fulllineitems}

\index{cplx\_writeLPsolution() (in module cbmpy.CBCPLEX)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBCPLEX.cplx_writeLPsolution}\pysiglinewithargsret{\code{cbmpy.CBCPLEX.}\bfcode{cplx\_writeLPsolution}}{\emph{fba\_sol}, \emph{objf\_name}, \emph{fname}, \emph{Dir=None}, \emph{separator='}, \emph{`}}{}
This function writes the optimal solution, produced wth \emph{cplx\_getOptimalSolution} to file
\begin{itemize}
\item {} 
\emph{fba\_sol} a dictionary of Flux : value pairs

\item {} 
\emph{objf\_name} the objective function flux id

\item {} 
\emph{fname} the output filename

\item {} 
\emph{Dir} {[}default=None{]} use directory if not None

\item {} 
\emph{separator} {[}default=','{]} the column separator

\end{itemize}

\end{fulllineitems}

\index{cplx\_writeLPtoLPTfile() (in module cbmpy.CBCPLEX)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBCPLEX.cplx_writeLPtoLPTfile}\pysiglinewithargsret{\code{cbmpy.CBCPLEX.}\bfcode{cplx\_writeLPtoLPTfile}}{\emph{c}, \emph{filename}, \emph{title=None}, \emph{Dir=None}}{}
Write out a CPLEX model as an LP format file

\end{fulllineitems}

\index{getReducedCosts() (in module cbmpy.CBCPLEX)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBCPLEX.getReducedCosts}\pysiglinewithargsret{\code{cbmpy.CBCPLEX.}\bfcode{getReducedCosts}}{\emph{fba}}{}
Get a dictionary of reduced costs for each reaction/flux

\end{fulllineitems}

\index{setReducedCosts() (in module cbmpy.CBCPLEX)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBCPLEX.setReducedCosts}\pysiglinewithargsret{\code{cbmpy.CBCPLEX.}\bfcode{setReducedCosts}}{\emph{fba}, \emph{reduced\_costs}}{}
For each reaction/flux, sets the attribute ``reduced\_cost'' from a dictionary of
reduced costs
\begin{itemize}
\item {} 
\emph{fba} an fba object

\item {} 
\emph{reduced\_costs} a dictionary of \{reaction : value\} pairs

\end{itemize}

\end{fulllineitems}

\phantomsection\label{modules_doc:module-cbmpy.CBDataStruct}\index{cbmpy.CBDataStruct (module)}

\section{CBMPy: CBDataStruct module}
\label{modules_doc:cbmpy-cbdatastruct-module}
PySCeS Constraint Based Modelling (\href{http://cbmpy.sourceforge.net}{http://cbmpy.sourceforge.net})
Copyright (C) 2009-2015 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see \textless{}\href{http://www.gnu.org/licenses/}{http://www.gnu.org/licenses/}\textgreater{}

Author: Brett G. Olivier
Contact email: \href{mailto:bgoli@users.sourceforge.net}{bgoli@users.sourceforge.net}
Last edit: \$Author: bgoli \$ (\$Id: CBDataStruct.py 331 2015-07-01 14:36:41Z bgoli \$)
\index{MIRIAMannotation (class in cbmpy.CBDataStruct)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBDataStruct.MIRIAMannotation}\pysigline{\strong{class }\code{cbmpy.CBDataStruct.}\bfcode{MIRIAMannotation}}
The MIRIAMannotation class MIRIAM annotations: Biological Qualifiers
\index{addIDorgURI() (cbmpy.CBDataStruct.MIRIAMannotation method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBDataStruct.MIRIAMannotation.addIDorgURI}\pysiglinewithargsret{\bfcode{addIDorgURI}}{\emph{qual}, \emph{uri}}{}
Add a URI directly into a qualifier collection:
\begin{itemize}
\item {} 
\emph{qual} a Biomodels biological qualifier e.g. ``is'' ``isEncodedBy''

\item {} 
\emph{uri} the complete identifiers.org uri e.g. \href{http://identifiers.org/chebi/CHEBI:58088}{http://identifiers.org/chebi/CHEBI:58088}

\end{itemize}

\end{fulllineitems}

\index{addMIRIAMannotation() (cbmpy.CBDataStruct.MIRIAMannotation method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBDataStruct.MIRIAMannotation.addMIRIAMannotation}\pysiglinewithargsret{\bfcode{addMIRIAMannotation}}{\emph{qual}, \emph{entity}, \emph{mid}}{}
Add a qualified MIRIAM annotation or entity:
\begin{itemize}
\item {} 
\emph{qual} a Biomodels biological qualifier e.g. ``is'' ``isEncodedBy''

\item {} 
\emph{entity} a MIRIAM resource entity e.g. ``ChEBI''

\item {} 
\emph{mid} the entity id e.g. CHEBI:17158

\end{itemize}

\end{fulllineitems}

\index{checkEntity() (cbmpy.CBDataStruct.MIRIAMannotation method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBDataStruct.MIRIAMannotation.checkEntity}\pysiglinewithargsret{\bfcode{checkEntity}}{\emph{entity}}{}
Check an entity entry, this is a MIRIAM resource name: ``chEBI''. The test is case insensitive and will correct the case
of wrongly capitalised entities automatically. If the entity is not recognised then a list of possible candidates
based on the first letters of the input is displayed.
\begin{itemize}
\item {} 
\emph{entity} a MIRIAM resource entity e.g. ``ChEBI''

\end{itemize}

\end{fulllineitems}

\index{checkEntityPattern() (cbmpy.CBDataStruct.MIRIAMannotation method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBDataStruct.MIRIAMannotation.checkEntityPattern}\pysiglinewithargsret{\bfcode{checkEntityPattern}}{\emph{entity}}{}
For an entity key compile the pattern to a regex, if necessary.
\begin{itemize}
\item {} 
\emph{entity} a MIRIAM resource entity

\end{itemize}

\end{fulllineitems}

\index{checkId() (cbmpy.CBDataStruct.MIRIAMannotation method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBDataStruct.MIRIAMannotation.checkId}\pysiglinewithargsret{\bfcode{checkId}}{\emph{entity}, \emph{mid}}{}
Check that a entity id e.g. CHEBI:17158
\begin{itemize}
\item {} 
\emph{mid} the entity id e.g. CHEBI:17158

\end{itemize}

\end{fulllineitems}

\index{deleteMIRIAMannotation() (cbmpy.CBDataStruct.MIRIAMannotation method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBDataStruct.MIRIAMannotation.deleteMIRIAMannotation}\pysiglinewithargsret{\bfcode{deleteMIRIAMannotation}}{\emph{qual}, \emph{entity}, \emph{mid}}{}
Deletes a qualified MIRIAM annotation or entity:
\begin{itemize}
\item {} 
\emph{qual} a Biomodels biological qualifier e.g. ``is'' ``isEncodedBy''

\item {} 
\emph{entity} a MIRIAM resource entity e.g. ``ChEBI''

\item {} 
\emph{mid} the entity id e.g. CHEBI:17158

\end{itemize}

\end{fulllineitems}

\index{getAllMIRIAMUris() (cbmpy.CBDataStruct.MIRIAMannotation method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBDataStruct.MIRIAMannotation.getAllMIRIAMUris}\pysiglinewithargsret{\bfcode{getAllMIRIAMUris}}{}{}
Return a dictionary of qualifiers that contain ID.org URL'S

\end{fulllineitems}

\index{getAndViewUrisForQualifier() (cbmpy.CBDataStruct.MIRIAMannotation method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBDataStruct.MIRIAMannotation.getAndViewUrisForQualifier}\pysiglinewithargsret{\bfcode{getAndViewUrisForQualifier}}{\emph{qual}}{}
Retrieve all url's associated with qualifier and attempt to open them all in a new browser tab
\begin{itemize}
\item {} 
\emph{qual} the qualifier e.g. ``is'' or ``isEncoded''

\end{itemize}

\end{fulllineitems}

\index{getMIRIAMUrisForQualifier() (cbmpy.CBDataStruct.MIRIAMannotation method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBDataStruct.MIRIAMannotation.getMIRIAMUrisForQualifier}\pysiglinewithargsret{\bfcode{getMIRIAMUrisForQualifier}}{\emph{qual}}{}
Return all list of urls associated with qualifier:
\begin{itemize}
\item {} 
\emph{qual} the qualifier e.g. ``is'' or ``isEncoded''

\end{itemize}

\end{fulllineitems}

\index{viewURL() (cbmpy.CBDataStruct.MIRIAMannotation method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBDataStruct.MIRIAMannotation.viewURL}\pysiglinewithargsret{\bfcode{viewURL}}{\emph{url}}{}
This will try to open the URL in a new tab of the default webbrowser
\begin{itemize}
\item {} 
\emph{url} the url

\end{itemize}

\end{fulllineitems}


\end{fulllineitems}

\index{StructMatrix (class in cbmpy.CBDataStruct)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBDataStruct.StructMatrix}\pysiglinewithargsret{\strong{class }\code{cbmpy.CBDataStruct.}\bfcode{StructMatrix}}{\emph{array}, \emph{ridx}, \emph{cidx}, \emph{row=None}, \emph{col=None}}{}
This class is specifically designed to store structural matrix information
give it an array and row/col index permutations it can generate its own
row/col labels given the label src.
\index{getColsByIdx() (cbmpy.CBDataStruct.StructMatrix method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBDataStruct.StructMatrix.getColsByIdx}\pysiglinewithargsret{\bfcode{getColsByIdx}}{\emph{*args}}{}
Return the columns referenced by index (1,3,5)

\end{fulllineitems}

\index{getColsByName() (cbmpy.CBDataStruct.StructMatrix method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBDataStruct.StructMatrix.getColsByName}\pysiglinewithargsret{\bfcode{getColsByName}}{\emph{*args}}{}
Return the columns referenced by label (`s','x','d')

\end{fulllineitems}

\index{getIndexes() (cbmpy.CBDataStruct.StructMatrix method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBDataStruct.StructMatrix.getIndexes}\pysiglinewithargsret{\bfcode{getIndexes}}{\emph{axis='all'}}{}
Return the matrix indexes ({[}rows{]},{[}cols{]}) where axis='row'/'col'/'all'

\end{fulllineitems}

\index{getLabels() (cbmpy.CBDataStruct.StructMatrix method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBDataStruct.StructMatrix.getLabels}\pysiglinewithargsret{\bfcode{getLabels}}{\emph{axis='all'}}{}
Return the matrix labels ({[}rows{]},{[}cols{]}) where axis='row'/'col'/'all'

\end{fulllineitems}

\index{getRowsByIdx() (cbmpy.CBDataStruct.StructMatrix method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBDataStruct.StructMatrix.getRowsByIdx}\pysiglinewithargsret{\bfcode{getRowsByIdx}}{\emph{*args}}{}
Return the rows referenced by index (1,3,5)

\end{fulllineitems}

\index{getRowsByName() (cbmpy.CBDataStruct.StructMatrix method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBDataStruct.StructMatrix.getRowsByName}\pysiglinewithargsret{\bfcode{getRowsByName}}{\emph{*args}}{}
Return the rows referenced by label (`s','x','d')

\end{fulllineitems}

\index{setCol() (cbmpy.CBDataStruct.StructMatrix method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBDataStruct.StructMatrix.setCol}\pysiglinewithargsret{\bfcode{setCol}}{\emph{src}}{}
Assuming that the col index array is a permutation (full/subset)
of a source label array by supplying that src to setCol
maps the row labels to cidx and creates self.col (col label list)

\end{fulllineitems}

\index{setRow() (cbmpy.CBDataStruct.StructMatrix method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBDataStruct.StructMatrix.setRow}\pysiglinewithargsret{\bfcode{setRow}}{\emph{src}}{}
Assuming that the row index array is a permutation (full/subset)
of a source label array by supplying that source to setRow it
maps the row labels to ridx and creates self.row (row label list)

\end{fulllineitems}


\end{fulllineitems}

\index{StructMatrixLP (class in cbmpy.CBDataStruct)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBDataStruct.StructMatrixLP}\pysiglinewithargsret{\strong{class }\code{cbmpy.CBDataStruct.}\bfcode{StructMatrixLP}}{\emph{array}, \emph{ridx}, \emph{cidx}, \emph{row=None}, \emph{col=None}, \emph{rhs=None}, \emph{operators=None}}{}
Adds some stuff to StructMatrix that makes it LP friendly
\index{getCopy() (cbmpy.CBDataStruct.StructMatrixLP method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBDataStruct.StructMatrixLP.getCopy}\pysiglinewithargsret{\bfcode{getCopy}}{\emph{attr\_str}, \emph{deep=False}}{}
Return a copy of the attribute with name attr\_str. Uses the copy module \emph{copy.copy} or \emph{copy.deepcopy}
\begin{itemize}
\item {} 
\emph{attr\_str} a string of the attribute name: `row', `col'

\item {} 
\emph{deep} {[}default=False{]} try to do a deepcopy. Use with caution see copy module docstring for details

\end{itemize}

\end{fulllineitems}


\end{fulllineitems}

\phantomsection\label{modules_doc:module-cbmpy.CBGUI}\index{cbmpy.CBGUI (module)}

\section{CBMPy: CBGUI module}
\label{modules_doc:cbmpy-cbgui-module}
PySCeS Constraint Based Modelling (\href{http://cbmpy.sourceforge.net}{http://cbmpy.sourceforge.net})
Copyright (C) 2009-2015 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see \textless{}\href{http://www.gnu.org/licenses/}{http://www.gnu.org/licenses/}\textgreater{}

Author: Brett G. Olivier
Contact email: \href{mailto:bgoli@users.sourceforge.net}{bgoli@users.sourceforge.net}
Last edit: \$Author: bgoli \$ (\$Id: CBGUI.py 305 2015-04-23 15:18:31Z bgoli \$)
\index{createReaction() (in module cbmpy.CBGUI)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBGUI.createReaction}\pysiglinewithargsret{\code{cbmpy.CBGUI.}\bfcode{createReaction}}{\emph{mod}}{}
Load the QT4 reaction creator widget
\begin{itemize}
\item {} 
\emph{mod} a PySCeS CBMPy model instance

\end{itemize}

\end{fulllineitems}

\index{loadCBGUI() (in module cbmpy.CBGUI)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBGUI.loadCBGUI}\pysiglinewithargsret{\code{cbmpy.CBGUI.}\bfcode{loadCBGUI}}{\emph{mod}, \emph{version=2}}{}
Load an FBA model instance into the quick editor to view or change basic model properties
\begin{itemize}
\item {} 
\emph{mod} a PySCeS CBMPy model instance

\end{itemize}

\end{fulllineitems}

\index{openFileName() (in module cbmpy.CBGUI)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBGUI.openFileName}\pysiglinewithargsret{\code{cbmpy.CBGUI.}\bfcode{openFileName}}{\emph{work\_dir=None}}{}
Load the QT4 file open selection dialogue
\begin{itemize}
\item {} 
\emph{work\_dir} the optional initial directory

\end{itemize}

\end{fulllineitems}

\index{saveFileName() (in module cbmpy.CBGUI)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBGUI.saveFileName}\pysiglinewithargsret{\code{cbmpy.CBGUI.}\bfcode{saveFileName}}{\emph{work\_dir=None}}{}
Load the QT4 file save selection dialogue
\begin{itemize}
\item {} 
\emph{work\_dir} the optional initial directory

\end{itemize}

\end{fulllineitems}

\phantomsection\label{modules_doc:module-cbmpy.CBModel}\index{cbmpy.CBModel (module)}

\section{CBMPy: CBModel module}
\label{modules_doc:cbmpy-cbmodel-module}
PySCeS Constraint Based Modelling (\href{http://cbmpy.sourceforge.net}{http://cbmpy.sourceforge.net})
Copyright (C) 2009-2014 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see \textless{}\href{http://www.gnu.org/licenses/}{http://www.gnu.org/licenses/}\textgreater{}

Author: Brett G. Olivier
Contact email: \href{mailto:bgoli@users.sourceforge.net}{bgoli@users.sourceforge.net}
Last edit: \$Author: bgoli \$ (\$Id: CBModel.py 416 2016-02-23 16:12:23Z bgoli \$)
\index{Compartment (class in cbmpy.CBModel)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Compartment}\pysiglinewithargsret{\strong{class }\code{cbmpy.CBModel.}\bfcode{Compartment}}{\emph{cid}, \emph{name=None}, \emph{size=1}, \emph{dimensions=3}, \emph{volume=None}}{}
A compartment
\index{containsReactions() (cbmpy.CBModel.Compartment method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Compartment.containsReactions}\pysiglinewithargsret{\bfcode{containsReactions}}{}{}
Lists the species contained in this compartment

\end{fulllineitems}

\index{containsSpecies() (cbmpy.CBModel.Compartment method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Compartment.containsSpecies}\pysiglinewithargsret{\bfcode{containsSpecies}}{}{}
Lists the species contained in this compartment

\end{fulllineitems}

\index{getDimensions() (cbmpy.CBModel.Compartment method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Compartment.getDimensions}\pysiglinewithargsret{\bfcode{getDimensions}}{}{}
Get the compartment dimensions

\end{fulllineitems}

\index{getSize() (cbmpy.CBModel.Compartment method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Compartment.getSize}\pysiglinewithargsret{\bfcode{getSize}}{}{}
Get the compartment size

\end{fulllineitems}

\index{setDimensions() (cbmpy.CBModel.Compartment method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Compartment.setDimensions}\pysiglinewithargsret{\bfcode{setDimensions}}{\emph{dimensions}}{}
Get the compartment dimensions
\begin{itemize}
\item {} 
\emph{dimensions} set the new compartment dimensions

\end{itemize}

\end{fulllineitems}

\index{setSize() (cbmpy.CBModel.Compartment method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Compartment.setSize}\pysiglinewithargsret{\bfcode{setSize}}{\emph{size}}{}
Set the compartment size
\begin{itemize}
\item {} 
\emph{size} the new compartment size

\end{itemize}

\end{fulllineitems}


\end{fulllineitems}

\index{Fbase (class in cbmpy.CBModel)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Fbase}\pysigline{\strong{class }\code{cbmpy.CBModel.}\bfcode{Fbase}}
Base class for CB Model objects
\index{addMIRIAMannotation() (cbmpy.CBModel.Fbase method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Fbase.addMIRIAMannotation}\pysiglinewithargsret{\bfcode{addMIRIAMannotation}}{\emph{qual}, \emph{entity}, \emph{mid}}{}
Add a qualified MIRIAM annotation or entity:
\begin{itemize}
\item {} 
\emph{qual} a Biomodels biological qualifier e.g. ``is'' ``isEncodedBy''

\item {} 
\emph{entity} a MIRIAM resource entity e.g. ``ChEBI''

\item {} 
\emph{mid} the entity id e.g. CHEBI:17158 or fully qualifies url (if only\_qual\_uri)

\end{itemize}

\end{fulllineitems}

\index{addMIRIAMuri() (cbmpy.CBModel.Fbase method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Fbase.addMIRIAMuri}\pysiglinewithargsret{\bfcode{addMIRIAMuri}}{\emph{qual}, \emph{uri}}{}
Add a qualified MIRIAM annotation or entity:
\begin{itemize}
\item {} 
\emph{qual} a Biomodels biological qualifier e.g. ``is'' ``isEncodedBy''

\item {} 
\emph{uri} the fully qualified entity id e.g. \href{http://identifiers.org/chebi/CHEBI:12345}{http://identifiers.org/chebi/CHEBI:12345} (no validity checking is done)

\end{itemize}

\end{fulllineitems}

\index{clone() (cbmpy.CBModel.Fbase method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Fbase.clone}\pysiglinewithargsret{\bfcode{clone}}{}{}
Return a clone of this object. Cloning performs a deepcop on the object which will also clone
any objects that exist as attributes of this object, in other words an independent copy of the
original. If this is not the desired behaviour override this method when subclassing or implement
your own.

\end{fulllineitems}

\index{deleteAnnotation() (cbmpy.CBModel.Fbase method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Fbase.deleteAnnotation}\pysiglinewithargsret{\bfcode{deleteAnnotation}}{\emph{key}}{}
Unsets (deltes) an objects annotation with key
\begin{itemize}
\item {} 
\emph{key} the annotation key

\end{itemize}

\end{fulllineitems}

\index{deleteMIRIAMannotation() (cbmpy.CBModel.Fbase method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Fbase.deleteMIRIAMannotation}\pysiglinewithargsret{\bfcode{deleteMIRIAMannotation}}{\emph{qual}, \emph{entity}, \emph{mid}}{}
Deletes a qualified MIRIAM annotation or entity:
\begin{itemize}
\item {} 
\emph{qual} a Biomodels biological qualifier e.g. ``is'' ``isEncodedBy''

\item {} 
\emph{entity} a MIRIAM resource entity e.g. ``ChEBI''

\item {} 
\emph{mid} the entity id e.g. CHEBI:17158

\end{itemize}

\end{fulllineitems}

\index{getAnnotation() (cbmpy.CBModel.Fbase method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Fbase.getAnnotation}\pysiglinewithargsret{\bfcode{getAnnotation}}{\emph{key}}{}
Return the object annotation associated with:
\begin{itemize}
\item {} 
\emph{key} the annotation key

\end{itemize}

\end{fulllineitems}

\index{getAnnotations() (cbmpy.CBModel.Fbase method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Fbase.getAnnotations}\pysiglinewithargsret{\bfcode{getAnnotations}}{}{}
Return the object annotation dictionary

\end{fulllineitems}

\index{getCompartmentId() (cbmpy.CBModel.Fbase method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Fbase.getCompartmentId}\pysiglinewithargsret{\bfcode{getCompartmentId}}{}{}
Return the compartment id where this element is located

\end{fulllineitems}

\index{getId() (cbmpy.CBModel.Fbase method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Fbase.getId}\pysiglinewithargsret{\bfcode{getId}}{}{}
Return the object ID.

\end{fulllineitems}

\index{getMIRIAMannotations() (cbmpy.CBModel.Fbase method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Fbase.getMIRIAMannotations}\pysiglinewithargsret{\bfcode{getMIRIAMannotations}}{}{}
Returns a dictionary of all MIRIAM annotations associated with this object
or None of there are none defined.

\end{fulllineitems}

\index{getMetaId() (cbmpy.CBModel.Fbase method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Fbase.getMetaId}\pysiglinewithargsret{\bfcode{getMetaId}}{}{}
Return the object metaId.

\end{fulllineitems}

\index{getName() (cbmpy.CBModel.Fbase method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Fbase.getName}\pysiglinewithargsret{\bfcode{getName}}{}{}
Return the object name.

\end{fulllineitems}

\index{getNotes() (cbmpy.CBModel.Fbase method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Fbase.getNotes}\pysiglinewithargsret{\bfcode{getNotes}}{}{}
Return the object's notes

\end{fulllineitems}

\index{getPid() (cbmpy.CBModel.Fbase method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Fbase.getPid}\pysiglinewithargsret{\bfcode{getPid}}{}{}
Return the object ID.

\end{fulllineitems}

\index{getSBOterm() (cbmpy.CBModel.Fbase method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Fbase.getSBOterm}\pysiglinewithargsret{\bfcode{getSBOterm}}{}{}
Return the SBO term for this object.

\end{fulllineitems}

\index{hasAnnotation() (cbmpy.CBModel.Fbase method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Fbase.hasAnnotation}\pysiglinewithargsret{\bfcode{hasAnnotation}}{\emph{key}}{}
Returns a boolean representing the presence/absence of the key in the objext annotation
\begin{itemize}
\item {} 
\emph{key} the annotation key

\end{itemize}

\end{fulllineitems}

\index{serialize() (cbmpy.CBModel.Fbase method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Fbase.serialize}\pysiglinewithargsret{\bfcode{serialize}}{\emph{protocol=0}}{}
Serialize object, returns a string by default
\begin{itemize}
\item {} \begin{description}
\item[{\emph{protocol} {[}default=0{]} serialize to a string or binary if required,}] \leavevmode
see pickle module documentation for details

\end{description}

\end{itemize}

\end{fulllineitems}

\index{serializeToDisk() (cbmpy.CBModel.Fbase method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Fbase.serializeToDisk}\pysiglinewithargsret{\bfcode{serializeToDisk}}{\emph{filename}, \emph{protocol=0}}{}
Serialize to disk using pickle protocol:
\begin{itemize}
\item {} 
\emph{filename} the name of the output file

\item {} \begin{description}
\item[{\emph{protocol} {[}default=0{]} serialize to a string or binary if required,}] \leavevmode
see pickle module documentation for details

\end{description}

\end{itemize}

\end{fulllineitems}

\index{setAnnotation() (cbmpy.CBModel.Fbase method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Fbase.setAnnotation}\pysiglinewithargsret{\bfcode{setAnnotation}}{\emph{key}, \emph{value}}{}
Set an objects annotation as a key : value pair.
\begin{itemize}
\item {} 
\emph{key} the annotation key

\item {} 
\emph{value} the annotation value

\end{itemize}

\end{fulllineitems}

\index{setCompartmentId() (cbmpy.CBModel.Fbase method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Fbase.setCompartmentId}\pysiglinewithargsret{\bfcode{setCompartmentId}}{\emph{compartment}}{}
Set the compartment id where this element is located

\end{fulllineitems}

\index{setId() (cbmpy.CBModel.Fbase method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Fbase.setId}\pysiglinewithargsret{\bfcode{setId}}{\emph{fid}}{}
Sets the object Id
\begin{itemize}
\item {} 
\emph{fid} a valid c variable style id string

\end{itemize}

\end{fulllineitems}

\index{setMetaId() (cbmpy.CBModel.Fbase method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Fbase.setMetaId}\pysiglinewithargsret{\bfcode{setMetaId}}{\emph{mid=None}}{}
Sets the object Id
\begin{itemize}
\item {} 
\emph{mid} {[}default=None{]} a valid c variable style metaid string, if None it will be set as meta+id

\end{itemize}

\end{fulllineitems}

\index{setName() (cbmpy.CBModel.Fbase method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Fbase.setName}\pysiglinewithargsret{\bfcode{setName}}{\emph{name}}{}
Set the object name:
\begin{itemize}
\item {} 
\emph{name} the name string

\end{itemize}

\end{fulllineitems}

\index{setNotes() (cbmpy.CBModel.Fbase method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Fbase.setNotes}\pysiglinewithargsret{\bfcode{setNotes}}{\emph{notes}}{}
Sets the object's notes:
\begin{itemize}
\item {} 
\emph{notes} the note string, should preferably be (X)HTML for SBML

\end{itemize}

\end{fulllineitems}

\index{setPid() (cbmpy.CBModel.Fbase method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Fbase.setPid}\pysiglinewithargsret{\bfcode{setPid}}{\emph{fid}}{}
Sets the object Id
\begin{itemize}
\item {} 
\emph{fid} a valid c variable style id string

\end{itemize}

\end{fulllineitems}

\index{setSBOterm() (cbmpy.CBModel.Fbase method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Fbase.setSBOterm}\pysiglinewithargsret{\bfcode{setSBOterm}}{\emph{sbo}}{}
Set the SBO term for this object.
\begin{itemize}
\item {} 
\emph{sbo} the SBOterm with format: SBO:nnnnnnn''

\end{itemize}

\end{fulllineitems}


\end{fulllineitems}

\index{FluxBound (class in cbmpy.CBModel)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.FluxBound}\pysiglinewithargsret{\strong{class }\code{cbmpy.CBModel.}\bfcode{FluxBound}}{\emph{fid}, \emph{reaction}, \emph{operation}, \emph{value}}{}
A reaction fluxbound
\index{getType() (cbmpy.CBModel.FluxBound method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.FluxBound.getType}\pysiglinewithargsret{\bfcode{getType}}{}{}
Returns the \emph{type} of FluxBound: `lower', `upper', `equality' or None

\end{fulllineitems}

\index{getValue() (cbmpy.CBModel.FluxBound method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.FluxBound.getValue}\pysiglinewithargsret{\bfcode{getValue}}{}{}
Returns the current value of the attribute (input/solution)

\end{fulllineitems}

\index{setReactionId() (cbmpy.CBModel.FluxBound method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.FluxBound.setReactionId}\pysiglinewithargsret{\bfcode{setReactionId}}{\emph{react}}{}
Sets the reaction attribute of the FluxBound

\end{fulllineitems}

\index{setValue() (cbmpy.CBModel.FluxBound method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.FluxBound.setValue}\pysiglinewithargsret{\bfcode{setValue}}{\emph{value}}{}
Sets the attribute `'value'`

\end{fulllineitems}


\end{fulllineitems}

\index{FluxObjective (class in cbmpy.CBModel)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.FluxObjective}\pysiglinewithargsret{\strong{class }\code{cbmpy.CBModel.}\bfcode{FluxObjective}}{\emph{pid}, \emph{reaction}, \emph{coefficient=1}}{}
A weighted flux that appears in an objective function

NOTE: reaction is a string containing a reaction id

\end{fulllineitems}

\index{Gene (class in cbmpy.CBModel)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Gene}\pysiglinewithargsret{\strong{class }\code{cbmpy.CBModel.}\bfcode{Gene}}{\emph{gid}, \emph{label=None}, \emph{active=True}}{}
Contains all the information about a gene (or gene+protein construct depending on your philosophy)

TODO: I will change the whole Gene/GPR structure to a dictionary data structure on the model which should simplify this all significantly.
\index{getLabel() (cbmpy.CBModel.Gene method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Gene.getLabel}\pysiglinewithargsret{\bfcode{getLabel}}{}{}
Returns the gene label

\end{fulllineitems}

\index{isActive() (cbmpy.CBModel.Gene method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Gene.isActive}\pysiglinewithargsret{\bfcode{isActive}}{}{}
Returns whether the gene is active or not

\end{fulllineitems}

\index{resetActivity() (cbmpy.CBModel.Gene method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Gene.resetActivity}\pysiglinewithargsret{\bfcode{resetActivity}}{}{}
Reset the gene to its default activity state

\end{fulllineitems}

\index{setActive() (cbmpy.CBModel.Gene method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Gene.setActive}\pysiglinewithargsret{\bfcode{setActive}}{}{}
Set the gene to be active

\end{fulllineitems}

\index{setInactive() (cbmpy.CBModel.Gene method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Gene.setInactive}\pysiglinewithargsret{\bfcode{setInactive}}{}{}
Set the gene to be inactive

\end{fulllineitems}

\index{setLabel() (cbmpy.CBModel.Gene method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Gene.setLabel}\pysiglinewithargsret{\bfcode{setLabel}}{\emph{label}}{}
Sets the gene label

\end{fulllineitems}


\end{fulllineitems}

\index{GeneProteinAssociation (class in cbmpy.CBModel)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.GeneProteinAssociation}\pysiglinewithargsret{\strong{class }\code{cbmpy.CBModel.}\bfcode{GeneProteinAssociation}}{\emph{gpid}, \emph{protein}}{}
This class associates genes to proteins.
TODO: I will change the whole Gene/GPR structure to a dictionary data structure on the model which should simplify this all significantly.
\index{addAssociation() (cbmpy.CBModel.GeneProteinAssociation method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.GeneProteinAssociation.addAssociation}\pysiglinewithargsret{\bfcode{addAssociation}}{\emph{assoc}}{}
Add a gene/protein association expression

\end{fulllineitems}

\index{addGeneref() (cbmpy.CBModel.GeneProteinAssociation method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.GeneProteinAssociation.addGeneref}\pysiglinewithargsret{\bfcode{addGeneref}}{\emph{geneid}}{}
Add a gene reference to the list of gene references
\begin{itemize}
\item {} 
\emph{geneid} a valid model Gene id

\end{itemize}

\end{fulllineitems}

\index{buildEvalFunc() (cbmpy.CBModel.GeneProteinAssociation method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.GeneProteinAssociation.buildEvalFunc}\pysiglinewithargsret{\bfcode{buildEvalFunc}}{}{}
Builds a function which evaluates the gene expressions and evaluates to an integer uisng
the following rules:
\begin{itemize}
\item {} 
True --\textgreater{} 1

\item {} 
False --\textgreater{} 0

\item {} 
and --\textgreater{} *

\item {} 
or --\textgreater{} +

\end{itemize}

\end{fulllineitems}

\index{createAssociationAndGeneRefs() (cbmpy.CBModel.GeneProteinAssociation method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.GeneProteinAssociation.createAssociationAndGeneRefs}\pysiglinewithargsret{\bfcode{createAssociationAndGeneRefs}}{\emph{assoc}, \emph{altlabels=None}}{}
Evaluate the gene/protein association and add the genes necessary to evaluate it
Note that this GPR should be added to a model with cmod.addGPRAssociation() before calling this method
\begin{itemize}
\item {} 
\emph{assoc} the COBRA style gene protein association

\item {} 
\emph{altlabels} {[}default=None{]} a dictionary containing a label\textless{}--\textgreater{}id mapping

\end{itemize}

\end{fulllineitems}

\index{deleteGeneref() (cbmpy.CBModel.GeneProteinAssociation method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.GeneProteinAssociation.deleteGeneref}\pysiglinewithargsret{\bfcode{deleteGeneref}}{\emph{gid}}{}
Deletes a gene reference
\begin{itemize}
\item {} 
\emph{geneid} a valid model Gene id

\end{itemize}

\end{fulllineitems}

\index{evalAssociation() (cbmpy.CBModel.GeneProteinAssociation method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.GeneProteinAssociation.evalAssociation}\pysiglinewithargsret{\bfcode{evalAssociation}}{}{}
Returns an integer value representing the logical associations or None.

\end{fulllineitems}

\index{getActiveGenes() (cbmpy.CBModel.GeneProteinAssociation method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.GeneProteinAssociation.getActiveGenes}\pysiglinewithargsret{\bfcode{getActiveGenes}}{}{}
Return a list of active gene objects

\end{fulllineitems}

\index{getAssociationStr() (cbmpy.CBModel.GeneProteinAssociation method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.GeneProteinAssociation.getAssociationStr}\pysiglinewithargsret{\bfcode{getAssociationStr}}{}{}
return the gene association string

\end{fulllineitems}

\index{getGene() (cbmpy.CBModel.GeneProteinAssociation method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.GeneProteinAssociation.getGene}\pysiglinewithargsret{\bfcode{getGene}}{\emph{gid}}{}
Return a gene object with id

\end{fulllineitems}

\index{getGeneIds() (cbmpy.CBModel.GeneProteinAssociation method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.GeneProteinAssociation.getGeneIds}\pysiglinewithargsret{\bfcode{getGeneIds}}{}{}
Return a list of gene id's

\end{fulllineitems}

\index{getGenes() (cbmpy.CBModel.GeneProteinAssociation method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.GeneProteinAssociation.getGenes}\pysiglinewithargsret{\bfcode{getGenes}}{}{}
Return a list of gene objects associated with this GPRass

\end{fulllineitems}

\index{getProtein() (cbmpy.CBModel.GeneProteinAssociation method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.GeneProteinAssociation.getProtein}\pysiglinewithargsret{\bfcode{getProtein}}{}{}
Return the protein associated with this set of genes

\end{fulllineitems}

\index{isProteinActive() (cbmpy.CBModel.GeneProteinAssociation method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.GeneProteinAssociation.isProteinActive}\pysiglinewithargsret{\bfcode{isProteinActive}}{}{}
This returns a boolean which indicates the result of evaluating the gene association. If the result is positive
then the protein is expressed and \emph{True} is returned, otherwise if the expression evaluates to a value of 0 then
the protein is not expressed and  \emph{False} is returned.

\end{fulllineitems}

\index{setAllGenesActive() (cbmpy.CBModel.GeneProteinAssociation method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.GeneProteinAssociation.setAllGenesActive}\pysiglinewithargsret{\bfcode{setAllGenesActive}}{}{}
Activate all genes in association

\end{fulllineitems}

\index{setAllGenesInactive() (cbmpy.CBModel.GeneProteinAssociation method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.GeneProteinAssociation.setAllGenesInactive}\pysiglinewithargsret{\bfcode{setAllGenesInactive}}{}{}
Deactivates all genes in association

\end{fulllineitems}

\index{setGeneActive() (cbmpy.CBModel.GeneProteinAssociation method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.GeneProteinAssociation.setGeneActive}\pysiglinewithargsret{\bfcode{setGeneActive}}{\emph{gid}}{}
Set a gene to be inactive

\end{fulllineitems}

\index{setGeneInactive() (cbmpy.CBModel.GeneProteinAssociation method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.GeneProteinAssociation.setGeneInactive}\pysiglinewithargsret{\bfcode{setGeneInactive}}{\emph{gid}}{}
Set a gene to be inactive

\end{fulllineitems}


\end{fulllineitems}

\index{Group (class in cbmpy.CBModel)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Group}\pysiglinewithargsret{\strong{class }\code{cbmpy.CBModel.}\bfcode{Group}}{\emph{pid}}{}
Container for SBML groups
\index{addMember() (cbmpy.CBModel.Group method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Group.addMember}\pysiglinewithargsret{\bfcode{addMember}}{\emph{obj}}{}
Add member CBMPy object(s) to the group
\begin{itemize}
\item {} 
\emph{obj} either a single, tuple or list of CBMPy objects

\end{itemize}

\end{fulllineitems}

\index{addSharedMIRIAMannotation() (cbmpy.CBModel.Group method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Group.addSharedMIRIAMannotation}\pysiglinewithargsret{\bfcode{addSharedMIRIAMannotation}}{\emph{qual}, \emph{entity}, \emph{mid}}{}
Add a qualified MIRIAM annotation or entity to the list of members (all) rather than the group itself:
\begin{itemize}
\item {} 
\emph{qual} a Biomodels biological qualifier e.g. ``is'' ``isEncodedBy''

\item {} 
\emph{entity} a MIRIAM resource entity e.g. ``ChEBI''

\item {} 
\emph{mid} the entity id e.g. CHEBI:17158 or fully qualifies url (if only\_qual\_uri)

\end{itemize}

\end{fulllineitems}

\index{assignAllSharedPropertiesToMembers() (cbmpy.CBModel.Group method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Group.assignAllSharedPropertiesToMembers}\pysiglinewithargsret{\bfcode{assignAllSharedPropertiesToMembers}}{\emph{overwrite=False}}{}
Assigns all group shared properties (notes, annotations, MIRIAM annotations, SBO) to the group members.
\begin{itemize}
\item {} 
\emph{overwrite} {[}default=False{]} overwrite the target notes if they are defined

\end{itemize}

\end{fulllineitems}

\index{assignSharedAnnotationToMembers() (cbmpy.CBModel.Group method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Group.assignSharedAnnotationToMembers}\pysiglinewithargsret{\bfcode{assignSharedAnnotationToMembers}}{}{}
This function merges or updates the group member objects annotations with the group shared annotation.

\end{fulllineitems}

\index{assignSharedMIRIAMannotationToMembers() (cbmpy.CBModel.Group method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Group.assignSharedMIRIAMannotationToMembers}\pysiglinewithargsret{\bfcode{assignSharedMIRIAMannotationToMembers}}{}{}
This function merges or updates the group member objects MIRIAM annotations with the group shared MIRIAM annotation.

\end{fulllineitems}

\index{assignSharedNotesToMembers() (cbmpy.CBModel.Group method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Group.assignSharedNotesToMembers}\pysiglinewithargsret{\bfcode{assignSharedNotesToMembers}}{\emph{overwrite=False}}{}
Assigns the group shared notes to the group members.
\begin{itemize}
\item {} 
\emph{overwrite} {[}default=False{]} overwrite the target notes if they are defined

\end{itemize}

\end{fulllineitems}

\index{assignSharedSBOtermsToMembers() (cbmpy.CBModel.Group method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Group.assignSharedSBOtermsToMembers}\pysiglinewithargsret{\bfcode{assignSharedSBOtermsToMembers}}{\emph{overwrite=False}}{}
Assigns the group shared member SBO term to the group members.
\begin{itemize}
\item {} 
\emph{overwrite} {[}default=False{]} overwrite the target SBO term if it is defined

\end{itemize}

\end{fulllineitems}

\index{clone() (cbmpy.CBModel.Group method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Group.clone}\pysiglinewithargsret{\bfcode{clone}}{}{}
Return a clone of this object. Note the for Groups this is a shallow copy, in that the reference
objects themselves are not cloned only the group (and attributes)

\end{fulllineitems}

\index{deleteMember() (cbmpy.CBModel.Group method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Group.deleteMember}\pysiglinewithargsret{\bfcode{deleteMember}}{\emph{oid}}{}
Deletes a group member with group id.
\begin{itemize}
\item {} 
\emph{oid} group member id

\end{itemize}

\end{fulllineitems}

\index{getKind() (cbmpy.CBModel.Group method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Group.getKind}\pysiglinewithargsret{\bfcode{getKind}}{}{}
Return the group kind

\end{fulllineitems}

\index{getMemberIDs() (cbmpy.CBModel.Group method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Group.getMemberIDs}\pysiglinewithargsret{\bfcode{getMemberIDs}}{\emph{as\_set=False}}{}
Return the ids of the member objects.
\begin{itemize}
\item {} 
\emph{as\_set} return id's as a set rather than a list

\end{itemize}

\end{fulllineitems}

\index{getMembers() (cbmpy.CBModel.Group method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Group.getMembers}\pysiglinewithargsret{\bfcode{getMembers}}{\emph{as\_set=False}}{}
Return the member objects of the group.
\begin{itemize}
\item {} 
\emph{as\_set} return objects as a set rather than a list

\end{itemize}

\end{fulllineitems}

\index{getSharedAnnotations() (cbmpy.CBModel.Group method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Group.getSharedAnnotations}\pysiglinewithargsret{\bfcode{getSharedAnnotations}}{}{}
Return a dictionary of the shared member annotations (rather than the group attribute).

\end{fulllineitems}

\index{getSharedMIRIAMannotations() (cbmpy.CBModel.Group method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Group.getSharedMIRIAMannotations}\pysiglinewithargsret{\bfcode{getSharedMIRIAMannotations}}{}{}
Return a dictionary of the shared member MIRIAM annotations (rather than the group attribute).

\end{fulllineitems}

\index{getSharedNotes() (cbmpy.CBModel.Group method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Group.getSharedNotes}\pysiglinewithargsret{\bfcode{getSharedNotes}}{}{}
Return the shared member notes (rather than the group attribute).

\end{fulllineitems}

\index{getSharedSBOterm() (cbmpy.CBModel.Group method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Group.getSharedSBOterm}\pysiglinewithargsret{\bfcode{getSharedSBOterm}}{}{}
Return the shared member SBO term (rather than the group attribute).

\end{fulllineitems}

\index{setKind() (cbmpy.CBModel.Group method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Group.setKind}\pysiglinewithargsret{\bfcode{setKind}}{\emph{kind}}{}
Sets the kind or type of the group, this must be one of: `collection', `partonomy', `classification'.
\begin{itemize}
\item {} 
\emph{kind} the kind

\end{itemize}

\end{fulllineitems}

\index{setSharedAnnotation() (cbmpy.CBModel.Group method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Group.setSharedAnnotation}\pysiglinewithargsret{\bfcode{setSharedAnnotation}}{\emph{key}, \emph{value}}{}
Sets the list of members (all) annotation as a key : value pair.
\begin{itemize}
\item {} 
\emph{key} the annotation key

\item {} 
\emph{value} the annotation value

\end{itemize}

\end{fulllineitems}

\index{setSharedNotes() (cbmpy.CBModel.Group method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Group.setSharedNotes}\pysiglinewithargsret{\bfcode{setSharedNotes}}{\emph{notes}}{}
Sets the group of objects notes attribute (all):
\begin{itemize}
\item {} 
\emph{notes} the note string, should preferably be (X)HTML for SBML

\end{itemize}

\end{fulllineitems}

\index{setSharedSBOterm() (cbmpy.CBModel.Group method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Group.setSharedSBOterm}\pysiglinewithargsret{\bfcode{setSharedSBOterm}}{\emph{sbo}}{}
Set the SBO term for the the members of the group (all).
\begin{itemize}
\item {} 
\emph{sbo} the SBOterm with format: ``SBO:\textless{}7 digit integer\textgreater{}''

\end{itemize}

\end{fulllineitems}


\end{fulllineitems}

\index{GroupMemberAttributes (class in cbmpy.CBModel)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.GroupMemberAttributes}\pysigline{\strong{class }\code{cbmpy.CBModel.}\bfcode{GroupMemberAttributes}}
Contains the shared attributes of the group members (equivalent to SBML annotation on ListOfMembers)

\end{fulllineitems}

\index{Model (class in cbmpy.CBModel)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model}\pysiglinewithargsret{\strong{class }\code{cbmpy.CBModel.}\bfcode{Model}}{\emph{pid}}{}
Container for constraint based model, adds methods for manipulating:
\begin{itemize}
\item {} 
objectives

\item {} 
constraints

\item {} 
reactions

\item {} 
species

\item {} 
compartments

\item {} 
groups

\item {} 
parameters

\item {} 
N a structmatrix object

\end{itemize}
\index{addCompartment() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.addCompartment}\pysiglinewithargsret{\bfcode{addCompartment}}{\emph{comp}}{}
Add an instantiated Compartment object to the CBM model
\begin{itemize}
\item {} 
\emph{comp} an instance of the Compartment class

\end{itemize}

\end{fulllineitems}

\index{addFluxBound() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.addFluxBound}\pysiglinewithargsret{\bfcode{addFluxBound}}{\emph{fluxbound}, \emph{fbexists=None}}{}
Add an instantiated FluxBound object to the FBA model
\begin{itemize}
\item {} 
\emph{fluxbound} an instance of the FluxBound class

\end{itemize}

\end{fulllineitems}

\index{addGPRAssociation() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.addGPRAssociation}\pysiglinewithargsret{\bfcode{addGPRAssociation}}{\emph{gpr}, \emph{update\_idx=True}}{}
Add a GeneProteinAssociation instance to the model
\begin{itemize}
\item {} 
\emph{gpr} an instantiated GeneProteinAssociation object

\end{itemize}

\end{fulllineitems}

\index{addGene() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.addGene}\pysiglinewithargsret{\bfcode{addGene}}{\emph{gene}}{}
Add an instantiated Gene object to the FBA model
\begin{itemize}
\item {} 
\emph{gene} an instance of the G class

\end{itemize}

\end{fulllineitems}

\index{addGroup() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.addGroup}\pysiglinewithargsret{\bfcode{addGroup}}{\emph{obj}}{}
Add an instantiated group object to the model
\begin{itemize}
\item {} 
\emph{obj} the Group instance

\end{itemize}

\end{fulllineitems}

\index{addMIRIAMannotation() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.addMIRIAMannotation}\pysiglinewithargsret{\bfcode{addMIRIAMannotation}}{\emph{qual}, \emph{entity}, \emph{mid}}{}
Add a qualified MIRIAM annotation or entity:
\begin{itemize}
\item {} 
\emph{qual} a Biomodels biological qualifier e.g. ``is'' ``isEncodedBy''

\item {} 
\emph{entity} a MIRIAM resource entity e.g. ``ChEBI''

\item {} 
\emph{mid} the entity id e.g. CHEBI:17158

\end{itemize}

\end{fulllineitems}

\index{addModelCreator() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.addModelCreator}\pysiglinewithargsret{\bfcode{addModelCreator}}{\emph{firstname}, \emph{lastname}, \emph{organisation=None}, \emph{email=None}}{}
Add a model creator to the list of model creators, only the first and fmaily names are mandatory:
\begin{itemize}
\item {} 
\emph{firstname}

\item {} 
\emph{lastname}

\item {} 
\emph{organisation} {[}default=None{]}

\item {} 
\emph{email}  {[}default=None{]}

\end{itemize}

\end{fulllineitems}

\index{addObjective() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.addObjective}\pysiglinewithargsret{\bfcode{addObjective}}{\emph{obj}, \emph{active=False}}{}
Add an instantiated Objective object to the FBA model
\begin{itemize}
\item {} 
\emph{obj} an instance of the Objective class

\item {} 
\emph{active} {[}default=False{]} flag this objective as the active objective (fba.activeObjIdx)

\end{itemize}

\end{fulllineitems}

\index{addParameter() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.addParameter}\pysiglinewithargsret{\bfcode{addParameter}}{\emph{par}}{}
Add an instantiated Parameter object to the model
\begin{itemize}
\item {} 
\emph{par} an instance of the Parameter class

\end{itemize}

\end{fulllineitems}

\index{addReaction() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.addReaction}\pysiglinewithargsret{\bfcode{addReaction}}{\emph{reaction}}{}
Adds a reaction object to the model
\begin{itemize}
\item {} 
\emph{reaction} an instance of the Reaction class

\end{itemize}

\end{fulllineitems}

\index{addSpecies() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.addSpecies}\pysiglinewithargsret{\bfcode{addSpecies}}{\emph{species}}{}
Add an instantiated Species object to the FBA model
\begin{itemize}
\item {} 
\emph{species} an instance of the Species class

\end{itemize}

\end{fulllineitems}

\index{addUserConstraint() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.addUserConstraint}\pysiglinewithargsret{\bfcode{addUserConstraint}}{\emph{pid}, \emph{fluxes=None}, \emph{operator='='}, \emph{rhs=0.0}}{}
Add a user defined constraint to FBA model, this is additional to the automatically determined Stoichiometric constraints.
\begin{itemize}
\item {} 
\emph{pid} user constraint name/id, use \emph{None} for auto-assign

\item {} 
\emph{fluxes} a list of (coefficient, reaction id) pairs where coefficient is a float

\item {} 
\emph{operator} is one of = \textgreater{} \textless{} \textgreater{}= \textless{}=

\item {} 
\emph{rhs} a float

\end{itemize}

\end{fulllineitems}

\index{buildStoichMatrix() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.buildStoichMatrix}\pysiglinewithargsret{\bfcode{buildStoichMatrix}}{\emph{matrix\_type='numpy'}, \emph{only\_return=False}}{}
Build the stoichiometric matrix N and additional constraint matrix CN (if required)
\begin{itemize}
\item {} 
\emph{matrix\_type} {[}default='numpy'{]} the type of matrix to use to generate constraints
\begin{itemize}
\item {} 
\emph{numpy} a NumPy matrix default

\item {} 
\emph{sympy} a SymPy symbolic matrix, if available note the denominator limit can be set in \code{CBModel.\_\_CBCONFIG\_\_{[}'SYMPY\_DENOM\_LIMIT'{]} = 10**12}

\item {} 
\emph{scipy\_csr} create using NumPy but store as SciPy csr\_sparse

\end{itemize}

\end{itemize}
\begin{itemize}
\item {} 
\emph{only\_return} {[}default=False{]} \textbf{IMPORTANT} only returns the stoichiometric matrix and constraint matrix (if required),
does not update the model

\end{itemize}

\end{fulllineitems}

\index{changeAllFluxBoundsWithValue() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.changeAllFluxBoundsWithValue}\pysiglinewithargsret{\bfcode{changeAllFluxBoundsWithValue}}{\emph{old}, \emph{new}}{}
Replaces all flux bounds with value ``old'' with a new value ``new'':
\begin{itemize}
\item {} 
\emph{old} value

\item {} 
\emph{new} value

\end{itemize}

\end{fulllineitems}

\index{clone() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.clone}\pysiglinewithargsret{\bfcode{clone}}{}{}
Return a clone of this object.

\end{fulllineitems}

\index{createCompartment() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.createCompartment}\pysiglinewithargsret{\bfcode{createCompartment}}{\emph{cid}, \emph{name=None}, \emph{size=1}, \emph{dimensions=3}, \emph{volume=None}}{}
Create a new compartment and add it to the model if the id does not exist
\begin{itemize}
\item {} 
\emph{cid} compartment id

\item {} 
\emph{name} {[}None{]} compartment name

\item {} 
\emph{size} {[}1{]} compartment size

\item {} 
\emph{dimensions} {[}3{]} compartment size dimensions

\item {} 
\emph{volume} {[}None{]} compartment volume

\end{itemize}

\end{fulllineitems}

\index{createGeneAssociationsFromAnnotations() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.createGeneAssociationsFromAnnotations}\pysiglinewithargsret{\bfcode{createGeneAssociationsFromAnnotations}}{\emph{annotation\_key='GENE ASSOCIATION'}, \emph{replace\_existing=True}}{}
Add genes to the model using the definitions stored in the annotation key. If this fails it tries some standard annotation
keys: GENE ASSOCIATION, GENE\_ASSOCIATION, gene\_association, gene association.
\begin{itemize}
\item {} 
\emph{annotation\_key} the annotation dictionary key that holds the gene association for the protein/enzyme

\item {} 
\emph{replace\_existing} {[}default=True{]} replace existing annotations, otherwise only new ones are added

\end{itemize}

\end{fulllineitems}

\index{createGeneProteinAssociation() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.createGeneProteinAssociation}\pysiglinewithargsret{\bfcode{createGeneProteinAssociation}}{\emph{protein}, \emph{assoc}, \emph{gid=None}, \emph{name=None}, \emph{gene\_pattern=None}, \emph{update\_idx=True}, \emph{altlabels=None}}{}
Create and add a gene protein relationship to the model, note genes are mapped on protein objects which may or may not be reactions
\begin{itemize}
\item {} 
\emph{protein} in this case the reaction

\item {} 
\emph{assoc} the COBRA style gene protein association

\item {} 
\emph{gid} the unique id

\item {} 
\emph{name} the optional name

\item {} 
\emph{gene\_pattern} deprecated, not needed anymore

\item {} 
\emph{update\_idx} update the model gene index, not used

\item {} 
\emph{altlabels} {[}default=None{]} alternative labels for genes, default uses geneIds

\end{itemize}

\end{fulllineitems}

\index{createGroup() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.createGroup}\pysiglinewithargsret{\bfcode{createGroup}}{\emph{gid}}{}
Create an empty group with
\begin{itemize}
\item {} 
\emph{gid} the unique group id

\end{itemize}

\end{fulllineitems}

\index{createObjectiveFunction() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.createObjectiveFunction}\pysiglinewithargsret{\bfcode{createObjectiveFunction}}{\emph{rid}, \emph{coefficient=1}, \emph{osense='maximize'}, \emph{active=True}, \emph{delete\_current\_obj=True}}{}
Create a single variable objective function:
\begin{itemize}
\item {} 
\textbf{rid} The

\item {} 
\textbf{coefficient} {[}default=1{]}

\item {} 
\textbf{osense} {[}default='maximize'{]}

\item {} 
\textbf{active} {[}default=True{]}

\item {} 
\textbf{delete\_current\_obj} {[}default=True{]}

\end{itemize}

\end{fulllineitems}

\index{createReaction() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.createReaction}\pysiglinewithargsret{\bfcode{createReaction}}{\emph{rid}, \emph{name=None}, \emph{reversible=True}, \emph{create\_default\_bounds=True}, \emph{silent=False}}{}
Create a new blank reaction and add it to the model:
\begin{itemize}
\item {} 
\textbf{id} the unique reaction ID

\item {} 
\textbf{name} the reaction name

\item {} 
\textbf{reversible} {[}default=True{]} the reaction reversibility. True is reversible, False is irreversible

\item {} 
\textbf{create\_default\_bounds} create default reaction bounds, irreversible 0 \textless{}= J \textless{}= INF, reversable -INF \textless{}= J \textless{}= INF

\end{itemize}

\end{fulllineitems}

\index{createReactionBounds() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.createReactionBounds}\pysiglinewithargsret{\bfcode{createReactionBounds}}{\emph{reaction}, \emph{lb\_value}, \emph{ub\_value}}{}
Create a new lower bound for a reaction: value \textless{}= reaction
\begin{itemize}
\item {} 
\textbf{reaction} the reaction id

\item {} 
\textbf{lb\_value} the value of the lower bound

\item {} 
\textbf{ub\_value} the value of the upper bound

\end{itemize}

\end{fulllineitems}

\index{createReactionLowerBound() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.createReactionLowerBound}\pysiglinewithargsret{\bfcode{createReactionLowerBound}}{\emph{reaction}, \emph{value}}{}
Create a new lower bound for a reaction: value \textless{}= reaction
\begin{itemize}
\item {} 
\textbf{reaction} the reaction id

\item {} 
\textbf{value} the value of the bound

\end{itemize}

\end{fulllineitems}

\index{createReactionReagent() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.createReactionReagent}\pysiglinewithargsret{\bfcode{createReactionReagent}}{\emph{reaction}, \emph{metabolite}, \emph{coefficient}, \emph{silent=False}}{}
Add a reagent to an existing reaction, both reaction and metabolites must exist
\begin{itemize}
\item {} 
\emph{reaction} a reaction id

\item {} 
\emph{metabolite} a species/metabolite id

\item {} 
\emph{coefficient} the reagent coefficient

\end{itemize}

\end{fulllineitems}

\index{createReactionUpperBound() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.createReactionUpperBound}\pysiglinewithargsret{\bfcode{createReactionUpperBound}}{\emph{reaction}, \emph{value}}{}
Create a new upper bound for a reaction: reaction \textless{}= value
\begin{itemize}
\item {} 
\textbf{reaction} the reaction id

\item {} 
\textbf{value} the value of the bound

\end{itemize}

\end{fulllineitems}

\index{createSingleGeneEffectMap() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.createSingleGeneEffectMap}\pysiglinewithargsret{\bfcode{createSingleGeneEffectMap}}{}{}
This takes a model and analyses the logical gene expression patterns. This only needs to be done once,
the result is a dictionary that has boolean effect patterns as keys and the (list of) genes that give rise to
those patterns as values. This map is used by the single gene deletion method for further analysis.

Note this dictionary can also be stored and retrieved separately as long as the model structure is not changed i.e.
the gene associations themselves or order of reactions (stored as the special entry `keyJ').

Stored as self.\_\_single\_gene\_effect\_map\_\_

\end{fulllineitems}

\index{createSpecies() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.createSpecies}\pysiglinewithargsret{\bfcode{createSpecies}}{\emph{sid}, \emph{boundary=False}, \emph{name='`}, \emph{value=nan}, \emph{compartment=None}, \emph{charge=None}, \emph{chemFormula=None}}{}
Create a new species and add it to the model:
\begin{itemize}
\item {} 
\textbf{id} the unique species id

\item {} 
\textbf{boundary} {[}default=False{]} whether the species is a variable (False) or is a boundary parameter (fixed)

\item {} 
\textbf{name} {[}default='`{]} the species name

\item {} 
\textbf{value} {[}default=nan{]} the value \emph{not currently used}

\item {} 
\textbf{compartment} {[}default=None{]} the compartment the species is located in

\item {} 
\textbf{charge} {[}default=None{]} the species charge

\item {} 
\textbf{chemFormula} {[}default=None{]} the chemical formula

\end{itemize}

\end{fulllineitems}

\index{deleteAllFluxBoundsWithValue() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.deleteAllFluxBoundsWithValue}\pysiglinewithargsret{\bfcode{deleteAllFluxBoundsWithValue}}{\emph{value}}{}
Delete all flux bounds which have a specified value:
\begin{itemize}
\item {} 
\emph{value} the value of the flux bound(s) to delete

\end{itemize}

\end{fulllineitems}

\index{deleteBoundsForReactionId() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.deleteBoundsForReactionId}\pysiglinewithargsret{\bfcode{deleteBoundsForReactionId}}{\emph{rid}, \emph{lower=True}, \emph{upper=True}}{}
Delete bounds connected to reaction, rid
\begin{itemize}
\item {} 
\emph{rid} a valid reaction id

\item {} 
\emph{upper} {[}default=True{]} delete the upper bound

\item {} 
\emph{lower} {[}default=True{]} delete the lower bound

\end{itemize}

\end{fulllineitems}

\index{deleteGroup() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.deleteGroup}\pysiglinewithargsret{\bfcode{deleteGroup}}{\emph{gid}}{}
Delete a group with
\begin{itemize}
\item {} 
\emph{gid} the unique group id

\end{itemize}

\end{fulllineitems}

\index{deleteNonReactingSpecies() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.deleteNonReactingSpecies}\pysiglinewithargsret{\bfcode{deleteNonReactingSpecies}}{\emph{simulate=True}}{}
Deletes all species that are not reagents (do not to take part in a reaction).
\emph{Warning} this deletion is permanent and greedy (not selective). Returns a list of (would be) deleted species
\begin{itemize}
\item {} 
\emph{simulate} {[}default=True{]} only return a list of the speciesId's that would have been deleted if False

\end{itemize}

\end{fulllineitems}

\index{deleteObjective() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.deleteObjective}\pysiglinewithargsret{\bfcode{deleteObjective}}{\emph{objective\_id}}{}
Delete objective function:
\begin{quote}

\emph{objective\_id} the id of the objective function. If objective\_id is given  as `active' then the active objective is deleted.
\end{quote}

\end{fulllineitems}

\index{deleteReactionAndBounds() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.deleteReactionAndBounds}\pysiglinewithargsret{\bfcode{deleteReactionAndBounds}}{\emph{rid}}{}
Delete all reaction and bounds connected to reaction
\begin{itemize}
\item {} 
\emph{rid} a valid reaction id

\end{itemize}

\end{fulllineitems}

\index{deleteSpecies() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.deleteSpecies}\pysiglinewithargsret{\bfcode{deleteSpecies}}{\emph{sid}}{}
Deletes a species object with id
\begin{itemize}
\item {} 
\emph{sid} the species id

\end{itemize}

\end{fulllineitems}

\index{exportFVAdata() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.exportFVAdata}\pysiglinewithargsret{\bfcode{exportFVAdata}}{}{}
Export the fva data as an array and list of reaction id's

\end{fulllineitems}

\index{findFluxesForConnectedSpecies() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.findFluxesForConnectedSpecies}\pysiglinewithargsret{\bfcode{findFluxesForConnectedSpecies}}{\emph{metab}}{}
Returns a list of (reaction, flux value) pairs that this metabolite appears as a reagent of
\begin{itemize}
\item {} 
\emph{metab} the metabolite name

\end{itemize}

\end{fulllineitems}

\index{getActiveObjective() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.getActiveObjective}\pysiglinewithargsret{\bfcode{getActiveObjective}}{}{}
Returns the active objective object.

\end{fulllineitems}

\index{getAllFluxBounds() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.getAllFluxBounds}\pysiglinewithargsret{\bfcode{getAllFluxBounds}}{}{}
Returns a dictionary of all flux bounds {[}id:value{]}

\end{fulllineitems}

\index{getAllGeneActivities() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.getAllGeneActivities}\pysiglinewithargsret{\bfcode{getAllGeneActivities}}{}{}
Returns a dictionary of genes (if defined) and whether they are active or not

\end{fulllineitems}

\index{getAllGeneProteinAssociations() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.getAllGeneProteinAssociations}\pysiglinewithargsret{\bfcode{getAllGeneProteinAssociations}}{}{}
Returns a dictionary of genes associated with each protein

\end{fulllineitems}

\index{getAllProteinActivities() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.getAllProteinActivities}\pysiglinewithargsret{\bfcode{getAllProteinActivities}}{}{}
Returns a dictionary of reactions (if genes and GPR's are defined) and whether they are active or not

\end{fulllineitems}

\index{getAllProteinGeneAssociations() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.getAllProteinGeneAssociations}\pysiglinewithargsret{\bfcode{getAllProteinGeneAssociations}}{}{}
Returns a dictionary of the proteins associated with each gene

\end{fulllineitems}

\index{getBoundarySpeciesIds() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.getBoundarySpeciesIds}\pysiglinewithargsret{\bfcode{getBoundarySpeciesIds}}{\emph{rid=None}}{}
Return all boundary species associated with reaction
\begin{itemize}
\item {} 
rid {[}default=None{]} by default return all boundary species in a model, alternatively a string containing a reaction id or list of reaction id's

\end{itemize}

\end{fulllineitems}

\index{getCompartment() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.getCompartment}\pysiglinewithargsret{\bfcode{getCompartment}}{\emph{cid}}{}
Returns a compartment object with \emph{cid}
\begin{itemize}
\item {} 
\emph{cid} compartment ID

\end{itemize}

\end{fulllineitems}

\index{getCompartmentIds() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.getCompartmentIds}\pysiglinewithargsret{\bfcode{getCompartmentIds}}{\emph{substring=None}}{}
Returns a list of compartment Ids, applies a substring search if substring is defined
\begin{itemize}
\item {} 
\emph{substring} search for this pattern anywhere in the id

\end{itemize}

\end{fulllineitems}

\index{getDescription() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.getDescription}\pysiglinewithargsret{\bfcode{getDescription}}{}{}
Returns the model description which was stored in the SBML \textless{}notes\textgreater{} field

\end{fulllineitems}

\index{getExchangeReactionIds() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.getExchangeReactionIds}\pysiglinewithargsret{\bfcode{getExchangeReactionIds}}{}{}
Returns id's of reactions where the `is\_exchange' attribute set to True. This is by default
reactions that contain a boundary species.

\end{fulllineitems}

\index{getExchangeReactions() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.getExchangeReactions}\pysiglinewithargsret{\bfcode{getExchangeReactions}}{}{}
Returns reaction instances where the `is\_exchange' attribute set to True. This is by default
reactions that contain a boundary species.

\end{fulllineitems}

\index{getFluxBoundByID() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.getFluxBoundByID}\pysiglinewithargsret{\bfcode{getFluxBoundByID}}{\emph{fid}}{}
Returns a FluxBound with id
\begin{itemize}
\item {} 
\emph{fid} the fluxBound ID

\end{itemize}

\end{fulllineitems}

\index{getFluxBoundByReactionID() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.getFluxBoundByReactionID}\pysiglinewithargsret{\bfcode{getFluxBoundByReactionID}}{\emph{rid}, \emph{bound}}{}
Returns a FluxBound instance
\begin{itemize}
\item {} 
\emph{rid} the reaction ID

\item {} 
\emph{bound} the bound: `upper', `lower', `equality'

\end{itemize}

\end{fulllineitems}

\index{getFluxBoundIds() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.getFluxBoundIds}\pysiglinewithargsret{\bfcode{getFluxBoundIds}}{\emph{substring=None}}{}
Returns a list of fluxbound Ids, applies a substring search if substring is defined
\begin{itemize}
\item {} 
\emph{substring} search for this pattern anywhere in the id

\end{itemize}

\end{fulllineitems}

\index{getFluxBoundsByReactionID() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.getFluxBoundsByReactionID}\pysiglinewithargsret{\bfcode{getFluxBoundsByReactionID}}{\emph{rid}}{}
Returns all FluxBound instances connected to a reactionId as a tuple of valid
(lower, upper, None) or (None, None, equality) or alternatively invalid (lower, upper, equality).
\begin{quote}
\begin{itemize}
\item {} 
\emph{rid} the reaction ID

\end{itemize}

\emph{under evaluation}
\end{quote}

\end{fulllineitems}

\index{getFluxesAssociatedWithSpecies() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.getFluxesAssociatedWithSpecies}\pysiglinewithargsret{\bfcode{getFluxesAssociatedWithSpecies}}{\emph{metab}}{}
Returns a list of (reaction, flux value) pairs that this metabolite appears as a reagent in
\begin{itemize}
\item {} 
\emph{metab} the metabolite name

\end{itemize}

\end{fulllineitems}

\index{getGPRassociation() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.getGPRassociation}\pysiglinewithargsret{\bfcode{getGPRassociation}}{\emph{gpr\_id}}{}
Returns a gene protein association object that has the identifier:
\begin{itemize}
\item {} 
\emph{gpr\_id} the gene protein identifier

\end{itemize}

\end{fulllineitems}

\index{getGPRforReaction() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.getGPRforReaction}\pysiglinewithargsret{\bfcode{getGPRforReaction}}{\emph{rid}}{}
Return the GPR associated with the reaction id:
\begin{itemize}
\item {} 
\emph{rid} a reaction id

\end{itemize}

\end{fulllineitems}

\index{getGene() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.getGene}\pysiglinewithargsret{\bfcode{getGene}}{\emph{g\_id}}{}
Returns a gene object that has the identifier:
\begin{itemize}
\item {} 
\emph{gid} the gene identifier

\end{itemize}

\end{fulllineitems}

\index{getGeneIdFromLabel() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.getGeneIdFromLabel}\pysiglinewithargsret{\bfcode{getGeneIdFromLabel}}{\emph{label}}{}
Given a gene label it returns the corresponding Gene id or None
\begin{itemize}
\item {} 
\emph{label}

\end{itemize}

\end{fulllineitems}

\index{getGeneIds() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.getGeneIds}\pysiglinewithargsret{\bfcode{getGeneIds}}{\emph{substring=None}}{}
Returns a list of gene Ids, applies a substring search if substring is defined
\begin{itemize}
\item {} 
\emph{substring} search for this pattern anywhere in the id

\end{itemize}

\end{fulllineitems}

\index{getGroup() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.getGroup}\pysiglinewithargsret{\bfcode{getGroup}}{\emph{gid}}{}
Return a group with
\begin{itemize}
\item {} 
\emph{gid} the unique group id

\end{itemize}

\end{fulllineitems}

\index{getGroupIds() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.getGroupIds}\pysiglinewithargsret{\bfcode{getGroupIds}}{}{}
Delete all group ids

\end{fulllineitems}

\index{getIrreversibleReactionIds() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.getIrreversibleReactionIds}\pysiglinewithargsret{\bfcode{getIrreversibleReactionIds}}{}{}
Return a list of irreversible reaction Id's

\end{fulllineitems}

\index{getModelCreators() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.getModelCreators}\pysiglinewithargsret{\bfcode{getModelCreators}}{}{}
Return model creator information

\end{fulllineitems}

\index{getObjFuncValue() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.getObjFuncValue}\pysiglinewithargsret{\bfcode{getObjFuncValue}}{}{}
Returns the objective function value

\end{fulllineitems}

\index{getObjectiveIds() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.getObjectiveIds}\pysiglinewithargsret{\bfcode{getObjectiveIds}}{\emph{substring=None}}{}
Returns a list of objective function Ids, applies a substring search if substring is defined
\begin{itemize}
\item {} 
\emph{substring} search for this pattern anywhere in the id

\end{itemize}

\end{fulllineitems}

\index{getReaction() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.getReaction}\pysiglinewithargsret{\bfcode{getReaction}}{\emph{rid}}{}
Returns a reaction object with \emph{id}
\begin{itemize}
\item {} 
\emph{rid} reaction ID

\end{itemize}

\end{fulllineitems}

\index{getReactionActivity() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.getReactionActivity}\pysiglinewithargsret{\bfcode{getReactionActivity}}{\emph{rid}}{}
If there is a GPR and genes associated with the reaction ID then return either active=True or inactive=False
Note if there is no gene associated information then this will return active.
\begin{itemize}
\item {} 
\emph{rid} a reaction id

\end{itemize}

\end{fulllineitems}

\index{getReactionBounds() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.getReactionBounds}\pysiglinewithargsret{\bfcode{getReactionBounds}}{\emph{rid}}{}
Get the bounds of a reaction, returns a tuple of rid, lowerbound value, upperbound value and equality value (None means bound does not exist).
\begin{itemize}
\item {} 
\emph{rid} the reaction ID

\end{itemize}

\end{fulllineitems}

\index{getReactionIds() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.getReactionIds}\pysiglinewithargsret{\bfcode{getReactionIds}}{\emph{substring=None}}{}
Returns a list of reaction Ids, applies a substring search if substring is defined
\begin{itemize}
\item {} 
\emph{substring} search for this pattern anywhere in the id

\end{itemize}

\end{fulllineitems}

\index{getReactionLowerBound() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.getReactionLowerBound}\pysiglinewithargsret{\bfcode{getReactionLowerBound}}{\emph{rid}}{}
Returns the lower bound of a reaction (it it exists) or None
\begin{itemize}
\item {} 
\emph{rid} the reaction ID

\end{itemize}

\end{fulllineitems}

\index{getReactionNames() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.getReactionNames}\pysiglinewithargsret{\bfcode{getReactionNames}}{\emph{substring=None}}{}
Returns a list of reaction names, applies a substring search if substring is defined
\begin{itemize}
\item {} 
\emph{substring} search for this pattern anywhere in the name

\end{itemize}

\end{fulllineitems}

\index{getReactionUpperBound() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.getReactionUpperBound}\pysiglinewithargsret{\bfcode{getReactionUpperBound}}{\emph{rid}}{}
Returns the upper bound of a reaction (it it exists) or None
\begin{itemize}
\item {} 
\emph{rid} the reaction ID

\end{itemize}

\end{fulllineitems}

\index{getReactionValues() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.getReactionValues}\pysiglinewithargsret{\bfcode{getReactionValues}}{\emph{only\_exchange=False}}{}
Returns a dictionary of ReactionID : ReactionValue pairs:
\begin{itemize}
\item {} 
\emph{only\_exchange} {[}default=False{]} only return the reactions labelled as exchange

\end{itemize}

\end{fulllineitems}

\index{getReversibleReactionIds() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.getReversibleReactionIds}\pysiglinewithargsret{\bfcode{getReversibleReactionIds}}{}{}
Return a list of reversible reaction Id's

\end{fulllineitems}

\index{getSolutionVector() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.getSolutionVector}\pysiglinewithargsret{\bfcode{getSolutionVector}}{\emph{names=False}}{}
Return a vector of solution values
\begin{itemize}
\item {} 
\emph{names} {[}default=False{]} if True return a solution vector and list of names

\end{itemize}

\end{fulllineitems}

\index{getSpecies() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.getSpecies}\pysiglinewithargsret{\bfcode{getSpecies}}{\emph{sid}}{}
Returns a species object with \emph{sid}
\begin{itemize}
\item {} 
\emph{sid} a specied ID

\end{itemize}

\end{fulllineitems}

\index{getSpeciesIds() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.getSpeciesIds}\pysiglinewithargsret{\bfcode{getSpeciesIds}}{\emph{substring=None}}{}
Returns a list of species Ids, applies a substring search if substring is defined
\begin{itemize}
\item {} 
\emph{substring} search for this pattern anywhere in the id

\end{itemize}

\end{fulllineitems}

\index{renameObjectIds() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.renameObjectIds}\pysiglinewithargsret{\bfcode{renameObjectIds}}{\emph{prefix=None}, \emph{suffix=None}, \emph{target='all'}, \emph{ignore=None}}{}
This method is designed for target=''all'' other use may result in incomplete models.
\begin{quote}
\begin{itemize}
\item {} 
\emph{prefix} {[}None{]} if supplied add as a prefix

\item {} 
\emph{suffix} {[}None{]} if supplied add as a suffix

\item {} 
\emph{target} {[}'all'{]} specify what class of objects to rename

\end{itemize}
\begin{itemize}
\item {} 
`species'

\item {} 
`reactions'

\item {} 
`bounds'

\item {} 
`objectives'

\item {} 
`all'

\end{itemize}
\begin{itemize}
\item {} 
\emph{ignore} {[}default=None{]} a list of id's to ignore

\end{itemize}
\end{quote}

\end{fulllineitems}

\index{resetAllGenes() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.resetAllGenes}\pysiglinewithargsret{\bfcode{resetAllGenes}}{\emph{update\_reactions=False}}{}
Resets all genes to their default activity state (normally on)
\begin{itemize}
\item {} 
\emph{update\_reactions} {[}default=False{]} update the associated reactions fluxbounds from the gene deletion bounds if they exist

\end{itemize}

\end{fulllineitems}

\index{resetAllInactiveGPRBounds() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.resetAllInactiveGPRBounds}\pysiglinewithargsret{\bfcode{resetAllInactiveGPRBounds}}{}{}
Resets all reaction bounds modified by the \code{cmod.setAllInactiveGeneReactionBounds()} method to their previous values

\end{fulllineitems}

\index{setAllFluxBounds() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.setAllFluxBounds}\pysiglinewithargsret{\bfcode{setAllFluxBounds}}{\emph{bounds}}{}
DEPRECATED! use setFluxBoundsFromDict()

Sets all the fluxbounds present in bounds
\begin{itemize}
\item {} 
\emph{bounds} a dictionary of {[}fluxbound\_id : value{]} pairs (not per reaction!!!)

\end{itemize}

\end{fulllineitems}

\index{setAllInactiveGPRBounds() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.setAllInactiveGPRBounds}\pysiglinewithargsret{\bfcode{setAllInactiveGPRBounds}}{\emph{lower=0.0}, \emph{upper=0.0}}{}
Set all reactions that are inactive (as determined by gene and gpr evaluation) to bounds:
\begin{itemize}
\item {} 
\emph{lower} {[}default=0.0{]} the new lower bound

\item {} 
\emph{upper} {[}default=0.0{]} the new upper bound

\end{itemize}

\end{fulllineitems}

\index{setAllProteinActivities() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.setAllProteinActivities}\pysiglinewithargsret{\bfcode{setAllProteinActivities}}{\emph{activites}, \emph{lower=0.0}, \emph{upper=0.0}}{}
Given a dictionary of activities {[}rid : boolean{]} pairs set all the corresponding reactions:
\begin{itemize}
\item {} 
\emph{activities} a dictionary of {[}rid : boolean{]} pairs

\item {} 
\emph{lower} {[}default=0.0{]} the lower bound of the deactivated flux

\item {} 
\emph{upper} {[}default=0.0{]} the upper bound of the deactivated flux

\end{itemize}

\end{fulllineitems}

\index{setBoundValueByName() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.setBoundValueByName}\pysiglinewithargsret{\bfcode{setBoundValueByName}}{\emph{rid}, \emph{value}, \emph{bound}}{}
Deprecated use setReactionBound
\begin{description}
\item[{Set a reaction bound}] \leavevmode\begin{itemize}
\item {} 
\emph{rid} the reactions id

\item {} 
\emph{value} the new value

\item {} 
\emph{bound} this is either `lower' or `upper'

\end{itemize}

\end{description}

\end{fulllineitems}

\index{setCreatedDate() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.setCreatedDate}\pysiglinewithargsret{\bfcode{setCreatedDate}}{\emph{date=None}}{}
Set the model created date tuple(year, month, day, hour, minute, second)
\begin{itemize}
\item {} 
\emph{date} {[}default=None{]} default is now (automatic) otherwise (year, month, day, hour, minute, second) e.g. (2012, 09, 24, 13, 34, 00)

\end{itemize}

\end{fulllineitems}

\index{setDescription() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.setDescription}\pysiglinewithargsret{\bfcode{setDescription}}{\emph{html}}{}
Sets the model description which translates into the SBML \textless{}notes\textgreater{} field.
\begin{itemize}
\item {} 
\emph{html} any valid html or the empty string to clear `'

\end{itemize}

\end{fulllineitems}

\index{setFluxBoundsFromDict() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.setFluxBoundsFromDict}\pysiglinewithargsret{\bfcode{setFluxBoundsFromDict}}{\emph{bounds}}{}
Sets all the fluxbounds present in bounds
\begin{itemize}
\item {} 
\emph{bounds} a dictionary of {[}fluxbound\_id : value{]} pairs (not per reaction!!!)

\end{itemize}

\end{fulllineitems}

\index{setGeneActive() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.setGeneActive}\pysiglinewithargsret{\bfcode{setGeneActive}}{\emph{g\_id}, \emph{update\_reactions=False}}{}
Effectively restores a gene by setting it's active flag
\begin{itemize}
\item {} 
\emph{g\_id} a gene ID

\item {} 
\emph{update\_reactions} {[}default=False{]} update the associated reactions fluxbounds from the gene deletion bounds if they exist

\end{itemize}

\end{fulllineitems}

\index{setGeneInactive() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.setGeneInactive}\pysiglinewithargsret{\bfcode{setGeneInactive}}{\emph{g\_id}, \emph{update\_reactions=False}, \emph{lower=0.0}, \emph{upper=0.0}}{}
Effectively deletes a gene by setting it's inactive flag while optionally updating the GPR associated reactions
\begin{itemize}
\item {} 
\emph{g\_id} a gene ID

\item {} 
\emph{update\_reactions} {[}default=False{]} update the associated reactions fluxbounds

\item {} 
\emph{lower} {[}default=0.0{]} the deactivated reaction lower bound

\item {} 
\emph{upper} {[}default=0.0{]} the deactivated reaction upper bound

\end{itemize}

\end{fulllineitems}

\index{setModifiedDate() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.setModifiedDate}\pysiglinewithargsret{\bfcode{setModifiedDate}}{\emph{date=None}}{}
Set the model modification date: tuple(year, month, day, hour, minute, second)
\begin{itemize}
\item {} 
\emph{date} {[}default=None{]} default is now (automatic) otherwise (year, month, day, hour, minute, second) e.g. (2012, 09, 24, 13, 34, 00)

\end{itemize}

\end{fulllineitems}

\index{setObjectiveFlux() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.setObjectiveFlux}\pysiglinewithargsret{\bfcode{setObjectiveFlux}}{\emph{rid}, \emph{coefficient=1}, \emph{osense='maximize'}, \emph{delete\_objflx=True}}{}
Set single target reaction flux for the current active objective function.
\begin{itemize}
\item {} 
\emph{rid} a string containing a reaction id

\item {} 
\emph{coefficient} {[}default=1{]} an objective flux coefficient

\item {} 
\emph{osense} the optimization sense must be \textbf{maximize} or \textbf{minimize}

\item {} 
\emph{delete\_objflx} {[}default=True{]} delete all existing fluxObjectives in the active objective function

\end{itemize}

\end{fulllineitems}

\index{setPrefix() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.setPrefix}\pysiglinewithargsret{\bfcode{setPrefix}}{\emph{prefix}, \emph{target}}{}
This is alpha stuff, target can be:
\begin{itemize}
\item {} 
`species'

\item {} 
`reactions'

\item {} 
`constraints'

\item {} 
`objectives'

\item {} 
`all'

\end{itemize}

\end{fulllineitems}

\index{setReactionBound() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.setReactionBound}\pysiglinewithargsret{\bfcode{setReactionBound}}{\emph{rid}, \emph{value}, \emph{bound}}{}
Set a reaction bound
\begin{itemize}
\item {} 
\emph{rid} the reactions id

\item {} 
\emph{value} the new value

\item {} 
\emph{bound} this is either `lower' or `upper', or `equal'

\end{itemize}

\end{fulllineitems}

\index{setReactionBounds() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.setReactionBounds}\pysiglinewithargsret{\bfcode{setReactionBounds}}{\emph{rid}, \emph{lower}, \emph{upper}}{}
Set both the upper and lower bound of a reaction:
\begin{itemize}
\item {} 
\emph{rid} the good old reaction id

\item {} 
\emph{lower} the lower bound value

\item {} 
\emph{upper} the upper bound value

\end{itemize}

\end{fulllineitems}

\index{setReactionLowerBound() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.setReactionLowerBound}\pysiglinewithargsret{\bfcode{setReactionLowerBound}}{\emph{rid}, \emph{value}}{}
Set a reactions lower bound (if it exists)
\begin{itemize}
\item {} 
\emph{rid} the reactions id

\item {} 
\emph{value} the new value

\end{itemize}

\end{fulllineitems}

\index{setReactionUpperBound() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.setReactionUpperBound}\pysiglinewithargsret{\bfcode{setReactionUpperBound}}{\emph{rid}, \emph{value}}{}
Set a reactions upper bound (if it exists)
\begin{itemize}
\item {} 
\emph{rid} the reaction id

\item {} 
\emph{value} the new value

\end{itemize}

\end{fulllineitems}

\index{setSuffix() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.setSuffix}\pysiglinewithargsret{\bfcode{setSuffix}}{\emph{suffix}, \emph{target}}{}
This is alpha stuff, target can be:
\begin{itemize}
\item {} 
`species'

\item {} 
`reactions'

\item {} 
`constraints'

\item {} 
`objectives'

\item {} 
`all'

\end{itemize}

\end{fulllineitems}

\index{sortReactionsById() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.sortReactionsById}\pysiglinewithargsret{\bfcode{sortReactionsById}}{}{}
Sorts the reactions by Reaction.id uses the python string sort

\end{fulllineitems}

\index{sortSpeciesById() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.sortSpeciesById}\pysiglinewithargsret{\bfcode{sortSpeciesById}}{}{}
Sorts the reaction list by Reaction.id uses the python string sort

\end{fulllineitems}

\index{splitEqualityFluxBounds() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.splitEqualityFluxBounds}\pysiglinewithargsret{\bfcode{splitEqualityFluxBounds}}{}{}
Splits any equalit flux bounds into lower and upper bounds.

\end{fulllineitems}

\index{testGeneProteinAssociations() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.testGeneProteinAssociations}\pysiglinewithargsret{\bfcode{testGeneProteinAssociations}}{}{}
This method will test the GeneProtein associations and return a list of protein, association pairs

\end{fulllineitems}

\index{undeleteObjective() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.undeleteObjective}\pysiglinewithargsret{\bfcode{undeleteObjective}}{\emph{objective\_id}}{}
Undeltes a deleted objective function:
\begin{itemize}
\item {} 
\emph{objective\_id} the id of an objeective function

\end{itemize}

\end{fulllineitems}

\index{undeleteReactionAndBounds() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.undeleteReactionAndBounds}\pysiglinewithargsret{\bfcode{undeleteReactionAndBounds}}{\emph{rid}}{}
Undelete a reaction and bounds deleted with the \textbf{deleteReactionAndBounds} method
\begin{itemize}
\item {} 
\emph{rid} a deleted reaction id

\end{itemize}

Please note this method is still experimental ;-)

\end{fulllineitems}

\index{updateNetwork() (cbmpy.CBModel.Model method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Model.updateNetwork}\pysiglinewithargsret{\bfcode{updateNetwork}}{\emph{lower=0.0}, \emph{upper=0.0}}{}
Update the reaction network based on gene activity. If reaction is deactivated then lower and upper bounds are used
\begin{itemize}
\item {} 
\emph{lower} {[}default=0.0{]} deactivated lower bound

\item {} 
\emph{upper} {[}default=0.0{]} deactivated upper bound

\end{itemize}

\end{fulllineitems}


\end{fulllineitems}

\index{Objective (class in cbmpy.CBModel)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Objective}\pysiglinewithargsret{\strong{class }\code{cbmpy.CBModel.}\bfcode{Objective}}{\emph{pid}, \emph{operation}}{}
An objective function
\index{addFluxObjective() (cbmpy.CBModel.Objective method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Objective.addFluxObjective}\pysiglinewithargsret{\bfcode{addFluxObjective}}{\emph{fobj}}{}
Adds a FluxObjective instance to the Objective

\end{fulllineitems}

\index{createFluxObjectives() (cbmpy.CBModel.Objective method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Objective.createFluxObjectives}\pysiglinewithargsret{\bfcode{createFluxObjectives}}{\emph{fluxlist}}{}
Create and add flux objective objects to this objective function.
\begin{itemize}
\item {} 
\emph{fluxlist} a list of one or more (`coefficient', `rid') pairs

\end{itemize}

\end{fulllineitems}

\index{deleteAllFluxObjectives() (cbmpy.CBModel.Objective method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Objective.deleteAllFluxObjectives}\pysiglinewithargsret{\bfcode{deleteAllFluxObjectives}}{}{}
Delete all flux objectives

\end{fulllineitems}

\index{getFluxObjective() (cbmpy.CBModel.Objective method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Objective.getFluxObjective}\pysiglinewithargsret{\bfcode{getFluxObjective}}{\emph{foid}}{}
Return the flux objective with id.
\begin{itemize}
\item {} 
\emph{foid} the flux objective id returns either an object or a list if there are multiply defined flux objectives

\end{itemize}

\end{fulllineitems}

\index{getFluxObjectiveData() (cbmpy.CBModel.Objective method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Objective.getFluxObjectiveData}\pysiglinewithargsret{\bfcode{getFluxObjectiveData}}{}{}
Returns a list of ObjectiveFunction components as (coefficient, flux) pairs

\end{fulllineitems}

\index{getFluxObjectiveForReaction() (cbmpy.CBModel.Objective method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Objective.getFluxObjectiveForReaction}\pysiglinewithargsret{\bfcode{getFluxObjectiveForReaction}}{\emph{rid}}{}
Returns the FluxObjective associated with the suplied rid. If there is more than fluxObjective associated with a reaction (illegal)
then a list of fluxObjectives is returned.
\begin{quote}

\emph{rid} a reaction id
\end{quote}

\end{fulllineitems}

\index{getFluxObjectiveIDs() (cbmpy.CBModel.Objective method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Objective.getFluxObjectiveIDs}\pysiglinewithargsret{\bfcode{getFluxObjectiveIDs}}{}{}
Returns a list of ObjectiveFlux ids, for the reaction id's use \emph{getFluxObjectiveReactions()}
or for coefficient, fluxobjective pairs use \emph{getFluxObjectiveData()}

\end{fulllineitems}

\index{getFluxObjectiveReactions() (cbmpy.CBModel.Objective method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Objective.getFluxObjectiveReactions}\pysiglinewithargsret{\bfcode{getFluxObjectiveReactions}}{}{}
Returns a list of reactions that are used as FluxObjectives

\end{fulllineitems}

\index{getFluxObjectives() (cbmpy.CBModel.Objective method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Objective.getFluxObjectives}\pysiglinewithargsret{\bfcode{getFluxObjectives}}{}{}
Returns the list of FluxObjective objects.

\end{fulllineitems}

\index{getOperation() (cbmpy.CBModel.Objective method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Objective.getOperation}\pysiglinewithargsret{\bfcode{getOperation}}{}{}
Returns the operation or sense of the objective

\end{fulllineitems}

\index{getValue() (cbmpy.CBModel.Objective method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Objective.getValue}\pysiglinewithargsret{\bfcode{getValue}}{}{}
Returns the current value of the attribute (input/solution)

\end{fulllineitems}

\index{setOperation() (cbmpy.CBModel.Objective method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Objective.setOperation}\pysiglinewithargsret{\bfcode{setOperation}}{\emph{operation}}{}
Sets the objective operation (sense)
\begin{itemize}
\item {} 
\emph{operation} {[}default='maximize'{]} one of `maximize', `maximise', `max', `minimize', `minimise', `min'

\end{itemize}

\end{fulllineitems}

\index{setValue() (cbmpy.CBModel.Objective method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Objective.setValue}\pysiglinewithargsret{\bfcode{setValue}}{\emph{value}}{}
Sets the attribute `'value'`

\end{fulllineitems}


\end{fulllineitems}

\index{Parameter (class in cbmpy.CBModel)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Parameter}\pysiglinewithargsret{\strong{class }\code{cbmpy.CBModel.}\bfcode{Parameter}}{\emph{pid}, \emph{value}, \emph{name=None}, \emph{constant=True}}{}
Holds parameter information
\index{addAssociation() (cbmpy.CBModel.Parameter method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Parameter.addAssociation}\pysiglinewithargsret{\bfcode{addAssociation}}{\emph{assoc}}{}
Add a fluxbound ID's to associate with this object

\end{fulllineitems}

\index{deleteAssociation() (cbmpy.CBModel.Parameter method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Parameter.deleteAssociation}\pysiglinewithargsret{\bfcode{deleteAssociation}}{\emph{assoc}}{}
Delete the fluxbound id associated with this object

\end{fulllineitems}

\index{getAssociations() (cbmpy.CBModel.Parameter method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Parameter.getAssociations}\pysiglinewithargsret{\bfcode{getAssociations}}{}{}
Return the FluxBounds ID's associated with this object

\end{fulllineitems}

\index{getValue() (cbmpy.CBModel.Parameter method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Parameter.getValue}\pysiglinewithargsret{\bfcode{getValue}}{}{}
Returns the current value of the attribute (input/solution)

\end{fulllineitems}

\index{setValue() (cbmpy.CBModel.Parameter method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Parameter.setValue}\pysiglinewithargsret{\bfcode{setValue}}{\emph{value}}{}
Sets the attribute `'value'`

\end{fulllineitems}


\end{fulllineitems}

\index{Reaction (class in cbmpy.CBModel)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Reaction}\pysiglinewithargsret{\strong{class }\code{cbmpy.CBModel.}\bfcode{Reaction}}{\emph{pid}, \emph{name=None}, \emph{reversible=True}}{}
Holds reaction information
\index{addReagent() (cbmpy.CBModel.Reaction method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Reaction.addReagent}\pysiglinewithargsret{\bfcode{addReagent}}{\emph{reag}}{}
Adds an instantiated Reagent object to the reaction

\end{fulllineitems}

\index{changeId() (cbmpy.CBModel.Reaction method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Reaction.changeId}\pysiglinewithargsret{\bfcode{changeId}}{\emph{pid}}{}
Changes the Id of the reaction and updates associated FluxBounds

\end{fulllineitems}

\index{changeReagentCoefficientForSpecies() (cbmpy.CBModel.Reaction method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Reaction.changeReagentCoefficientForSpecies}\pysiglinewithargsret{\bfcode{changeReagentCoefficientForSpecies}}{\emph{s\_id}, \emph{coefficient}}{}
Change the coefficient of reagent which refers to s\_id. If there is more than one reagent that refers
to this species return a warning and a list of reagents otherwise None.
\begin{itemize}
\item {} 
\emph{s\_id} a species/metabolite id

\item {} 
\emph{coefficient} the new coefficient

\end{itemize}

\end{fulllineitems}

\index{createReagent() (cbmpy.CBModel.Reaction method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Reaction.createReagent}\pysiglinewithargsret{\bfcode{createReagent}}{\emph{metabolite}, \emph{coefficient}}{}
Create a new reagent and add it to the reaction:
\begin{quote}
\begin{itemize}
\item {} 
\textbf{metabolite} the metabolite name

\item {} 
\textbf{coefficient} the

\end{itemize}

\begin{flushright}
---negative coefficient is a substrate
-- positive coefficient is a product
\end{flushright}
\end{quote}

Will fail if a species reference already exists

\end{fulllineitems}

\index{deactivateReaction() (cbmpy.CBModel.Reaction method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Reaction.deactivateReaction}\pysiglinewithargsret{\bfcode{deactivateReaction}}{\emph{lower=0.0}, \emph{upper=0.0}}{}
Deactivates a reaction by setting its bounds to lower and upper. Restore with reactivateReaction()
\begin{itemize}
\item {} 
\emph{lower} {[}default=0.0{]} bound

\item {} 
\emph{upper} {[}default=0.0{]} bound

\end{itemize}

\end{fulllineitems}

\index{deleteReagentWithSpeciesRef() (cbmpy.CBModel.Reaction method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Reaction.deleteReagentWithSpeciesRef}\pysiglinewithargsret{\bfcode{deleteReagentWithSpeciesRef}}{\emph{species}}{}
Delete a reagent that refers to the species id:
\begin{itemize}
\item {} 
\emph{species} a species/metabolite id

\end{itemize}

\end{fulllineitems}

\index{getEquation() (cbmpy.CBModel.Reaction method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Reaction.getEquation}\pysiglinewithargsret{\bfcode{getEquation}}{\emph{reverse\_symb='='}, \emph{irreverse\_symb='\textgreater{}'}, \emph{use\_names=False}}{}
Return a pretty printed string containing the reaction equation
\begin{itemize}
\item {} 
\emph{reverse\_symb} {[}default = `='{]} the symbol to use for reversible reactions

\item {} 
\emph{irreverse\_symb} {[}default = `\textgreater{}'{]} the symbol to use for irreversible reactions

\item {} 
\emph{use\_names} {[}defualt = False{]} use species names rather than id's

\end{itemize}

\end{fulllineitems}

\index{getFVAdata() (cbmpy.CBModel.Reaction method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Reaction.getFVAdata}\pysiglinewithargsret{\bfcode{getFVAdata}}{\emph{roundnum=None}, \emph{silent=True}}{}
Returns the data generated by CBSolver.FluxVariabilityAnalysis() for this reaction as a tuple of
(Flux, FVAmin, FVAmax, span) where span is abs(FVAmax - FVAmin). FVAmin or FVAmax is None this indicates no solution
to that particular optimization (infeasible).
\begin{itemize}
\item {} 
\emph{roundnum} {[}default=None{]} the integer number of roundoff decimals the default is no rounding

\item {} 
\emph{silent} {[}default=True{]} supress output to stdout

\end{itemize}

\end{fulllineitems}

\index{getLowerBound() (cbmpy.CBModel.Reaction method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Reaction.getLowerBound}\pysiglinewithargsret{\bfcode{getLowerBound}}{}{}
Get the value of the reactions lower bound

\end{fulllineitems}

\index{getProductIds() (cbmpy.CBModel.Reaction method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Reaction.getProductIds}\pysiglinewithargsret{\bfcode{getProductIds}}{\emph{use\_names=False}}{}
Returns a list of the reaction products, species identifiers
\begin{itemize}
\item {} 
\emph{use\_names} {[}defualt = False{]} use species names rather than id's

\end{itemize}

\end{fulllineitems}

\index{getReagent() (cbmpy.CBModel.Reaction method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Reaction.getReagent}\pysiglinewithargsret{\bfcode{getReagent}}{\emph{rid}}{}
Return the one or more reagent objects which have \emph{rid}:
\begin{itemize}
\item {} 
\emph{rid} a reagent \emph{rid}

\end{itemize}

\end{fulllineitems}

\index{getReagentObjIds() (cbmpy.CBModel.Reaction method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Reaction.getReagentObjIds}\pysiglinewithargsret{\bfcode{getReagentObjIds}}{}{}
Returns a list of the reagent id's. For the name of the reagents/metabolites use \emph{\textless{}reaction\textgreater{}.getSpeciesIds()}

\end{fulllineitems}

\index{getReagentRefs() (cbmpy.CBModel.Reaction method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Reaction.getReagentRefs}\pysiglinewithargsret{\bfcode{getReagentRefs}}{}{}
Returns a list of the reagents/metabolites

\end{fulllineitems}

\index{getReagentWithSpeciesRef() (cbmpy.CBModel.Reaction method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Reaction.getReagentWithSpeciesRef}\pysiglinewithargsret{\bfcode{getReagentWithSpeciesRef}}{\emph{species}}{}
Return the reagent object which refers to the \emph{species} id:
\begin{itemize}
\item {} 
\emph{species} the species/metabolite id

\end{itemize}

\end{fulllineitems}

\index{getSpeciesIds() (cbmpy.CBModel.Reaction method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Reaction.getSpeciesIds}\pysiglinewithargsret{\bfcode{getSpeciesIds}}{}{}
Returns a list of the reagents/metabolites

\end{fulllineitems}

\index{getSpeciesObj() (cbmpy.CBModel.Reaction method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Reaction.getSpeciesObj}\pysiglinewithargsret{\bfcode{getSpeciesObj}}{}{}
Returns a list of the species objects that are reagents

\end{fulllineitems}

\index{getStoichiometry() (cbmpy.CBModel.Reaction method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Reaction.getStoichiometry}\pysiglinewithargsret{\bfcode{getStoichiometry}}{\emph{use\_names=False}, \emph{altout=False}}{}
Returns a list of (coefficient, species) pairs for this reaction
\begin{itemize}
\item {} 
\emph{use\_names} {[}default = False{]} use species names rather than id's

\item {} 
\emph{altout} {[}default = False{]} returns a dictionary

\end{itemize}

\end{fulllineitems}

\index{getSubstrateIds() (cbmpy.CBModel.Reaction method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Reaction.getSubstrateIds}\pysiglinewithargsret{\bfcode{getSubstrateIds}}{\emph{use\_names=False}}{}
Returns a list of the reaction substrates, species identifiers
\begin{itemize}
\item {} 
\emph{use\_names} {[}defualt = False{]} use species names rather than id's

\end{itemize}

\end{fulllineitems}

\index{getUpperBound() (cbmpy.CBModel.Reaction method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Reaction.getUpperBound}\pysiglinewithargsret{\bfcode{getUpperBound}}{}{}
Get the value of the reactions upper bound

\end{fulllineitems}

\index{getValue() (cbmpy.CBModel.Reaction method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Reaction.getValue}\pysiglinewithargsret{\bfcode{getValue}}{}{}
Returns the current value of the flux.

\end{fulllineitems}

\index{reactivateReaction() (cbmpy.CBModel.Reaction method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Reaction.reactivateReaction}\pysiglinewithargsret{\bfcode{reactivateReaction}}{}{}
Activates a reaction deactivated with deactivateReaction

\end{fulllineitems}

\index{setLowerBound() (cbmpy.CBModel.Reaction method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Reaction.setLowerBound}\pysiglinewithargsret{\bfcode{setLowerBound}}{\emph{value}}{}
Set the value of the reactions lower bound
\begin{itemize}
\item {} 
\emph{value} a floating point value

\end{itemize}

\end{fulllineitems}

\index{setStoichCoefficient() (cbmpy.CBModel.Reaction method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Reaction.setStoichCoefficient}\pysiglinewithargsret{\bfcode{setStoichCoefficient}}{\emph{sid}, \emph{value}}{}
Sets the stoichiometric coefficient of a reagent that refers to a metabolite. Note \emph{negative coefficients} are \emph{substrates}
while \emph{positive} ones are \emph{products}. At this point zero coefficients are not allowed
\begin{itemize}
\item {} 
\emph{sid} the species/metabolite id

\item {} 
\emph{value} a floating point value != 0

\end{itemize}

\end{fulllineitems}

\index{setUpperBound() (cbmpy.CBModel.Reaction method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Reaction.setUpperBound}\pysiglinewithargsret{\bfcode{setUpperBound}}{\emph{value}}{}
Set the value of the reactions upper bound
\begin{itemize}
\item {} 
\emph{value} a floating point value

\end{itemize}

\end{fulllineitems}

\index{setValue() (cbmpy.CBModel.Reaction method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Reaction.setValue}\pysiglinewithargsret{\bfcode{setValue}}{\emph{value}}{}
Sets the attribute \emph{value} in this case the flux.

\end{fulllineitems}

\index{undeleteReagentWithSpeciesRef() (cbmpy.CBModel.Reaction method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Reaction.undeleteReagentWithSpeciesRef}\pysiglinewithargsret{\bfcode{undeleteReagentWithSpeciesRef}}{\emph{species}}{}
Attempts to unDelete reagent deleted with deleteReagent() that refers to the species id:
\begin{itemize}
\item {} 
\emph{species} a species/metabolite id

\end{itemize}

\end{fulllineitems}


\end{fulllineitems}

\index{Reagent (class in cbmpy.CBModel)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Reagent}\pysiglinewithargsret{\strong{class }\code{cbmpy.CBModel.}\bfcode{Reagent}}{\emph{reid}, \emph{species\_ref}, \emph{coef}}{}~\begin{description}
\item[{Has a reactive species id and stoichiometric coefficient:}] \leavevmode\begin{itemize}
\item {} 
negative = substrate

\item {} 
positive = product

\item {} 
species\_ref a reference to a species obj

\end{itemize}

\end{description}
\index{getCoefficient() (cbmpy.CBModel.Reagent method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Reagent.getCoefficient}\pysiglinewithargsret{\bfcode{getCoefficient}}{}{}
Returns the reagent coefficient

\end{fulllineitems}

\index{getRole() (cbmpy.CBModel.Reagent method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Reagent.getRole}\pysiglinewithargsret{\bfcode{getRole}}{}{}
Returns the reagents role, ``substrate'', ``product'' or None

\end{fulllineitems}

\index{getSpecies() (cbmpy.CBModel.Reagent method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Reagent.getSpecies}\pysiglinewithargsret{\bfcode{getSpecies}}{}{}
Returns the metabolite/species that the reagent reference refers to

\end{fulllineitems}

\index{setCoefficient() (cbmpy.CBModel.Reagent method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Reagent.setCoefficient}\pysiglinewithargsret{\bfcode{setCoefficient}}{\emph{coef}}{}
Sets the reagent coefficient and role, negative coefficients are substrates and positive ones are products
\begin{itemize}
\item {} 
\emph{coeff} the new coefficient

\end{itemize}

\end{fulllineitems}

\index{setSpecies() (cbmpy.CBModel.Reagent method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Reagent.setSpecies}\pysiglinewithargsret{\bfcode{setSpecies}}{\emph{spe}}{}
Sets the metabolite/species that the reagent reference refers to

\end{fulllineitems}


\end{fulllineitems}

\index{Species (class in cbmpy.CBModel)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Species}\pysiglinewithargsret{\strong{class }\code{cbmpy.CBModel.}\bfcode{Species}}{\emph{pid}, \emph{boundary=False}, \emph{name=None}, \emph{value=nan}, \emph{compartment=None}, \emph{charge=None}, \emph{chemFormula=None}}{}
Holds species/metabolite information
\index{getCharge() (cbmpy.CBModel.Species method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Species.getCharge}\pysiglinewithargsret{\bfcode{getCharge}}{}{}
Returns the species charge

\end{fulllineitems}

\index{getChemFormula() (cbmpy.CBModel.Species method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Species.getChemFormula}\pysiglinewithargsret{\bfcode{getChemFormula}}{}{}
Returns the species chemical formula

\end{fulllineitems}

\index{getReagentOf() (cbmpy.CBModel.Species method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Species.getReagentOf}\pysiglinewithargsret{\bfcode{getReagentOf}}{}{}
Returns a list of reaction id's that this metabolite occurs in

\end{fulllineitems}

\index{getValue() (cbmpy.CBModel.Species method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Species.getValue}\pysiglinewithargsret{\bfcode{getValue}}{}{}
Returns the current value of the attribute (input/solution)

\end{fulllineitems}

\index{isReagentOf() (cbmpy.CBModel.Species method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Species.isReagentOf}\pysiglinewithargsret{\bfcode{isReagentOf}}{}{}
Returns a dynamically generated list of reactions that this species occurs as a reagent

\end{fulllineitems}

\index{setBoundary() (cbmpy.CBModel.Species method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Species.setBoundary}\pysiglinewithargsret{\bfcode{setBoundary}}{}{}
Sets the species so it is a boundary metabolite or fixed which does not occur in the stoichiometric matrix N

\end{fulllineitems}

\index{setCharge() (cbmpy.CBModel.Species method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Species.setCharge}\pysiglinewithargsret{\bfcode{setCharge}}{\emph{charge}}{}
Sets the species charge:
\begin{itemize}
\item {} 
\emph{charge} a signed double but generally a signed int is used

\end{itemize}

\end{fulllineitems}

\index{setChemFormula() (cbmpy.CBModel.Species method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Species.setChemFormula}\pysiglinewithargsret{\bfcode{setChemFormula}}{\emph{cf}}{}
Sets the species chemical formula
\begin{itemize}
\item {} 
\emph{cf} a chemical formula e.g. CH3NO2

\end{itemize}

\end{fulllineitems}

\index{setReagentOf() (cbmpy.CBModel.Species method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Species.setReagentOf}\pysiglinewithargsret{\bfcode{setReagentOf}}{\emph{rid}}{}
Adds the supplied reaction id to the reagent\_of list (if it isn't one already)
\begin{itemize}
\item {} 
\emph{rid} a valid reaction id

\end{itemize}

\end{fulllineitems}

\index{setValue() (cbmpy.CBModel.Species method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Species.setValue}\pysiglinewithargsret{\bfcode{setValue}}{\emph{value}}{}
Sets the attribute `'value'`

\end{fulllineitems}

\index{unsetBoundary() (cbmpy.CBModel.Species method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBModel.Species.unsetBoundary}\pysiglinewithargsret{\bfcode{unsetBoundary}}{}{}
Unsets the species boundary attribute so that the metabolite is free and therefore occurs in the stoichiometric matrix N

\end{fulllineitems}


\end{fulllineitems}

\phantomsection\label{modules_doc:module-cbmpy.CBModelTools}\index{cbmpy.CBModelTools (module)}

\section{CBMPy: CBModelTools module}
\label{modules_doc:cbmpy-cbmodeltools-module}
PySCeS Constraint Based Modelling (\href{http://cbmpy.sourceforge.net}{http://cbmpy.sourceforge.net})
Copyright (C) 2009-2015 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see \textless{}\href{http://www.gnu.org/licenses/}{http://www.gnu.org/licenses/}\textgreater{}

Author: Brett G. Olivier
Contact email: \href{mailto:bgoli@users.sourceforge.net}{bgoli@users.sourceforge.net}
Last edit: \$Author: bgoli \$ (\$Id: CBModelTools.py 305 2015-04-23 15:18:31Z bgoli \$)
\phantomsection\label{modules_doc:module-cbmpy.CBMultiCore}\index{cbmpy.CBMultiCore (module)}

\section{CBMPy: CBMultiCore module}
\label{modules_doc:cbmpy-cbmulticore-module}
PySCeS Constraint Based Modelling (\href{http://cbmpy.sourceforge.net}{http://cbmpy.sourceforge.net})
Copyright (C) 2009-2015 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see \textless{}\href{http://www.gnu.org/licenses/}{http://www.gnu.org/licenses/}\textgreater{}

Author: Brett G. Olivier
Contact email: \href{mailto:bgoli@users.sourceforge.net}{bgoli@users.sourceforge.net}
Last edit: \$Author: bgoli \$ (\$Id: CBMultiCore.py 305 2015-04-23 15:18:31Z bgoli \$)
\index{grouper() (in module cbmpy.CBMultiCore)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBMultiCore.grouper}\pysiglinewithargsret{\code{cbmpy.CBMultiCore.}\bfcode{grouper}}{\emph{3}, \emph{`abcdefg'}, \emph{`x') --\textgreater{} (`a'}, \emph{`b'}, \emph{`c')}, \emph{(`d'}, \emph{`e'}, \emph{`f')}, \emph{(`g'}, \emph{`x'}, \emph{`x'}}{}
\end{fulllineitems}

\index{runMultiCoreFVA() (in module cbmpy.CBMultiCore)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBMultiCore.runMultiCoreFVA}\pysiglinewithargsret{\code{cbmpy.CBMultiCore.}\bfcode{runMultiCoreFVA}}{\emph{fba}, \emph{selected\_reactions=None}, \emph{pre\_opt=True}, \emph{tol=None}, \emph{objF2constr=True}, \emph{rhs\_sense='lower'}, \emph{optPercentage=100.0}, \emph{work\_dir=None}, \emph{quiet=True}, \emph{debug=False}, \emph{oldlpgen=False}, \emph{markupmodel=True}, \emph{procs=2}}{}
Run a multicore FVA where:
\begin{itemize}
\item {} 
\emph{fba} is an fba model instance

\item {} 
\emph{procs} {[}default=2{]} number of processing threads (optimum seems to be about the number of physical cores)

\end{itemize}

\end{fulllineitems}

\index{runMultiCoreMultiEnvFVA() (in module cbmpy.CBMultiCore)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBMultiCore.runMultiCoreMultiEnvFVA}\pysiglinewithargsret{\code{cbmpy.CBMultiCore.}\bfcode{runMultiCoreMultiEnvFVA}}{\emph{lp}, \emph{selected\_reactions=None}, \emph{tol=None}, \emph{rhs\_sense='lower'}, \emph{optPercentage=100.0}, \emph{work\_dir=None}, \emph{debug=False}, \emph{procs=2}}{}
Run a multicore FVA where:
\begin{itemize}
\item {} 
\emph{lp} is a multienvironment lp model instance

\item {} 
\emph{procs} {[}default=2{]} number of processing threads (optimum seems to be about the number of physical cores)

\end{itemize}

\end{fulllineitems}

\phantomsection\label{modules_doc:module-cbmpy.CBMultiEnv}\index{cbmpy.CBMultiEnv (module)}

\section{CBMPy: CBMultiEnv module}
\label{modules_doc:cbmpy-cbmultienv-module}
PySCeS Constraint Based Modelling (\href{http://cbmpy.sourceforge.net}{http://cbmpy.sourceforge.net})
Copyright (C) 2009-2015 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see \textless{}\href{http://www.gnu.org/licenses/}{http://www.gnu.org/licenses/}\textgreater{}

Author: Brett G. Olivier
Contact email: \href{mailto:bgoli@users.sourceforge.net}{bgoli@users.sourceforge.net}
Last edit: \$Author: bgoli \$ (\$Id: CBMultiEnv.py 305 2015-04-23 15:18:31Z bgoli \$)
\phantomsection\label{modules_doc:module-cbmpy.CBNetDB}\index{cbmpy.CBNetDB (module)}

\section{CBMPy: CBNetDB module}
\label{modules_doc:cbmpy-cbnetdb-module}
PySCeS Constraint Based Modelling (\href{http://cbmpy.sourceforge.net}{http://cbmpy.sourceforge.net})
Copyright (C) 2009-2015 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see \textless{}\href{http://www.gnu.org/licenses/}{http://www.gnu.org/licenses/}\textgreater{}

Author: Brett G. Olivier
Contact email: \href{mailto:bgoli@users.sourceforge.net}{bgoli@users.sourceforge.net}
Last edit: \$Author: bgoli \$ (\$Id: CBNetDB.py 375 2015-09-04 15:44:42Z bgoli \$)
\index{DBTools (class in cbmpy.CBNetDB)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBNetDB.DBTools}\pysigline{\strong{class }\code{cbmpy.CBNetDB.}\bfcode{DBTools}}
Some user friendly tools to work with SQLite2 DB's
\index{checkEntry() (cbmpy.CBNetDB.DBTools method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBNetDB.DBTools.checkEntry}\pysiglinewithargsret{\bfcode{checkEntry}}{\emph{table}, \emph{id}}{}
Check if an entry exists in a table
\begin{itemize}
\item {} 
\emph{table} the table name

\item {} 
\emph{id} the table row to search for

\end{itemize}

\end{fulllineitems}

\index{connectSQLiteDB() (cbmpy.CBNetDB.DBTools method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBNetDB.DBTools.connectSQLiteDB}\pysiglinewithargsret{\bfcode{connectSQLiteDB}}{\emph{db\_name}, \emph{work\_dir=None}}{}
Connect to a sqlite database.
\begin{itemize}
\item {} 
\emph{db\_name} the name of the sqlite database

\item {} 
\emph{work\_dir} the optional database path

\end{itemize}

\end{fulllineitems}

\index{createDBTable() (cbmpy.CBNetDB.DBTools method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBNetDB.DBTools.createDBTable}\pysiglinewithargsret{\bfcode{createDBTable}}{\emph{table}, \emph{sqlcols}}{}
Create a database table if it does not exist:
\begin{itemize}
\item {} 
\emph{table} the table name

\item {} 
\emph{sqlcols} a list containing the SQL definitions of the table columns: \textless{}id\textgreater{} \textless{}type\textgreater{} for examepl \emph{{[}'gene TEXT PRIMARY KEY', `aa\_seq TEXT', `nuc\_seq TEXT', `aa\_len INT', `nuc\_len INT'{]}}

\end{itemize}

Effectively writes CREATE TABLE ``table'' (\textless{}id\textgreater{} \textless{}type\textgreater{}, gene TEXT PRIMARY KEY, aa\_seq TEXT, nuc\_seq TEXT, aa\_len INT, nuc\_len INT) \% table

\end{fulllineitems}

\index{dumpTableToTxt() (cbmpy.CBNetDB.DBTools method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBNetDB.DBTools.dumpTableToTxt}\pysiglinewithargsret{\bfcode{dumpTableToTxt}}{\emph{table}, \emph{filename}}{}
Save a table as tab separated txt file
\begin{itemize}
\item {} 
\emph{table} the table to export

\item {} 
\emph{filename} the filename of the table dump

\end{itemize}

\end{fulllineitems}

\index{executeSQL() (cbmpy.CBNetDB.DBTools method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBNetDB.DBTools.executeSQL}\pysiglinewithargsret{\bfcode{executeSQL}}{\emph{sql}}{}
Execute a SQL command:
\begin{itemize}
\item {} 
\emph{sql} a string containing a SQL command

\end{itemize}

\end{fulllineitems}

\index{fetchAll() (cbmpy.CBNetDB.DBTools method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBNetDB.DBTools.fetchAll}\pysiglinewithargsret{\bfcode{fetchAll}}{\emph{sql}}{}
E.g. SELECT aa\_len FROM gene\_data WHERE gene=''G'''

\end{fulllineitems}

\index{getTable() (cbmpy.CBNetDB.DBTools method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBNetDB.DBTools.getTable}\pysiglinewithargsret{\bfcode{getTable}}{\emph{table}, \emph{colOut=False}}{}
Returns an entire database table
\begin{itemize}
\item {} 
\emph{table} the table name

\item {} 
\emph{colOut} optionally return a tuple of (data,ColNames)

\end{itemize}

\end{fulllineitems}

\index{insertData() (cbmpy.CBNetDB.DBTools method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBNetDB.DBTools.insertData}\pysiglinewithargsret{\bfcode{insertData}}{\emph{table}, \emph{data}, \emph{commit=True}}{}~\begin{description}
\item[{Insert data into a table: ``INSERT INTO \%s (gene, aa\_seq, nuc\_seq, aa\_len, nuc\_len) VALUES (?, ?, ?, ?, ?)'' \% tablename,}] \leavevmode\begin{quote}

(str(ecg), str(prot2), str(gene2), int(len(prot2)), int(len(gene2))) )
\end{quote}
\begin{itemize}
\item {} 
\emph{table} the DB table name

\item {} 
\emph{data} a list of (column\_id, value) pairs

\item {} 
\emph{commit} whether to commit the data insertions

\end{itemize}

\end{description}

\end{fulllineitems}

\index{updateData() (cbmpy.CBNetDB.DBTools method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBNetDB.DBTools.updateData}\pysiglinewithargsret{\bfcode{updateData}}{\emph{table}, \emph{id}}{}
Update already defined data (primary key)
\begin{itemize}
\item {} 
\emph{table} the table name

\item {} 
\emph{id} the table row to search for

\end{itemize}

\end{fulllineitems}


\end{fulllineitems}

\index{KeGGSequenceTools (class in cbmpy.CBNetDB)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBNetDB.KeGGSequenceTools}\pysiglinewithargsret{\strong{class }\code{cbmpy.CBNetDB.}\bfcode{KeGGSequenceTools}}{\emph{url}, \emph{db\_name}, \emph{work\_dir}}{}
Using the KeGG connector this class provides tools to construct an organims specific sequence database

\end{fulllineitems}

\index{KeGGTools (class in cbmpy.CBNetDB)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBNetDB.KeGGTools}\pysiglinewithargsret{\strong{class }\code{cbmpy.CBNetDB.}\bfcode{KeGGTools}}{\emph{url}}{}
Class that holds useful methods for querying KeGG via a SUDS provided soap client
\index{fetchSeqfromKeGG() (cbmpy.CBNetDB.KeGGTools method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBNetDB.KeGGTools.fetchSeqfromKeGG}\pysiglinewithargsret{\bfcode{fetchSeqfromKeGG}}{\emph{k\_gene}}{}
Given a gene name try and retrieve the gene and amino acid sequence

\end{fulllineitems}


\end{fulllineitems}

\index{MIRIAMTools (class in cbmpy.CBNetDB)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBNetDB.MIRIAMTools}\pysigline{\strong{class }\code{cbmpy.CBNetDB.}\bfcode{MIRIAMTools}}
Tools dealing with MIRIAM annotations

\end{fulllineitems}

\index{RESTClient (class in cbmpy.CBNetDB)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBNetDB.RESTClient}\pysigline{\strong{class }\code{cbmpy.CBNetDB.}\bfcode{RESTClient}}
Class that provides the basis for application specific connectors to REST web services
\index{Close() (cbmpy.CBNetDB.RESTClient method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBNetDB.RESTClient.Close}\pysiglinewithargsret{\bfcode{Close}}{}{}
Close the currently active connection

\end{fulllineitems}

\index{Connect() (cbmpy.CBNetDB.RESTClient method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBNetDB.RESTClient.Connect}\pysiglinewithargsret{\bfcode{Connect}}{\emph{root}}{}
Establish HTTP connection to
\begin{itemize}
\item {} 
\emph{root} the site root ``www.google.com''

\end{itemize}

\end{fulllineitems}

\index{Get() (cbmpy.CBNetDB.RESTClient method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBNetDB.RESTClient.Get}\pysiglinewithargsret{\bfcode{Get}}{\emph{query}}{}
Perform an http GET using:
\begin{itemize}
\item {} 
\emph{query} e.g.

\item {} 
\emph{reply\_mode} {[}default='`{]} this is the reply mode

\end{itemize}

For example ``/semanticSBML/annotate/search.xml?q=ATP''

\end{fulllineitems}

\index{GetLog() (cbmpy.CBNetDB.RESTClient method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBNetDB.RESTClient.GetLog}\pysiglinewithargsret{\bfcode{GetLog}}{}{}
Return the logged history

\end{fulllineitems}

\index{Log() (cbmpy.CBNetDB.RESTClient method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBNetDB.RESTClient.Log}\pysiglinewithargsret{\bfcode{Log}}{\emph{txt}}{}
Add txt to logfile history
\begin{itemize}
\item {} 
\emph{txt} a string

\end{itemize}

\end{fulllineitems}

\index{URLDecode() (cbmpy.CBNetDB.RESTClient method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBNetDB.RESTClient.URLDecode}\pysiglinewithargsret{\bfcode{URLDecode}}{\emph{txt}}{}
Decodes a URL encoded string

\end{fulllineitems}

\index{URLEncode() (cbmpy.CBNetDB.RESTClient method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBNetDB.RESTClient.URLEncode}\pysiglinewithargsret{\bfcode{URLEncode}}{\emph{txt}}{}
URL encodes a string.

\end{fulllineitems}


\end{fulllineitems}

\index{SemanticSBML (class in cbmpy.CBNetDB)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBNetDB.SemanticSBML}\pysigline{\strong{class }\code{cbmpy.CBNetDB.}\bfcode{SemanticSBML}}
REST client for connecting to SemanticSBML services
\index{parseXMLtoText() (cbmpy.CBNetDB.SemanticSBML method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBNetDB.SemanticSBML.parseXMLtoText}\pysiglinewithargsret{\bfcode{parseXMLtoText}}{\emph{xml}}{}
Parse the xml output by quickLookup() into a list of URL
\begin{itemize}
\item {} 
\emph{xml} XML returns from SemanticSBML

\end{itemize}

\end{fulllineitems}

\index{quickLookup() (cbmpy.CBNetDB.SemanticSBML method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBNetDB.SemanticSBML.quickLookup}\pysiglinewithargsret{\bfcode{quickLookup}}{\emph{txt}}{}
Do a quick lookpup for txt using SemanticSBML (connectic if required) and return results. Returns
a list of identifiers.org id's in descending priority (as return)
\begin{itemize}
\item {} 
\emph{txt} the string to lookup

\end{itemize}

\end{fulllineitems}

\index{viewDataInWebrowser() (cbmpy.CBNetDB.SemanticSBML method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBNetDB.SemanticSBML.viewDataInWebrowser}\pysiglinewithargsret{\bfcode{viewDataInWebrowser}}{\emph{maxres=10}}{}
Attempt to view \#maxres results returned by SemanticSBML in the default browser
\begin{itemize}
\item {} 
\emph{maxres} default maximum number of results to display.

\end{itemize}

\end{fulllineitems}


\end{fulllineitems}

\phantomsection\label{modules_doc:module-cbmpy.CBPlot}\index{cbmpy.CBPlot (module)}

\section{CBMPy: CBPlot module}
\label{modules_doc:cbmpy-cbplot-module}
PySCeS Constraint Based Modelling (\href{http://cbmpy.sourceforge.net}{http://cbmpy.sourceforge.net})
Copyright (C) 2009-2015 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see \textless{}\href{http://www.gnu.org/licenses/}{http://www.gnu.org/licenses/}\textgreater{}

Author: Brett G. Olivier
Contact email: \href{mailto:bgoli@users.sourceforge.net}{bgoli@users.sourceforge.net}
Last edit: \$Author: bgoli \$ (\$Id: CBPlot.py 305 2015-04-23 15:18:31Z bgoli \$)
\index{plotFluxVariability() (in module cbmpy.CBPlot)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBPlot.plotFluxVariability}\pysiglinewithargsret{\code{cbmpy.CBPlot.}\bfcode{plotFluxVariability}}{\emph{fva\_data}, \emph{fva\_names}, \emph{fname}, \emph{work\_dir=None}, \emph{title=None}, \emph{ySlice=None}, \emph{minHeight=None}, \emph{maxHeight=None}, \emph{roundec=None}, \emph{autoclose=True}, \emph{fluxval=True}, \emph{type='png'}}{}
Plots and saves as an image the flux variability results as generated by CBSolver.FluxVariabilityAnalysis.
\begin{itemize}
\item {} 
\emph{fva\_data} FluxVariabilityAnalysis() FVA OUTPUT\_ARRAY

\item {} 
\emph{fva\_names} FluxVariabilityAnalysis() FVA OUTPUT\_NAMES

\item {} 
\emph{fname} filename\_base for the CSV output

\item {} 
\emph{work\_dir} {[}default=None{]} if set the output directory for the csv files

\item {} 
\emph{title} {[}default=None{]} the user defined title for the graph

\item {} 
\emph{ySlice} {[}default=None{]} this sets an absolute (fixed) limit on the Y-axis (+- ySlice)

\item {} 
\emph{minHeight} {[}default=None{]} the minimum length that defined a span

\item {} 
\emph{maxHeight} {[}default=None{]} the maximum length a span can obtain, bar will be limited to maxHeight and coloured yellow

\item {} 
\emph{roundec} {[}default=None{]} an integer indicating at which decimal to round off output. Default is no rounding.

\item {} 
\emph{autoclose} {[}default=True{]} autoclose plot after save

\item {} 
\emph{fluxval} {[}default=True{]} plot the flux value

\item {} 
\emph{type} {[}default='png'{]} the output format, depends on matplotlib backend e.g. `png', `pdf', `eps'

\end{itemize}

\end{fulllineitems}

\phantomsection\label{modules_doc:module-cbmpy.CBQt4}\index{cbmpy.CBQt4 (module)}

\section{CBMPy: CBQt4 module}
\label{modules_doc:cbmpy-cbqt4-module}
Constraint Based Modelling in Python (\href{http://pysces.sourceforge.net/getNewReaction}{http://pysces.sourceforge.net/getNewReaction})
Copyright (C) 2009-2015 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see \textless{}\href{http://www.gnu.org/licenses/}{http://www.gnu.org/licenses/}\textgreater{}

Author: Brett G. Olivier
Contact email: \href{mailto:bgoli@users.sourceforge.net}{bgoli@users.sourceforge.net}
Last edit: \$Author: bgoli \$ (\$Id: CBQt4.py 305 2015-04-23 15:18:31Z bgoli \$)
\index{createReaction() (in module cbmpy.CBQt4)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBQt4.createReaction}\pysiglinewithargsret{\code{cbmpy.CBQt4.}\bfcode{createReaction}}{\emph{mod}}{}
Create a reaction using the graphical Reaction Creator
\begin{itemize}
\item {} 
\emph{mod} a CBMPy model object

\end{itemize}

\end{fulllineitems}

\phantomsection\label{modules_doc:module-cbmpy.CBRead}\index{cbmpy.CBRead (module)}

\section{CBMPy: CBRead module}
\label{modules_doc:cbmpy-cbread-module}
PySCeS Constraint Based Modelling (\href{http://cbmpy.sourceforge.net}{http://cbmpy.sourceforge.net})
Copyright (C) 2009-2015 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see \textless{}\href{http://www.gnu.org/licenses/}{http://www.gnu.org/licenses/}\textgreater{}

Author: Brett G. Olivier
Contact email: \href{mailto:bgoli@users.sourceforge.net}{bgoli@users.sourceforge.net}
Last edit: \$Author: bgoli \$ (\$Id: CBRead.py 416 2016-02-23 16:12:23Z bgoli \$)
\index{readCOBRASBML() (in module cbmpy.CBRead)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBRead.readCOBRASBML}\pysiglinewithargsret{\code{cbmpy.CBRead.}\bfcode{readCOBRASBML}}{\emph{fname}, \emph{work\_dir=None}, \emph{return\_sbml\_model=False}, \emph{delete\_intermediate=False}, \emph{fake\_boundary\_species\_search=False}, \emph{output\_dir=None}}{}
Read in a COBRA format SBML Level 2 file with FBA annotation where and return either a CBM model object
or a (cbm\_mod, sbml\_mod) pair if return\_sbml\_model=True
\begin{itemize}
\item {} 
\emph{fname} is the filename

\item {} 
\emph{work\_dir} is the working directory

\item {} 
\emph{return\_sbml\_model} {[}default=False{]} return a a (cbm\_mod, sbml\_mod) pair

\item {} 
\emph{delete\_intermediate} {[}default=False{]} delete the intermediate SBML Level 3 FBC file

\item {} 
\emph{fake\_boundary\_species\_search} {[}default=False{]} after looking for the boundary\_condition of a species search for overloaded id's \textless{}id\textgreater{}\_b

\item {} 
\emph{output\_dir} {[}default=None{]} the directory to output the intermediate SBML L3 files (if generated) default to input directory

\end{itemize}

\end{fulllineitems}

\index{readExcel97Model() (in module cbmpy.CBRead)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBRead.readExcel97Model}\pysiglinewithargsret{\code{cbmpy.CBRead.}\bfcode{readExcel97Model}}{\emph{xlname}, \emph{write\_sbml=True}, \emph{sbml\_level=3}, \emph{return\_dictionaries=False}}{}
Reads a model encoded as an Excel97 workbook and returns it as a CBMPy model object and SBML file. Note the workbook must be formatted
exactly like those produced by cbm.writeModelToExcel97(). Note that reactions have to be defined in \textbf{both} the \emph{reaction}
and \emph{network\_react} sheets to be included in the model.
\begin{itemize}
\item {} 
\emph{xlpath} the filename of the Excel workbook

\item {} 
\emph{return\_model} {[}default=True{]} construct and return the CBMPy model

\item {} 
\emph{write\_sbml} {[}default=True{]} write the SBML file to fname

\item {} 
\emph{return\_dictionaries} {[}default=False{]} return the dictionaries constructed when reading the Excel file (in place of the model)

\item {} 
\emph{sbml\_level} {[}default=3{]} write the SBML file as either SBML L2 FBA or SBML L3 FBC file.

\end{itemize}

\end{fulllineitems}

\index{readSBML2FBA() (in module cbmpy.CBRead)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBRead.readSBML2FBA}\pysiglinewithargsret{\code{cbmpy.CBRead.}\bfcode{readSBML2FBA}}{\emph{fname}, \emph{work\_dir=None}, \emph{return\_sbml\_model=False}, \emph{fake\_boundary\_species\_search=False}}{}
Read in an SBML Level 2 file with FBA annotation where:
\begin{itemize}
\item {} 
\emph{fname} is the filename

\item {} 
\emph{work\_dir} is the working directory if None then only fname is used

\item {} 
\emph{return\_sbml\_model} {[}default=False{]} return a a (cbm\_mod, sbml\_mod) pair

\item {} 
\emph{fake\_boundary\_species\_search} {[}default=False{]} after looking for the boundary\_condition of a species search for overloaded id's \textless{}id\textgreater{}\_b

\end{itemize}

\end{fulllineitems}

\index{readSBML3FBC() (in module cbmpy.CBRead)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBRead.readSBML3FBC}\pysiglinewithargsret{\code{cbmpy.CBRead.}\bfcode{readSBML3FBC}}{\emph{fname}, \emph{work\_dir=None}, \emph{return\_sbml\_model=False}, \emph{xoptions=\{`validate': False\}}}{}
Read in an SBML Level 3 file with FBC annotation where and return either a CBM model object
or a (cbm\_mod, sbml\_mod) pair if return\_sbml\_model=True
\begin{itemize}
\item {} 
\emph{fname} is the filename

\item {} 
\emph{work\_dir} is the working directory

\item {} 
\emph{return\_sbml\_model} {[}default=False{]} return a a (cbm\_mod, sbml\_mod) pair

\item {} 
\emph{xoptions} special load options enable with option = True
- \emph{nogenes} do not load/process genes
- \emph{noannot} do not load/process any annotations
- \emph{validate} validate model and display errors and warnings before loading

\end{itemize}

\end{fulllineitems}

\index{readSK\_FVA() (in module cbmpy.CBRead)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBRead.readSK_FVA}\pysiglinewithargsret{\code{cbmpy.CBRead.}\bfcode{readSK\_FVA}}{\emph{filename}}{}
Read Stevens FVA results (opt.fva) file and return a list of dictionaries

\end{fulllineitems}

\index{readSK\_vertex() (in module cbmpy.CBRead)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBRead.readSK_vertex}\pysiglinewithargsret{\code{cbmpy.CBRead.}\bfcode{readSK\_vertex}}{\emph{fname}, \emph{bigfile=True}, \emph{fast\_rational=False}, \emph{nformat='\%.14f'}, \emph{compression=None}, \emph{hdf5file=None}}{}
Reads in Stevens vertex analysis file:
\begin{itemize}
\item {} 
\emph{fname} the input filename (.all file that results from Stevens pipeline)

\item {} 
\emph{bigfile} {[}default=True{]} this option is now always true and is left in for backwards compatability

\item {} 
\emph{fast\_rational} {[}default=False{]} by default off and uses SymPy for rational--\textgreater{}float conversion, when on uses float decomposition with a slight (2th decimal) decrease in accuracy

\item {} 
\emph{nformat} {[}default='\%.14f'{]} the number format used in output files

\item {} 
\emph{compression} {[}default=None{]} compression to be used in hdf5 files can be one of {[}None, `lzf', `gz?', `szip'{]}

\item {} 
\emph{hdf5file} {[}default=None{]} if None then generic filename `\_vtx\_.tmp.hdf5' is uses otherwise \textless{}hdf5file\textgreater{}.hdf5

\end{itemize}

and returns an hdf5 \emph{filename} of the results with a single group named \textbf{data} which countains datasets
\begin{itemize}
\item {} 
vertices

\item {} 
rays

\item {} 
lin

\end{itemize}

where all vectors are in terms of the column space of N.

\end{fulllineitems}

\index{readSK\_vertexOld() (in module cbmpy.CBRead)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBRead.readSK_vertexOld}\pysiglinewithargsret{\code{cbmpy.CBRead.}\bfcode{readSK\_vertexOld}}{\emph{fname}, \emph{bigfile=False}, \emph{fast\_rational=False}, \emph{nformat='\%.14f'}, \emph{compresslevel=3}}{}
Reads in Stevens vertex analysis file and returns, even more optimized for large datasets than the original.
\begin{itemize}
\item {} 
a list of vertex vectors

\item {} 
a list of ray vectors

\item {} 
the basis of the lineality space as a list of vectors

\end{itemize}

all vectors in terms of the column space of N

\end{fulllineitems}

\phantomsection\label{modules_doc:module-cbmpy.CBReadtxt}\index{cbmpy.CBReadtxt (module)}

\section{CBMPy: CBReadtxt module}
\label{modules_doc:cbmpy-cbreadtxt-module}
PySCeS Constraint Based Modelling (\href{http://cbmpy.sourceforge.net}{http://cbmpy.sourceforge.net})
Copyright (C) 2009-2015 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see \textless{}\href{http://www.gnu.org/licenses/}{http://www.gnu.org/licenses/}\textgreater{}

Author: Brett G. Olivier
Contact email: \href{mailto:bgoli@users.sourceforge.net}{bgoli@users.sourceforge.net}
Last edit: \$Author: bgoli \$ (\$Id: CBReadtxt.py 386 2015-09-28 14:05:35Z bgoli \$)
\index{readCSV() (in module cbmpy.CBReadtxt)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBReadtxt.readCSV}\pysiglinewithargsret{\code{cbmpy.CBReadtxt.}\bfcode{readCSV}}{\emph{model\_file}, \emph{bounds\_file=None}, \emph{biomass\_flux=None}, \emph{model\_id='FBAModel'}, \emph{reaction\_prefix='R\_'}, \emph{has\_header=False}}{}
This function loads a CSV file and translates it into a Python object:

\begin{Verbatim}[commandchars=\\\{\}]
\PYGZhy{} *model\PYGZus{}file* the name of the CSV file that contains the model
\PYGZhy{} *bounds\PYGZus{}file* the name of the CSV file that contains the flux bounds
\PYGZhy{} *biomass\PYGZus{}flux* the name of the reaction that is the objective function
\PYGZhy{} *reaction\PYGZus{}prefix* [default=\PYGZsq{}R \PYGZus{}\PYGZsq{}] the prefix to add to input reaction ID\PYGZsq{}s
\PYGZhy{} *has\PYGZus{}header* [default=False] if there is a header row in the csv file
\end{Verbatim}

\end{fulllineitems}

\phantomsection\label{modules_doc:module-cbmpy.CBSolver}\index{cbmpy.CBSolver (module)}

\section{CBMPy: CBSolver module}
\label{modules_doc:cbmpy-cbsolver-module}
PySCeS Constraint Based Modelling (\href{http://cbmpy.sourceforge.net}{http://cbmpy.sourceforge.net})
Copyright (C) 2009-2015 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see \textless{}\href{http://www.gnu.org/licenses/}{http://www.gnu.org/licenses/}\textgreater{}

Author: Brett G. Olivier
Contact email: \href{mailto:bgoli@users.sourceforge.net}{bgoli@users.sourceforge.net}
Last edit: \$Author: bgoli \$ (\$Id: CBSolver.py 305 2015-04-23 15:18:31Z bgoli \$)
\phantomsection\label{modules_doc:module-cbmpy.CBTools}\index{cbmpy.CBTools (module)}

\section{CBMPy: CBTools module}
\label{modules_doc:cbmpy-cbtools-module}
PySCeS Constraint Based Modelling (\href{http://cbmpy.sourceforge.net}{http://cbmpy.sourceforge.net})
Copyright (C) 2009-2015 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see \textless{}\href{http://www.gnu.org/licenses/}{http://www.gnu.org/licenses/}\textgreater{}

Author: Brett G. Olivier
Contact email: \href{mailto:bgoli@users.sourceforge.net}{bgoli@users.sourceforge.net}
Last edit: \$Author: bgoli \$ (\$Id: CBTools.py 414 2016-02-17 08:21:03Z bgoli \$)
\index{addFluxAsActiveObjective() (in module cbmpy.CBTools)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBTools.addFluxAsActiveObjective}\pysiglinewithargsret{\code{cbmpy.CBTools.}\bfcode{addFluxAsActiveObjective}}{\emph{f}, \emph{reaction\_id}, \emph{osense}, \emph{coefficient=1}}{}
Adds a flux as an active objective function
\begin{itemize}
\item {} 
\emph{reaction\_id} a string containing a reaction id

\item {} 
\emph{osense} objective sense must be \textbf{maximize} or \textbf{minimize}

\item {} 
\emph{coefficient} the objective funtion coefficient {[}default=1{]}

\end{itemize}

\end{fulllineitems}

\index{addGenesFromAnnotations() (in module cbmpy.CBTools)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBTools.addGenesFromAnnotations}\pysiglinewithargsret{\code{cbmpy.CBTools.}\bfcode{addGenesFromAnnotations}}{\emph{fba}, \emph{annotation\_key='GENE ASSOCIATION'}, \emph{gene\_pattern=None}}{}
THIS METHOD IS DERPRECATED PLEASE USE cmod.createGeneAssociationsFromAnnotations()

Add genes to the model using the definitions stored in the annotation key
\begin{itemize}
\item {} 
\emph{fba} and fba object

\item {} 
\emph{annotation\_key} the annotation dictionary key that holds the gene association for the protein/enzyme

\item {} 
\emph{gene\_pattern} deprecated, not needed anymore

\end{itemize}

\end{fulllineitems}

\index{addSinkReaction() (in module cbmpy.CBTools)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBTools.addSinkReaction}\pysiglinewithargsret{\code{cbmpy.CBTools.}\bfcode{addSinkReaction}}{\emph{fbam}, \emph{species}, \emph{lb=0.0}, \emph{ub=1000.0}}{}
Adds a sink reactions that consumes a model \emph{species} so that X --\textgreater{}
\begin{itemize}
\item {} 
\emph{fbam} an fba model object

\item {} 
\emph{species} a valid species name

\item {} 
\emph{lb} lower flux bound {[}default = 0.0{]}

\item {} 
\emph{ub} upper flux bound {[}default = 1000.0{]}

\end{itemize}

\end{fulllineitems}

\index{addSourceReaction() (in module cbmpy.CBTools)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBTools.addSourceReaction}\pysiglinewithargsret{\code{cbmpy.CBTools.}\bfcode{addSourceReaction}}{\emph{fbam}, \emph{species}, \emph{lb=0.0}, \emph{ub=1000.0}}{}
Adds a source reactions that produces a model \emph{species} so that --\textgreater{} X
\begin{itemize}
\item {} 
\emph{fbam} an fba model object

\item {} 
\emph{species} a valid species name

\item {} 
\emph{lb} lower flux bound {[}default = 0.0{]}

\item {} 
\emph{ub} upper flux bound {[}default = 1000.0{]}

\end{itemize}

Note reversiblity is determined by the lower bound, default 0 = irreversible. If
negative then reversible.

\end{fulllineitems}

\index{addStoichToFBAModel() (in module cbmpy.CBTools)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBTools.addStoichToFBAModel}\pysiglinewithargsret{\code{cbmpy.CBTools.}\bfcode{addStoichToFBAModel}}{\emph{fm}}{}
Build stoichiometry: this method has been refactored into the model class - cmod.buildStoichMatrix()

\end{fulllineitems}

\index{checkExchangeReactions() (in module cbmpy.CBTools)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBTools.checkExchangeReactions}\pysiglinewithargsret{\code{cbmpy.CBTools.}\bfcode{checkExchangeReactions}}{\emph{fba}, \emph{autocorrect=True}}{}
Scan all reactions for exchange reactions (reactions containing a boundary species), return a list of
inconsistent reactions or correct automatically.
\begin{itemize}
\item {} 
\emph{fba} a CBMPy model

\item {} 
\emph{autocorrect} {[}default=True{]} correctly set the ``is\_exchange'' attribute on a reaction

\end{itemize}

\end{fulllineitems}

\index{checkFluxBoundConsistency() (in module cbmpy.CBTools)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBTools.checkFluxBoundConsistency}\pysiglinewithargsret{\code{cbmpy.CBTools.}\bfcode{checkFluxBoundConsistency}}{\emph{fba}}{}
Check flux bound consistency checks for multiply defined bounds, bounds without a reaction, inconsistent bounds with respect to each other
and reaction reversbility. Returns a dictionary of bounds/reactions where errors occur.

\end{fulllineitems}

\index{checkIds() (in module cbmpy.CBTools)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBTools.checkIds}\pysiglinewithargsret{\code{cbmpy.CBTools.}\bfcode{checkIds}}{\emph{fba}, \emph{items='all'}}{}
Checks the id's of the specified model attributes to see if the name is legal and if there are duplicates.
Returns a list of items with errors.
\begin{itemize}
\item {} 
\emph{fba} a CBMPy model instance

\item {} 
\emph{items} {[}default='all'{]} `all' means `species,reactions,flux\_bounds,objectives' of which one or more can be specified

\end{itemize}

\end{fulllineitems}

\index{checkProducibility() (in module cbmpy.CBTools)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBTools.checkProducibility}\pysiglinewithargsret{\code{cbmpy.CBTools.}\bfcode{checkProducibility}}{\emph{mod}, \emph{metabolites=None}, \emph{reactions=None}, \emph{retOnlyZeroEntr=False}, \emph{zeroLimit=1e-11}}{}
Check for blocked metabolites by adding a sink reaction and maximizing its output. If no metabolites
are defined all metabolites are used by default. Returns a dictionary of metabolite
id and sink flux pairs:
\begin{itemize}
\item {} 
\emph{mod} a CBMPy model

\item {} 
\emph{metabolites} {[}default={[}{]}{]} if not specified by default uses all metabolites defined in model

\item {} 
\emph{reactions} {[}default={[}{]}{]} if defined, the reagents of each reaction listed here will be tested

\item {} 
\emph{retOnlyZeroEntr} {[}default=False{]} default returns all results, if this is try only blocked metabolites are returned

\item {} 
\emph{zeroLimit} {[}default=1.0e-11{]} values smaller than this are considered to be zero

\end{itemize}

This function was contributed by Willi Gottstein, Amsterdam, 2015.

\end{fulllineitems}

\index{checkProducibilityMetabolites() (in module cbmpy.CBTools)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBTools.checkProducibilityMetabolites}\pysiglinewithargsret{\code{cbmpy.CBTools.}\bfcode{checkProducibilityMetabolites}}{\emph{mod}, \emph{metabolites=None}, \emph{retOnlyZeroEntr=False}, \emph{zeroLimit=1e-11}}{}
Check for blocked metabolites by adding a sink reaction and maximizing its output. If no metabolites
are defined all metabolites are used by default. Returns a dictionary of metabolite
id and sink flux pairs:
\begin{itemize}
\item {} 
\emph{mod} a CBMPy model

\item {} 
\emph{metabolites} {[}default={[}{]}{]} if not specified by default uses all metabolites defined in model

\item {} 
\emph{reactions} {[}default={[}{]}{]} if defined, the reagents of each reaction listed here will be tested

\item {} 
\emph{retOnlyZeroEntr} {[}default=False{]} default returns all results, if this is try only blocked metabolites are returned

\item {} 
\emph{zeroLimit} {[}default=1.0e-11{]} values smaller than this are considered to be zero

\end{itemize}

This function was contributed by Willi Gottstein, Amsterdam, 2015.

\end{fulllineitems}

\index{checkProducibilityReactions() (in module cbmpy.CBTools)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBTools.checkProducibilityReactions}\pysiglinewithargsret{\code{cbmpy.CBTools.}\bfcode{checkProducibilityReactions}}{\emph{mod}, \emph{reactions=None}, \emph{retOnlyZeroEntr=False}, \emph{zeroLimit=1e-11}}{}
Check for blocked metabolites by adding a sink reaction to each reaction reagent and maximizing
its output. Returns a dictionary of reagent/metabolite id and sink flux pairs:
\begin{itemize}
\item {} 
\emph{mod} a CBMPy model

\item {} 
\emph{reactions} {[}default={[}{]}{]} if defined, the reagents of each reaction listed here will be tested

\item {} 
\emph{retOnlyZeroEntr} {[}default=False{]} default returns all results, if this is try only blocked metabolites are returned

\item {} 
\emph{zeroLimit} {[}default=1.0e-11{]} values smaller than this are considered to be zero

\end{itemize}

This function was contributed by Willi Gottstein, Amsterdam, 2015.

\end{fulllineitems}

\index{checkReactionBalanceElemental() (in module cbmpy.CBTools)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBTools.checkReactionBalanceElemental}\pysiglinewithargsret{\code{cbmpy.CBTools.}\bfcode{checkReactionBalanceElemental}}{\emph{f}, \emph{Rid=None}, \emph{zero\_tol=1e-12}}{}
Check if the reaction is balanced using the chemical formula
\begin{itemize}
\item {} 
\emph{f} the FBA object

\item {} 
\emph{Rid} {[}default = None{]} the reaction to check, defaults to all

\item {} 
\emph{zero\_tol} {[}default=1.0e-12{]} the floating point zero used for elemental balancing

\end{itemize}

This function is derived from the code found here: \href{http://pyparsing.wikispaces.com/file/view/chemicalFormulas.py}{http://pyparsing.wikispaces.com/file/view/chemicalFormulas.py}

\end{fulllineitems}

\index{checkSuffixes() (in module cbmpy.CBTools)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBTools.checkSuffixes}\pysiglinewithargsret{\code{cbmpy.CBTools.}\bfcode{checkSuffixes}}{\emph{aList}, \emph{suf1}, \emph{suf2}}{}
Check whether there are strings in aList with the suffixes suf1 and suf2, respectively
used in the function getReaByMetSuf

\end{fulllineitems}

\index{createTempFileName() (in module cbmpy.CBTools)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBTools.createTempFileName}\pysiglinewithargsret{\code{cbmpy.CBTools.}\bfcode{createTempFileName}}{}{}
Return a temporary filename

\end{fulllineitems}

\index{createZipArchive() (in module cbmpy.CBTools)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBTools.createZipArchive}\pysiglinewithargsret{\code{cbmpy.CBTools.}\bfcode{createZipArchive}}{\emph{zipname}, \emph{files}, \emph{move=False}, \emph{compression='normal'}}{}
Create a zip archive which contains one or more files
\begin{itemize}
\item {} 
\emph{zipname} the name of the zip archive to create (fully qualified)

\item {} 
\emph{files} either a valid filename or a list of filenames (fully qualified)

\item {} 
\emph{move} {[}default=False{]} attempt to delete input files after zip-archive creation

\item {} 
\emph{compression} {[}default='normal'{]} normal zip compression, set as None for no compression only store files (zlib not required)

\end{itemize}

\end{fulllineitems}

\index{deSerialize() (in module cbmpy.CBTools)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBTools.deSerialize}\pysiglinewithargsret{\code{cbmpy.CBTools.}\bfcode{deSerialize}}{\emph{s}}{}
Deserializes a serialised object contained in a string

\end{fulllineitems}

\index{exportArray2CSV() (in module cbmpy.CBTools)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBTools.exportArray2CSV}\pysiglinewithargsret{\code{cbmpy.CBTools.}\bfcode{exportArray2CSV}}{\emph{arr}, \emph{fname}}{}
Export an array to fname.csv
\begin{itemize}
\item {} 
\emph{arr} the an array like object

\item {} 
\emph{fname} the output filename

\item {} 
\emph{sep} {[}default=','{]} the column separator

\end{itemize}

\end{fulllineitems}

\index{exportArray2TXT() (in module cbmpy.CBTools)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBTools.exportArray2TXT}\pysiglinewithargsret{\code{cbmpy.CBTools.}\bfcode{exportArray2TXT}}{\emph{arr}, \emph{fname}}{}
Export an array to fname.txt
\begin{itemize}
\item {} 
\emph{arr} the an array like object

\item {} 
\emph{fname} the output filename

\item {} 
\emph{sep} {[}default=','{]} the column separator

\end{itemize}

\end{fulllineitems}

\index{exportLabelledArray() (in module cbmpy.CBTools)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBTools.exportLabelledArray}\pysiglinewithargsret{\code{cbmpy.CBTools.}\bfcode{exportLabelledArray}}{\emph{arr}, \emph{fname}, \emph{names=None}, \emph{sep='}, \emph{`}, \emph{fmt='\%f'}}{}
Write a 2D array type object to file
\begin{itemize}
\item {} 
\emph{arr} the an array like object

\item {} 
\emph{names} {[}default=None{]} the list of row names

\item {} 
\emph{fname} the output filename

\item {} 
\emph{sep} {[}default=','{]} the column separator

\item {} 
\emph{fmt} {[}default='\%s'{]} the output number format

\end{itemize}

\end{fulllineitems}

\index{exportLabelledArray2CSV() (in module cbmpy.CBTools)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBTools.exportLabelledArray2CSV}\pysiglinewithargsret{\code{cbmpy.CBTools.}\bfcode{exportLabelledArray2CSV}}{\emph{arr}, \emph{fname}, \emph{names=None}}{}
Export an array with row names to fname.csv
\begin{itemize}
\item {} 
\emph{arr} the an array like object

\item {} 
\emph{fname} the output filename

\item {} 
\emph{names} {[}default=None{]} the list of row names

\end{itemize}

\end{fulllineitems}

\index{exportLabelledArray2TXT() (in module cbmpy.CBTools)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBTools.exportLabelledArray2TXT}\pysiglinewithargsret{\code{cbmpy.CBTools.}\bfcode{exportLabelledArray2TXT}}{\emph{arr}, \emph{fname}, \emph{names=None}}{}
Export an array with row names to fname.txt
\begin{itemize}
\item {} 
\emph{arr} the an array like object

\item {} 
\emph{names} {[}default=None{]} the list of row names

\item {} 
\emph{fname} the output filename

\end{itemize}

\end{fulllineitems}

\index{exportLabelledArrayWithHeader() (in module cbmpy.CBTools)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBTools.exportLabelledArrayWithHeader}\pysiglinewithargsret{\code{cbmpy.CBTools.}\bfcode{exportLabelledArrayWithHeader}}{\emph{arr}, \emph{fname}, \emph{names=None}, \emph{header=None}, \emph{sep='}, \emph{`}, \emph{fmt='\%f'}}{}
Export an array with row names and header
\begin{itemize}
\item {} 
\emph{arr} the an array like object

\item {} 
\emph{names} {[}default=None{]} the list of row names

\item {} 
\emph{header} {[}default=None{]} the list of column names

\item {} 
\emph{fname} the output filename

\item {} 
\emph{sep} {[}default=','{]} the column separator

\item {} 
\emph{fmt} {[}default='\%s'{]} the output number format

\item {} 
\emph{appendlist} {[}default=False{]} if True append the array to \emph{fname} otherwise create a new file

\end{itemize}

\end{fulllineitems}

\index{exportLabelledArrayWithHeader2CSV() (in module cbmpy.CBTools)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBTools.exportLabelledArrayWithHeader2CSV}\pysiglinewithargsret{\code{cbmpy.CBTools.}\bfcode{exportLabelledArrayWithHeader2CSV}}{\emph{arr}, \emph{fname}, \emph{names=None}, \emph{header=None}}{}
Export an array with row names and header to fname.csv
\begin{itemize}
\item {} 
\emph{arr} the an array like object

\item {} 
\emph{fname} the output filename

\item {} 
\emph{names} {[}default=None{]} the list of row names

\item {} 
\emph{header} {[}default=None{]} the list of column names

\end{itemize}

\end{fulllineitems}

\index{exportLabelledArrayWithHeader2TXT() (in module cbmpy.CBTools)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBTools.exportLabelledArrayWithHeader2TXT}\pysiglinewithargsret{\code{cbmpy.CBTools.}\bfcode{exportLabelledArrayWithHeader2TXT}}{\emph{arr}, \emph{fname}, \emph{names=None}, \emph{header=None}}{}
Export an array with row names and header to fname.txt
\begin{itemize}
\item {} 
\emph{arr} the an array like object

\item {} 
\emph{names} the list of row names

\item {} 
\emph{header} the list of column names

\item {} 
\emph{fname} the output filename

\end{itemize}

\end{fulllineitems}

\index{exportLabelledLinkedList() (in module cbmpy.CBTools)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBTools.exportLabelledLinkedList}\pysiglinewithargsret{\code{cbmpy.CBTools.}\bfcode{exportLabelledLinkedList}}{\emph{arr}, \emph{fname}, \emph{names=None}, \emph{sep='}, \emph{`}, \emph{fmt='\%s'}, \emph{appendlist=False}}{}
Write a 2D linked list {[}{[}...{]},{[}...{]},{[}...{]},{[}...{]}{]} and optionally a list of row labels to file:
\begin{itemize}
\item {} 
\emph{arr} the linked list

\item {} 
\emph{fname} the output filename

\item {} 
\emph{names} {[}default=None{]} the list of row names

\item {} 
\emph{sep} {[}default=','{]} the column separator

\item {} 
\emph{fmt} {[}default='\%s'{]} the output number format

\item {} 
\emph{appendlist} {[}default=False{]} if True append the array to \emph{fname} otherwise create a new file

\end{itemize}

\end{fulllineitems}

\index{findDeadEndMetabolites() (in module cbmpy.CBTools)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBTools.findDeadEndMetabolites}\pysiglinewithargsret{\code{cbmpy.CBTools.}\bfcode{findDeadEndMetabolites}}{\emph{fbam}}{}
Finds dead-end (single reaction) metabolites rows in N with a single entry), returns a list of (metabolite, reaction) ids

\end{fulllineitems}

\index{findDeadEndReactions() (in module cbmpy.CBTools)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBTools.findDeadEndReactions}\pysiglinewithargsret{\code{cbmpy.CBTools.}\bfcode{findDeadEndReactions}}{\emph{fbam}}{}
Finds dead-end (single substrate/product) reactions (cols in N with a single entry), returns a list of (metabolite, reaction) ids

\end{fulllineitems}

\index{fixReversibility() (in module cbmpy.CBTools)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBTools.fixReversibility}\pysiglinewithargsret{\code{cbmpy.CBTools.}\bfcode{fixReversibility}}{\emph{fbam}, \emph{auto\_correct=False}}{}
Set fluxbound lower bound from reactions reversibility information.
\begin{itemize}
\item {} 
\emph{fbam} and FBAModel instance

\item {} 
\emph{auto\_correct} (default=False) if True automatically sets lower bound to zero if required, otherwise prints a warning if false.

\end{itemize}

\end{fulllineitems}

\index{getBoundsDict() (in module cbmpy.CBTools)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBTools.getBoundsDict}\pysiglinewithargsret{\code{cbmpy.CBTools.}\bfcode{getBoundsDict}}{\emph{fbamod}, \emph{substring=None}}{}
Return a dictionary of reactions\&bounds

\end{fulllineitems}

\index{getExchBoundsDict() (in module cbmpy.CBTools)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBTools.getExchBoundsDict}\pysiglinewithargsret{\code{cbmpy.CBTools.}\bfcode{getExchBoundsDict}}{\emph{fbamod}}{}
Return a dictionary of all exchange reactions (as determined by the is\_exchange attribute of Reaction)
\begin{itemize}
\item {} 
\emph{fbamod} a CBMPy model

\end{itemize}

\end{fulllineitems}

\index{getModelGenesPerReaction() (in module cbmpy.CBTools)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBTools.getModelGenesPerReaction}\pysiglinewithargsret{\code{cbmpy.CBTools.}\bfcode{getModelGenesPerReaction}}{\emph{fba}, \emph{gene\_pattern=None}, \emph{gene\_annotation\_key='GENE ASSOCIATION'}}{}
Parse a BiGG style gene annotation string using default gene\_pattern='((W*w*W*))' or
(\textless{}any non-alphanum\textgreater{}\textless{}any alphanum\textgreater{}\textless{}any non-alphanum\textgreater{})

Old eColi specific pattern `(bw*W)'

It is advisable to use the model methods directly rather than this function

\end{fulllineitems}

\index{getReaByMetSuf() (in module cbmpy.CBTools)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBTools.getReaByMetSuf}\pysiglinewithargsret{\code{cbmpy.CBTools.}\bfcode{getReaByMetSuf}}{\emph{fba\_mod}, \emph{suf1}, \emph{suf2}, \emph{retSpec=False}}{}~\begin{itemize}
\item {} 
can be used to determine all reactions in which at least two species with different suffixes are involved

\item {} 
e.g. getReaByMetSuf(fba\_mod, `\_e', `\_c') returns all reactions IDs between the extracellular compartment (suffix

\end{itemize}

`\_e') and the cytosol (suffix `\_c)'.

INPUT:
fba\_mod: a model instance
suf1: suffix one (string)
suf2: suffix two (string)

OUTPUT:
if retSpec=True, a dictionary of reaction IDs and their associated species are returned
if retSpec=False, a list with reaction IDs is returned

\end{fulllineitems}

\index{loadObj() (in module cbmpy.CBTools)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBTools.loadObj}\pysiglinewithargsret{\code{cbmpy.CBTools.}\bfcode{loadObj}}{\emph{filename}}{}
Loads a serialised Python pickle from \emph{filename}.dat returns the Python object(s)

\end{fulllineitems}

\index{merge2Models() (in module cbmpy.CBTools)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBTools.merge2Models}\pysiglinewithargsret{\code{cbmpy.CBTools.}\bfcode{merge2Models}}{\emph{m1}, \emph{m2}, \emph{ignore=None}, \emph{ignore\_duplicate\_ids=False}}{}
Merge 2 models, this method does a raw merge of model 2 into model 1 without any model checking.
Component id's in ignore are ignored in both models and the first objective of model 1 is arbitrarily
set as active. Compartments are also merged and a new ``OuterMerge'' compartment is also created.

In all cases duplicate id's are tracked and ignored, essentially using the object id encountered first -
usually that of model 1. Duplicate checking can be disabled by setting the \emph{ignore\_duplicate\_ids} flag.
\begin{itemize}
\item {} 
\emph{m1} model 1

\item {} 
\emph{m2} model 2

\item {} 
\emph{ignore} {[}{[}{]}{]} do not merge these id's

\item {} 
\emph{ignore\_duplicate\_ids} {[}False{]} default behaviour that can be enabled

\end{itemize}

In development: merging genes and gpr's.

\end{fulllineitems}

\index{processBiGGannotationNote() (in module cbmpy.CBTools)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBTools.processBiGGannotationNote}\pysiglinewithargsret{\code{cbmpy.CBTools.}\bfcode{processBiGGannotationNote}}{\emph{fba}, \emph{annotation\_key='note'}}{}
Parse the HTML formatted reaction information stored in the BiGG notes field.
This function is being deprecated and replaced by \emph{CBTools.processSBMLAnnotationNotes()}
\begin{itemize}
\item {} 
requires an \emph{annotation\_key} which contains a BiGG HTML fragment

\end{itemize}

\end{fulllineitems}

\index{processBiGGchemFormula() (in module cbmpy.CBTools)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBTools.processBiGGchemFormula}\pysiglinewithargsret{\code{cbmpy.CBTools.}\bfcode{processBiGGchemFormula}}{\emph{fba}}{}
Disambiguates the overloaded BiGG name NAME\_CHEMFORMULA into
\begin{itemize}
\item {} 
\emph{species.name} NAME

\item {} 
\emph{species.chemFormula} CHEMFORMULA

\end{itemize}

\end{fulllineitems}

\index{processExchangeReactions() (in module cbmpy.CBTools)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBTools.processExchangeReactions}\pysiglinewithargsret{\code{cbmpy.CBTools.}\bfcode{processExchangeReactions}}{\emph{fba}, \emph{key}}{}
Extract exchange reactions from model using \emph{key} and return:
\begin{itemize}
\item {} 
a dictionary of all exchange reactions without \emph{medium} reactions

\item {} 
a dictionary of \emph{medium} exchange reactions (negative lower bound)

\end{itemize}

\end{fulllineitems}

\index{processSBMLAnnotationNotes() (in module cbmpy.CBTools)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBTools.processSBMLAnnotationNotes}\pysiglinewithargsret{\code{cbmpy.CBTools.}\bfcode{processSBMLAnnotationNotes}}{\emph{fba}, \emph{annotation\_key='note'}}{}
Parse the HTML formatted reaction information stored in the SBML notes field currently
processes BiGG and PySCeSCBM style annotations it looks for the the annotation indexed
with the \emph{annotation\_key}
\begin{itemize}
\item {} 
\emph{annotation\_key} {[}default='note'{]} which contains a HTML/XHTML fragment in BiGG/PySCeSCBM format

\end{itemize}

\end{fulllineitems}

\index{removeFixedSpeciesReactions() (in module cbmpy.CBTools)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBTools.removeFixedSpeciesReactions}\pysiglinewithargsret{\code{cbmpy.CBTools.}\bfcode{removeFixedSpeciesReactions}}{\emph{f}}{}
This function is a hack that removes reactions which only have boundary species as reactants
and products. These are typically gene associations encoded in the Manchester style and there
is probably a better way of working around this problem ...
\begin{itemize}
\item {} 
\emph{f} an instantiated fba model object

\end{itemize}

\end{fulllineitems}

\index{roundOffWithSense() (in module cbmpy.CBTools)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBTools.roundOffWithSense}\pysiglinewithargsret{\code{cbmpy.CBTools.}\bfcode{roundOffWithSense}}{\emph{val}, \emph{osense='max'}, \emph{tol=1e-08}}{}
Round of a value in a way that takes into consideration the sense of the operation that generated it
\begin{itemize}
\item {} 
\emph{val} the value

\item {} 
\emph{osense} {[}default='max'{]} the sense

\item {} 
\emph{tol} {[}default=1e-8{]} the tolerance of the roundoff factor

\end{itemize}

\end{fulllineitems}

\index{scanForReactionDuplicates() (in module cbmpy.CBTools)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBTools.scanForReactionDuplicates}\pysiglinewithargsret{\code{cbmpy.CBTools.}\bfcode{scanForReactionDuplicates}}{\emph{f}, \emph{ignore\_coefficients=False}}{}
This method uses uses a brute force apprach to finding reactions with matching
stoichiometry

\end{fulllineitems}

\index{scanForUnbalancedReactions() (in module cbmpy.CBTools)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBTools.scanForUnbalancedReactions}\pysiglinewithargsret{\code{cbmpy.CBTools.}\bfcode{scanForUnbalancedReactions}}{\emph{f}, \emph{output='all'}}{}
Scan a model for unbalanced reactions, returns a tuple of dictionaries balanced and unbalanced:
\begin{itemize}
\item {} 
\emph{f} an FBA model instance

\item {} 
\emph{output} {[}default='all'{]} can be one of {[}'all','charge','element'{]}

\item {} 
\emph{charge} return all charge \textbf{un} balanced reactions

\item {} 
\emph{element} return all element \textbf{un} balanced reactions

\end{itemize}

\end{fulllineitems}

\index{setSpeciesPropertiesFromAnnotations() (in module cbmpy.CBTools)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBTools.setSpeciesPropertiesFromAnnotations}\pysiglinewithargsret{\code{cbmpy.CBTools.}\bfcode{setSpeciesPropertiesFromAnnotations}}{\emph{fbam}, \emph{overwriteCharge=False}, \emph{overwriteChemFormula=False}}{}
This will attempt to set the model Species properties from the annotation. With the default options
it will only replace missing data. With ChemicalFormula this is easy to detect however charge may
have an ``unknown value'' of 0. Setting the optional values to true will replace any existing value
with any valid annotation.
\begin{itemize}
\item {} 
\emph{overwriteChemFormula} {[}default=False{]}

\item {} 
\emph{overwriteCharge} {[}default=False{]}

\end{itemize}

\end{fulllineitems}

\index{splitReversibleReactions() (in module cbmpy.CBTools)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBTools.splitReversibleReactions}\pysiglinewithargsret{\code{cbmpy.CBTools.}\bfcode{splitReversibleReactions}}{\emph{fba}, \emph{selected\_reactions=None}}{}
Split a (set of) reactions into reversible reactions returns a copy of the original model

R1: A = B
R1f: A -\textgreater{} B
R1r: B -\textgreater{} A
\begin{itemize}
\item {} 
\emph{fba} an instantiated CBMPy model object

\item {} 
\emph{selected\_reactions} if a reversible reaction id is in here split it

\end{itemize}

\end{fulllineitems}

\index{splitSingleReversibleReaction() (in module cbmpy.CBTools)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBTools.splitSingleReversibleReaction}\pysiglinewithargsret{\code{cbmpy.CBTools.}\bfcode{splitSingleReversibleReaction}}{\emph{fba}, \emph{rid}, \emph{fwd\_id=None}, \emph{rev\_id=None}}{}
Split a single reversible reaction into two irreversible reactions, returns the original reversible reaction and bounds
while deleting them from model.

R1: A = B
R1\_fwd: A -\textgreater{} B
R1\_rev: B -\textgreater{} A
\begin{itemize}
\item {} 
\emph{fba} an instantiated CBMPy model object

\item {} 
\emph{rid} a valid reaction id

\item {} 
\emph{fwd\_id} {[}default=None{]} the new forward reaction id, defaults to rid\_fwd

\item {} 
\emph{rev\_id} {[}default=None{]} the new forward reaction id, defaults to rid\_rev

\end{itemize}

\end{fulllineitems}

\index{storeObj() (in module cbmpy.CBTools)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBTools.storeObj}\pysiglinewithargsret{\code{cbmpy.CBTools.}\bfcode{storeObj}}{\emph{obj}, \emph{filename}, \emph{compress=False}}{}
Stores a Python \emph{obj} as a serialised binary object in \emph{filename}.dat
\begin{itemize}
\item {} 
\emph{obj} a python object

\item {} 
\emph{filename} the base filename

\item {} 
\emph{compress} {[}False{]} use gzip compression not \emph{implemented}

\end{itemize}

\end{fulllineitems}

\index{stringReplace() (in module cbmpy.CBTools)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBTools.stringReplace}\pysiglinewithargsret{\code{cbmpy.CBTools.}\bfcode{stringReplace}}{\emph{fbamod}, \emph{old}, \emph{new}, \emph{target}}{}
This is alpha stuff, target can be:
\begin{itemize}
\item {} 
`species'

\item {} 
`reactions'

\item {} 
`constraints'

\item {} 
`objectives'

\item {} 
`all'

\end{itemize}

\end{fulllineitems}

\phantomsection\label{modules_doc:module-cbmpy.CBWrite}\index{cbmpy.CBWrite (module)}

\section{CBMPy: CBWrite module}
\label{modules_doc:cbmpy-cbwrite-module}
PySCeS Constraint Based Modelling (\href{http://cbmpy.sourceforge.net}{http://cbmpy.sourceforge.net})
Copyright (C) 2009-2015 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see \textless{}\href{http://www.gnu.org/licenses/}{http://www.gnu.org/licenses/}\textgreater{}

Author: Brett G. Olivier
Contact email: \href{mailto:bgoli@users.sourceforge.net}{bgoli@users.sourceforge.net}
Last edit: \$Author: bgoli \$ (\$Id: CBWrite.py 411 2016-01-27 12:32:01Z bgoli \$)
\index{BuildHformatFluxBounds() (in module cbmpy.CBWrite)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBWrite.BuildHformatFluxBounds}\pysiglinewithargsret{\code{cbmpy.CBWrite.}\bfcode{BuildHformatFluxBounds}}{\emph{fba}, \emph{infinity\_replace=None}, \emph{use\_rational=False}}{}
Build and return a csio that contains the flux bounds in H format
\begin{itemize}
\item {} 
\emph{fba} a PySCeS-CBM FBA object

\item {} 
\emph{infinity\_replace} {[}default=None{]} if defined this is the abs(value) of +-\textless{}infinity\textgreater{}

\end{itemize}

\end{fulllineitems}

\index{BuildLPConstraints() (in module cbmpy.CBWrite)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBWrite.BuildLPConstraints}\pysiglinewithargsret{\code{cbmpy.CBWrite.}\bfcode{BuildLPConstraints}}{\emph{fba}, \emph{use\_rational=False}}{}
Build and return a csio that contains constraint constructed from
the StoichiometeryLP object
\begin{itemize}
\item {} 
\emph{fba} an fba model object which has a stoichiometry

\item {} 
\emph{use\_rational} write rational number output {[}default=False{]}

\end{itemize}

\end{fulllineitems}

\index{BuildLPConstraintsMath() (in module cbmpy.CBWrite)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBWrite.BuildLPConstraintsMath}\pysiglinewithargsret{\code{cbmpy.CBWrite.}\bfcode{BuildLPConstraintsMath}}{\emph{fba}, \emph{use\_rational=False}}{}
Build and return a csio that contains the constaints in LP format
Strict refers to dS/dt =\textgreater{} 0 and dS/dt \textless{}= 0

\end{fulllineitems}

\index{BuildLPConstraintsRelaxed() (in module cbmpy.CBWrite)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBWrite.BuildLPConstraintsRelaxed}\pysiglinewithargsret{\code{cbmpy.CBWrite.}\bfcode{BuildLPConstraintsRelaxed}}{\emph{fba}}{}
Build and return a csio that contains the constaints in LP format
Relaxed refers to dS/dt \textgreater{}= 0

\end{fulllineitems}

\index{BuildLPConstraintsStrict() (in module cbmpy.CBWrite)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBWrite.BuildLPConstraintsStrict}\pysiglinewithargsret{\code{cbmpy.CBWrite.}\bfcode{BuildLPConstraintsStrict}}{\emph{fba}, \emph{use\_rational=False}}{}
Build and return a csio that contains the constaints in LP format
Strict refers to dS/dt = 0

\end{fulllineitems}

\index{BuildLPFluxBounds() (in module cbmpy.CBWrite)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBWrite.BuildLPFluxBounds}\pysiglinewithargsret{\code{cbmpy.CBWrite.}\bfcode{BuildLPFluxBounds}}{\emph{fba}, \emph{use\_rational=False}}{}
Build and return a csio that contains the flux bounds in LP format

\end{fulllineitems}

\index{BuildLPUserConstraints() (in module cbmpy.CBWrite)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBWrite.BuildLPUserConstraints}\pysiglinewithargsret{\code{cbmpy.CBWrite.}\bfcode{BuildLPUserConstraints}}{\emph{fba}, \emph{use\_rational=False}}{}
Build and return a csio that contains constraint constructed from
the StoichiometeryLP object
\begin{itemize}
\item {} 
\emph{fba} an fba model object which has a stoichiometry

\item {} 
\emph{use\_rational} write rational number output {[}default=False{]}

\end{itemize}

\end{fulllineitems}

\index{WriteFVAdata() (in module cbmpy.CBWrite)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBWrite.WriteFVAdata}\pysiglinewithargsret{\code{cbmpy.CBWrite.}\bfcode{WriteFVAdata}}{\emph{fva}, \emph{names}, \emph{fname}, \emph{work\_dir=None}, \emph{roundec=None}, \emph{scale\_min=False}, \emph{appendfile=False}, \emph{info=None}}{}
INFO: this method will be deprecated please update your scripts to use ``writeFVAdata()''

\end{fulllineitems}

\index{WriteFVAtoCSV() (in module cbmpy.CBWrite)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBWrite.WriteFVAtoCSV}\pysiglinewithargsret{\code{cbmpy.CBWrite.}\bfcode{WriteFVAtoCSV}}{\emph{id}, \emph{fva}, \emph{names}, \emph{Dir=None}, \emph{fbaObj=None}}{}
INFO: this method will be deprecated please update your scripts to use ``writeFVAtoCSV()''

\end{fulllineitems}

\index{WriteModelHFormatFBA() (in module cbmpy.CBWrite)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBWrite.WriteModelHFormatFBA}\pysiglinewithargsret{\code{cbmpy.CBWrite.}\bfcode{WriteModelHFormatFBA}}{\emph{fba}, \emph{work\_dir=None}, \emph{use\_rational=False}, \emph{fullLP=True}, \emph{format='\%s'}, \emph{infinity\_replace=None}}{}
INFO: this method will be deprecated please update your scripts to use ``writeModelHFormatFBA2()''

\end{fulllineitems}

\index{WriteModelHFormatFBA2() (in module cbmpy.CBWrite)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBWrite.WriteModelHFormatFBA2}\pysiglinewithargsret{\code{cbmpy.CBWrite.}\bfcode{WriteModelHFormatFBA2}}{\emph{fba}, \emph{fname=None}, \emph{work\_dir=None}, \emph{use\_rational=False}, \emph{fullLP=True}, \emph{format='\%s'}, \emph{infinity\_replace=None}}{}
INFO: this method will be deprecated please update your scripts to use ``writeModelHFormatFBA2()''

\end{fulllineitems}

\index{WriteModelLP() (in module cbmpy.CBWrite)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBWrite.WriteModelLP}\pysiglinewithargsret{\code{cbmpy.CBWrite.}\bfcode{WriteModelLP}}{\emph{fba}, \emph{work\_dir=None}, \emph{fname=None}, \emph{multisymb=' `}, \emph{format='\%s'}, \emph{use\_rational=False}, \emph{constraint\_mode=None}, \emph{quiet=False}}{}
INFO: this method will be deprecated please update your scripts to use ``writeModelLP()''

\end{fulllineitems}

\index{WriteModelLPOld() (in module cbmpy.CBWrite)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBWrite.WriteModelLPOld}\pysiglinewithargsret{\code{cbmpy.CBWrite.}\bfcode{WriteModelLPOld}}{\emph{fba}, \emph{work\_dir=None}, \emph{multisymb=' `}, \emph{lpt=True}, \emph{constraint\_mode='strict'}, \emph{use\_rational=False}, \emph{format='\%s'}}{}
INFO: this method will be deprecated please update your scripts to use ``writeModelLPOld()''

\end{fulllineitems}

\index{WriteModelRaw() (in module cbmpy.CBWrite)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBWrite.WriteModelRaw}\pysiglinewithargsret{\code{cbmpy.CBWrite.}\bfcode{WriteModelRaw}}{\emph{fba}, \emph{work\_dir=None}}{}
INFO: this method will be deprecated please update your scripts to use ``writeModelRaw()''

\end{fulllineitems}

\index{convertExcelToFloat() (in module cbmpy.CBWrite)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBWrite.convertExcelToFloat}\pysiglinewithargsret{\code{cbmpy.CBWrite.}\bfcode{convertExcelToFloat}}{\emph{num}}{}
Converts an Excel ``number'' to a float
\begin{itemize}
\item {} 
\emph{num} a number

\end{itemize}

\end{fulllineitems}

\index{convertFloatToExcel() (in module cbmpy.CBWrite)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBWrite.convertFloatToExcel}\pysiglinewithargsret{\code{cbmpy.CBWrite.}\bfcode{convertFloatToExcel}}{\emph{num}, \emph{roundoff}}{}
Converts a float to Excel compatible ``number''
\begin{itemize}
\item {} 
\emph{num} a number

\item {} 
\emph{roundoff} the number of roundoff digits for round()

\end{itemize}

\end{fulllineitems}

\index{exportModel() (in module cbmpy.CBWrite)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBWrite.exportModel}\pysiglinewithargsret{\code{cbmpy.CBWrite.}\bfcode{exportModel}}{\emph{fba}, \emph{fname=None}, \emph{fmt='lp'}, \emph{work\_dir=None}, \emph{use\_rational='both'}}{}
Export the FBA model in different formats:
\begin{itemize}
\item {} 
\emph{fba} the FBA model

\item {} 
\emph{fname} {[}default=None{]} the exported filename if None then \emph{fba.getPid()} is used

\item {} 
\emph{fmt} {[}default='lp'{]} the export format can be one of: `lp' (CPLEX), `hformat' (Polyhedra), `all' (both)

\item {} 
\emph{use\_rational} {[}default='both'{]} if \emph{all} or \emph{hformat} is specified should hformat files be written using rational math or not. The default \emph{both} is the legacy behaviour and writes both.

\end{itemize}

Note that `hformat' ignores `fname' and only uses fba.getPid() this is a legacy behaviour

\end{fulllineitems}

\index{generateBGID() (in module cbmpy.CBWrite)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBWrite.generateBGID}\pysiglinewithargsret{\code{cbmpy.CBWrite.}\bfcode{generateBGID}}{\emph{num}, \emph{prefix}}{}~\begin{description}
\item[{Create a BGID generator, which is \textless{}prefix\textgreater{}\textless{}num\textgreater{} where perfix is two letters num is padded to 6 figures}] \leavevmode\begin{itemize}
\item {} 
\emph{num} the starting number

\item {} 
\emph{prefix} the two letter prefix

\end{itemize}

\end{description}

\end{fulllineitems}

\index{printFBASolution() (in module cbmpy.CBWrite)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBWrite.printFBASolution}\pysiglinewithargsret{\code{cbmpy.CBWrite.}\bfcode{printFBASolution}}{\emph{fba}, \emph{include\_all=False}}{}
Prints the FBA optimal solution to the screen.
\begin{itemize}
\item {} 
\emph{fba} an FBA model object

\item {} 
\emph{include\_all} include all variables

\end{itemize}

\end{fulllineitems}

\index{writeCOBRASBML() (in module cbmpy.CBWrite)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBWrite.writeCOBRASBML}\pysiglinewithargsret{\code{cbmpy.CBWrite.}\bfcode{writeCOBRASBML}}{\emph{fba}, \emph{fname}, \emph{directory=None}}{}
Takes an FBA model object and writes it to file as a COBRA compatible :
\begin{itemize}
\item {} 
\emph{fba} an fba model object

\item {} 
\emph{fname} the model will be written as XML to \emph{fname}

\item {} 
\emph{directory} {[}default=None{]} if defined it is prepended to fname

\end{itemize}

\end{fulllineitems}

\index{writeFVAdata() (in module cbmpy.CBWrite)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBWrite.writeFVAdata}\pysiglinewithargsret{\code{cbmpy.CBWrite.}\bfcode{writeFVAdata}}{\emph{fvadata}, \emph{names}, \emph{fname}, \emph{work\_dir=None}, \emph{roundec=None}, \emph{scale\_min=False}, \emph{appendfile=False}, \emph{info=None}}{}
Takes the resuls of a FluxVariabilityAnalysis method and writes it to a nice
csv file. Note this method replaces the glpk/cplx\_WriteFVAtoCSV methods. Data is output as a csv file
with columns: FluxName, FVA\_MIN, FVA\_MAX, OPT\_VAL, SPAN
\begin{itemize}
\item {} 
\emph{fvadata} FluxVariabilityAnalysis() FVA OUTPUT\_ARRAY

\item {} 
\emph{names} FluxVariabilityAnalysis() FVA OUTPUT\_NAMES

\item {} 
\emph{fname} filename\_base for the CSV output

\item {} 
\emph{work\_dir} {[}default=None{]} if set the output directory for the csv files

\item {} 
\emph{roundec} {[}default=None{]} an integer indicating at which decimal to round off output. Default is no rounding.

\item {} 
\emph{scale\_min} {[}default=False{]} normalise each flux such that that FVA\_MIN = 0.0

\item {} 
\emph{appendfile} {[}default=False{]} instead of opening a new file try and append the data

\item {} 
\emph{info} {[}default=None{]} a string added to the results as an extra column, useful with \emph{appendfile}

\end{itemize}

\end{fulllineitems}

\index{writeFVAtoCSV() (in module cbmpy.CBWrite)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBWrite.writeFVAtoCSV}\pysiglinewithargsret{\code{cbmpy.CBWrite.}\bfcode{writeFVAtoCSV}}{\emph{fvadata}, \emph{names}, \emph{fname}, \emph{Dir=None}, \emph{fbaObj=None}}{}
Takes the resuls of a FluxVariabilityAnalysis method and writes it to a nice
csv file. Note this method replaces the glpk/cplx\_WriteFVAtoCSV methods.
\begin{itemize}
\item {} 
\emph{fvadata} FluxVariabilityAnalysis() OUTPUT\_ARRAY

\item {} 
\emph{names} FluxVariabilityAnalysis() OUTPUT\_NAMES

\item {} 
\emph{fname} filename\_base for the CSV output

\item {} 
\emph{Dir} {[}default=None{]} if set the output directory for the csv files

\item {} 
\emph{fbaObj} {[}default=None{]} if supplied adds extra model information into the output tables

\end{itemize}

\end{fulllineitems}

\index{writeMinDistanceLPwithCost() (in module cbmpy.CBWrite)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBWrite.writeMinDistanceLPwithCost}\pysiglinewithargsret{\code{cbmpy.CBWrite.}\bfcode{writeMinDistanceLPwithCost}}{\emph{fname}, \emph{fbas}, \emph{work\_dir=None}, \emph{ignoreDistance={[}{]}}, \emph{constraint\_mode='strict'}}{}
For backwards compatability only

\end{fulllineitems}

\index{writeModelHFormatFBA() (in module cbmpy.CBWrite)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBWrite.writeModelHFormatFBA}\pysiglinewithargsret{\code{cbmpy.CBWrite.}\bfcode{writeModelHFormatFBA}}{\emph{fba}, \emph{work\_dir=None}, \emph{use\_rational=False}, \emph{fullLP=True}, \emph{format='\%s'}, \emph{infinity\_replace=None}}{}
Write an FBA-LP in polynomial H-Format file. This version has been replaced by \emph{writeModelHFormatFBA2()}
but is kept for backwards compatability.
\begin{itemize}
\item {} 
\emph{fba} a PySCeS-CBM FBA object

\item {} 
\emph{Work\_dir} {[}default=None{]} the output directory

\item {} 
\emph{use\_rational} {[}default=false{]} use rational numbers in output (requires sympy)

\item {} 
\emph{fullLP} {[}default=True{]} include the default objective function as a maximization target

\item {} 
\emph{format} {[}default='\%s'{]} the number format string

\item {} 
\emph{infinity\_replace} {[}default=None{]} if defined this is the abs(value) of +-\textless{}infinity\textgreater{}

\end{itemize}

\end{fulllineitems}

\index{writeModelHFormatFBA2() (in module cbmpy.CBWrite)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBWrite.writeModelHFormatFBA2}\pysiglinewithargsret{\code{cbmpy.CBWrite.}\bfcode{writeModelHFormatFBA2}}{\emph{fba}, \emph{fname=None}, \emph{work\_dir=None}, \emph{use\_rational=False}, \emph{fullLP=True}, \emph{format='\%s'}, \emph{infinity\_replace=None}}{}
Write an FBA-LP in polynomial H-Format file. This is an improved version of \emph{WriteModelHFormatFBA()}
which it replaces. Note that if a SymPy matrix is used as input then use\_rational is automatically enabled.
\begin{itemize}
\item {} 
\emph{fba} a PySCeS-CBM FBA object

\item {} 
\emph{fname} {[}default=None{]} the output filename, fba.getPid() if not defined

\item {} 
\emph{Work\_dir} {[}default=None{]} the output directory

\item {} 
\emph{use\_rational} {[}default=false{]} use rational numbers in output (requires sympy)

\item {} 
\emph{fullLP} {[}default=True{]} include the default objective function as a maximization target

\item {} 
\emph{format} {[}default='\%s'{]} the number format string

\item {} 
\emph{infinity\_replace} {[}default=None{]} if defined this is the abs(value) of +-\textless{}infinity\textgreater{}

\end{itemize}

\end{fulllineitems}

\index{writeModelInfoToFile() (in module cbmpy.CBWrite)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBWrite.writeModelInfoToFile}\pysiglinewithargsret{\code{cbmpy.CBWrite.}\bfcode{writeModelInfoToFile}}{\emph{fba}, \emph{fname}, \emph{Dir=None}, \emph{separator='}, \emph{`}, \emph{only\_exchange=False}, \emph{met\_type='all'}}{}
This function writes a CBModel to file
\begin{itemize}
\item {} 
\emph{fba} an instance of an PySCeSCBM model

\item {} 
\emph{fname} the output filename

\item {} 
\emph{Dir} {[}default=None{]} use directory if not None

\item {} 
\emph{separator} {[}default=','{]} the column separator

\item {} 
\emph{only\_exchange} {[}default=False{]} only output fluxes labelled as exchange reactions

\item {} 
\emph{type} {[}default='all'{]} only output certain type of species: `all','boundary' or `variable'

\end{itemize}

\end{fulllineitems}

\index{writeModelLP() (in module cbmpy.CBWrite)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBWrite.writeModelLP}\pysiglinewithargsret{\code{cbmpy.CBWrite.}\bfcode{writeModelLP}}{\emph{fba}, \emph{work\_dir=None}, \emph{fname=None}, \emph{multisymb=' `}, \emph{format='\%s'}, \emph{use\_rational=False}, \emph{constraint\_mode=None}, \emph{quiet=False}}{}
Writes an FBA object as an LP in CPLEX LP format
\begin{itemize}
\item {} 
\emph{fba} an instantiated FBAmodel instance

\item {} 
\emph{work\_dir} directory designated for output

\item {} 
\emph{fname} the file name {[}default=fba.getPid(){]}

\item {} 
\emph{multisymb} the multiplication symbol (default: \textless{}space\textgreater{})

\item {} 
\emph{format} the number format of the output

\item {} 
\emph{use\_rational} output rational numbers {[}default=False{]}

\item {} 
\emph{quiet} {[}default=False{]} supress information messages

\end{itemize}

\end{fulllineitems}

\index{writeModelLPOld() (in module cbmpy.CBWrite)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBWrite.writeModelLPOld}\pysiglinewithargsret{\code{cbmpy.CBWrite.}\bfcode{writeModelLPOld}}{\emph{fba}, \emph{work\_dir=None}, \emph{multisymb=' `}, \emph{lpt=True}, \emph{constraint\_mode='strict'}, \emph{use\_rational=False}, \emph{format='\%s'}}{}
Writes a fba as an LP/LPT
\begin{itemize}
\item {} 
\emph{fba} an instantiated FBAmodel instance

\item {} 
\emph{work\_dir} directory designated for output

\item {} 
\emph{multisymb} the multiplication symbol (default: \textless{}space\textgreater{})

\item {} 
\emph{lpt} the file format (default: True for lpt) or False for lp

\end{itemize}

\end{fulllineitems}

\index{writeModelRaw() (in module cbmpy.CBWrite)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBWrite.writeModelRaw}\pysiglinewithargsret{\code{cbmpy.CBWrite.}\bfcode{writeModelRaw}}{\emph{fba}, \emph{work\_dir=None}}{}
Writes a fba (actually just dumps it) to a text file.
\begin{itemize}
\item {} 
\emph{fba} an instantiated FBAmodel instance

\item {} 
\emph{work\_dir} directory designated for output

\end{itemize}

\end{fulllineitems}

\index{writeModelToCOMBINEarchive() (in module cbmpy.CBWrite)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBWrite.writeModelToCOMBINEarchive}\pysiglinewithargsret{\code{cbmpy.CBWrite.}\bfcode{writeModelToCOMBINEarchive}}{\emph{mod}, \emph{fname=None}, \emph{directory=None}, \emph{sbmlname=None}, \emph{withExcel=True}, \emph{vc\_given='CBMPy'}, \emph{vc\_family='Software'}, \emph{vc\_email='None'}, \emph{vc\_org='cbmpy.sourceforge.net'}, \emph{add\_cbmpy\_annot=True}, \emph{add\_cobra\_annot=True}}{}
Write a model in SBML and Excel format to a COMBINE archive using the following information:
\begin{itemize}
\item {} 
\emph{mod} a model object

\item {} 
\emph{fname} the output base filename, archive will be \textless{}fname\textgreater{}.zip

\item {} 
\emph{directory} {[}default=None{]} created the combine archive `directory'

\item {} 
\emph{sbmlname} {[}default='None'{]} If \emph{sbmlname} is defined then SBML file is \textless{}sbmlname\textgreater{}.xml otherwise sbml will be \textless{}fname\textgreater{}.xml.

\item {} 
\emph{withExcel} {[}default=True{]} include a human readable Excel spreadsheet version of the model

\item {} 
\emph{vc\_given} {[}default='CBMPy'{]} first name

\item {} 
\emph{vc\_family} {[}default='Software'{]} family name

\item {} 
\emph{vc\_email} {[}default='None'{]} email

\item {} 
\emph{vc\_org} {[}default='None'{]} organisation

\item {} 
\emph{add\_cbmpy\_annot} {[}default=True{]} add CBMPy KeyValueData annotation. Replaces \textless{}notes\textgreater{}

\item {} 
\emph{add\_cobra\_annot} {[}default=True{]} add COBRA \textless{}notes\textgreater{} annotation

\end{itemize}

\end{fulllineitems}

\index{writeModelToExcel97() (in module cbmpy.CBWrite)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBWrite.writeModelToExcel97}\pysiglinewithargsret{\code{cbmpy.CBWrite.}\bfcode{writeModelToExcel97}}{\emph{fba}, \emph{filename}, \emph{roundoff=6}}{}
Exports the model as an Excel 97 spreadsheet
\begin{itemize}
\item {} 
\emph{fba} a CBMPy model instance

\item {} 
\emph{filename} the filename of the workbook

\item {} 
\emph{roundoff} {[}default=6{]} the number of digits to round off to

\end{itemize}

\end{fulllineitems}

\index{writeOptimalSolution() (in module cbmpy.CBWrite)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBWrite.writeOptimalSolution}\pysiglinewithargsret{\code{cbmpy.CBWrite.}\bfcode{writeOptimalSolution}}{\emph{fba}, \emph{fname}, \emph{Dir=None}, \emph{separator='}, \emph{`}, \emph{only\_exchange=False}}{}
This function writes the optimal solution to file
\begin{itemize}
\item {} 
\emph{fba} an instance of an PySCeSCBM model

\item {} 
\emph{fname} the output filename

\item {} 
\emph{Dir} {[}default=None{]} use current directory if not None

\item {} 
\emph{separator} {[}default=','{]} the column separator

\item {} 
\emph{only\_exchange} {[}default=False{]} only output fluxes labelled as exchange reactions

\end{itemize}

\end{fulllineitems}

\index{writeProteinCostToCSV() (in module cbmpy.CBWrite)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBWrite.writeProteinCostToCSV}\pysiglinewithargsret{\code{cbmpy.CBWrite.}\bfcode{writeProteinCostToCSV}}{\emph{fba}, \emph{fname}}{}
Writes the protein costs `CBM\_PEPTIDE\_COST' annotation toa csv file.
\begin{itemize}
\item {} 
\emph{fba} an instantiated FBA object

\item {} 
\emph{fname} the exported file name

\end{itemize}

\end{fulllineitems}

\index{writeReactionInfoToFile() (in module cbmpy.CBWrite)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBWrite.writeReactionInfoToFile}\pysiglinewithargsret{\code{cbmpy.CBWrite.}\bfcode{writeReactionInfoToFile}}{\emph{fba}, \emph{fname}, \emph{Dir=None}, \emph{separator='}, \emph{`}, \emph{only\_exchange=False}}{}
This function writes a CBModel to file
\begin{itemize}
\item {} 
\emph{fba} an instance of an PySCeSCBM model

\item {} 
\emph{fname} the output filename

\item {} 
\emph{Dir} {[}default=None{]} use directory if not None

\item {} 
\emph{separator} {[}default=','{]} the column separator

\item {} 
\emph{only\_exchange} {[}default=False{]} only output fluxes labelled as exchange reactions

\end{itemize}

\end{fulllineitems}

\index{writeSBML2FBA() (in module cbmpy.CBWrite)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBWrite.writeSBML2FBA}\pysiglinewithargsret{\code{cbmpy.CBWrite.}\bfcode{writeSBML2FBA}}{\emph{fba}, \emph{fname}, \emph{directory=None}, \emph{sbml\_level\_version=None}}{}
Takes an FBA model object and writes it to file as SBML L2 with FBA annotations.
Note if you want to write BiGG/FAME style annotations then you must use \emph{sbml\_level\_version=(2,1)}
\begin{itemize}
\item {} 
\emph{fba} an fba model object

\item {} 
\emph{fname} the model will be written as XML to \emph{fname}

\item {} 
\emph{sbml\_level\_version} {[}default=None{]} a tuple containing the SBML level and version e.g. (2,1)

\end{itemize}

This is a utility wrapper for the function \emph{CBXML.sbml\_writeSBML2FBA}

\end{fulllineitems}

\index{writeSBML3FBC() (in module cbmpy.CBWrite)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBWrite.writeSBML3FBC}\pysiglinewithargsret{\code{cbmpy.CBWrite.}\bfcode{writeSBML3FBC}}{\emph{fba}, \emph{fname}, \emph{directory=None}, \emph{sbml\_level\_version=(3}, \emph{1)}, \emph{autofix=True}, \emph{gpr\_from\_annot=False}, \emph{add\_groups=False}, \emph{add\_cbmpy\_annot=True}, \emph{add\_cobra\_annot=False}, \emph{xoptions=\{`fbc\_version': 1}, \emph{`validate': False}, \emph{`compress\_bounds': True\}}}{}
Takes an FBA model object and writes it to file as SBML L3 FBC:
\begin{itemize}
\item {} 
\emph{fba} an fba model object

\item {} 
\emph{fname} the model will be written as XML to \emph{fname}

\item {} 
\emph{directory} {[}default=None{]} if defined it is prepended to fname

\item {} 
\emph{sbml\_level\_version} {[}default=(3,1){]} a tuple containing the SBML level and version e.g. (3,1)

\item {} 
\emph{autofix} convert \textless{}\textgreater{} to \textless{}=\textgreater{}=

\item {} 
\emph{gpr\_from\_annot} {[}default=True{]} if enabled will attempt to add the gene protein associations from the annotations
if no gene protein association objects exist

\item {} 
\emph{add\_cbmpy\_annot} {[}default=True{]} add CBMPy KeyValueData annotation. Replaces \textless{}notes\textgreater{}

\item {} 
\emph{add\_cobra\_annot} {[}default=True{]} add COBRA \textless{}notes\textgreater{} annotation

\item {} 
\emph{xoptions} extended options
\begin{itemize}
\item {} 
\emph{fbc\_version} {[}default=1{]} write SBML3FBC using version 1 (2013) or version 2 (2015)

\item {} 
\emph{validate} {[}default=False{]} validate the output SBML file

\item {} 
\emph{compress\_bounds} {[}default=False{]} try compress output flux bound parameters

\end{itemize}

\end{itemize}

\end{fulllineitems}

\index{writeSBML3FBCV2() (in module cbmpy.CBWrite)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBWrite.writeSBML3FBCV2}\pysiglinewithargsret{\code{cbmpy.CBWrite.}\bfcode{writeSBML3FBCV2}}{\emph{fba}, \emph{fname}, \emph{directory=None}, \emph{gpr\_from\_annot=False}, \emph{add\_groups=False}, \emph{add\_cbmpy\_annot=True}, \emph{add\_cobra\_annot=False}, \emph{validate=False}, \emph{compress\_bounds=True}}{}
Takes an FBA model object and writes it to file as SBML L3 FBC:
\begin{itemize}
\item {} 
\emph{fba} an fba model object

\item {} 
\emph{fname} the model will be written as XML to \emph{fname}

\item {} 
\emph{directory} {[}default=None{]} if defined it is prepended to fname

\item {} 
\emph{gpr\_from\_annot} {[}default=False{]} if enabled will attempt to add the gene protein associations from the annotations

\item {} 
\emph{add\_groups} {[}default=False{]} add SBML3 groups (if supported by libSBML)

\item {} 
\emph{add\_cbmpy\_annot} {[}default=True{]} add CBMPy KeyValueData annotation. Replaces \textless{}notes\textgreater{}

\item {} 
\emph{add\_cobra\_annot} {[}default=False{]} add COBRA \textless{}notes\textgreater{} annotation

\item {} 
\emph{validate} {[}default=False{]} validate the output SBML file

\item {} 
\emph{compress\_bounds} {[}default=True{]} try compress output flux bound parameters

\end{itemize}

\end{fulllineitems}

\index{writeSensitivitiesToCSV() (in module cbmpy.CBWrite)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBWrite.writeSensitivitiesToCSV}\pysiglinewithargsret{\code{cbmpy.CBWrite.}\bfcode{writeSensitivitiesToCSV}}{\emph{sensitivities}, \emph{fname}}{}
Write out a sensitivity report using the objective sensitivities and
bound sensitivity dictionaries created by e.g. cplx\_getSensitivities().
\begin{quote}
\begin{itemize}
\item {} 
\emph{sensitivity} tuple containing

\end{itemize}
\begin{itemize}
\item {} 
\emph{obj\_sens} dictionary of objective coefficient sensitivities (per flux)

\item {} 
\emph{rhs\_sens} dictionary of constraint rhs sensitivities (per constraint)

\item {} 
\emph{bound\_sens} dictionary of bound sensitivities (per flux)

\end{itemize}
\begin{itemize}
\item {} 
\emph{fname} output filename e.g. fname.csv

\end{itemize}
\end{quote}

\end{fulllineitems}

\index{writeSolutions() (in module cbmpy.CBWrite)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBWrite.writeSolutions}\pysiglinewithargsret{\code{cbmpy.CBWrite.}\bfcode{writeSolutions}}{\emph{fname}, \emph{sols={[}{]}}, \emph{sep='}, \emph{`}, \emph{extra\_output=None}, \emph{fba=None}}{}
Write 2 or more solutions where a solution is a dictionary of flux:value pairs:
\begin{itemize}
\item {} 
\emph{fname} the export filename

\item {} 
\emph{sols} a list of dictionaries containing flux:value pairs (e.g. output by cmod.getReactionValues())

\item {} 
\emph{sep} {[}default=','{]} the column separator

\item {} 
\emph{extra\_output} {[}default=None{]} add detailed information to output e.g. reaction names by giving a CBModel object as an argument to \emph{extra\_output}.

\item {} 
\emph{fba} an fba model that canbe used for extra\_output

\end{itemize}

\end{fulllineitems}

\index{writeSpeciesInfoToFile() (in module cbmpy.CBWrite)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBWrite.writeSpeciesInfoToFile}\pysiglinewithargsret{\code{cbmpy.CBWrite.}\bfcode{writeSpeciesInfoToFile}}{\emph{fba}, \emph{fname}, \emph{Dir=None}, \emph{separator='}, \emph{`}, \emph{met\_type='all'}}{}
This function writes a CBModel to file
\begin{itemize}
\item {} 
\emph{fba} an instance of an PySCeSCBM model

\item {} 
\emph{fname} the output filename

\item {} 
\emph{Dir} {[}default=None{]} use directory if not None

\item {} 
\emph{separator} {[}default=','{]} the column separator

\item {} 
\emph{met\_type} {[}default='all'{]} only output certain type of species: `all','boundary' or `variable'

\end{itemize}

\end{fulllineitems}

\index{writeStoichiometricMatrix() (in module cbmpy.CBWrite)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBWrite.writeStoichiometricMatrix}\pysiglinewithargsret{\code{cbmpy.CBWrite.}\bfcode{writeStoichiometricMatrix}}{\emph{fba}, \emph{fname=None}, \emph{work\_dir=None}, \emph{use\_rational=False}, \emph{fullLP=True}, \emph{format='\%s'}, \emph{infinity\_replace=None}}{}
Write an FBA-LP in polynomial H-Format file. This is an improved version of \emph{WriteModelHFormatFBA()}
which it replaces but is kept for backwards compatability.
\begin{itemize}
\item {} 
\emph{fba} a PySCeS-CBM FBA object

\item {} 
\emph{fname} {[}default=None{]} the output filename, fba.getPid() if not defined

\item {} 
\emph{Work\_dir} {[}default=None{]} the output directory

\item {} 
\emph{use\_rational} {[}default=false{]} use rational numbers in output (requires sympy)

\item {} 
\emph{fullLP} {[}default=True{]} include the default objective function as a maximization target

\item {} 
\emph{format} {[}default='\%s'{]} the number format string

\item {} 
\emph{infinity\_replace} {[}default=None{]} if defined this is the abs(value) of +-\textless{}infinity\textgreater{}

\end{itemize}

\end{fulllineitems}

\phantomsection\label{modules_doc:module-cbmpy.CBWx}\index{cbmpy.CBWx (module)}

\section{CBMPy: CBWx module}
\label{modules_doc:cbmpy-cbwx-module}
PySCeS Constraint Based Modelling (\href{http://cbmpy.sourceforge.net}{http://cbmpy.sourceforge.net})
Copyright (C) 2009-2015 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see \textless{}\href{http://www.gnu.org/licenses/}{http://www.gnu.org/licenses/}\textgreater{}

Author: Brett G. Olivier
Contact email: \href{mailto:bgoli@users.sourceforge.net}{bgoli@users.sourceforge.net}
Last edit: \$Author: bgoli \$ (\$Id: CBWx.py 378 2015-09-14 16:32:38Z bgoli \$)
\index{HtmlWindowMod (class in cbmpy.CBWx)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBWx.HtmlWindowMod}\pysiglinewithargsret{\strong{class }\code{cbmpy.CBWx.}\bfcode{HtmlWindowMod}}{\emph{*args}, \emph{**kwargs}}{}
Overrides `OnLinkClicked' to open links in external browser

\end{fulllineitems}

\index{circlePoints() (in module cbmpy.CBWx)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBWx.circlePoints}\pysiglinewithargsret{\code{cbmpy.CBWx.}\bfcode{circlePoints}}{\emph{totalPoints=4}, \emph{startAngle=0}, \emph{arc=360}, \emph{circleradius=1}, \emph{centerxy=(0}, \emph{0)}, \emph{direction='forward'}, \emph{evenDistribution=True}}{}
Returns a list of points evenly spread around a circle:
\begin{itemize}
\item {} 
\emph{totalPoints} how many points

\item {} 
\emph{startAngle} where to start

\item {} 
\emph{arc} how far to go

\item {} 
\emph{circleradius} radius

\item {} 
\emph{centerxy} origin

\item {} 
\emph{direction} `forward' or `backward'

\item {} 
\emph{evenDistribution} True/False

\end{itemize}

This code has been adapted from the Flash example that can be found here:
\href{http://www.lextalkington.com/blog/2009/12/generate-points-around-a-circles-circumference/}{http://www.lextalkington.com/blog/2009/12/generate-points-around-a-circles-circumference/}

\end{fulllineitems}

\phantomsection\label{modules_doc:module-cbmpy.CBXML}\index{cbmpy.CBXML (module)}

\section{CBMPy: CBXML module}
\label{modules_doc:cbmpy-cbxml-module}
PySCeS Constraint Based Modelling (\href{http://cbmpy.sourceforge.net}{http://cbmpy.sourceforge.net})
Copyright (C) 2009-2015 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see \textless{}\href{http://www.gnu.org/licenses/}{http://www.gnu.org/licenses/}\textgreater{}

Author: Brett G. Olivier
Contact email: \href{mailto:bgoli@users.sourceforge.net}{bgoli@users.sourceforge.net}
Last edit: \$Author: bgoli \$ (\$Id: CBXML.py 416 2016-02-23 16:12:23Z bgoli \$)
\index{MLStripper (class in cbmpy.CBXML)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBXML.MLStripper}\pysigline{\strong{class }\code{cbmpy.CBXML.}\bfcode{MLStripper}}
Class for stripping a string of HTML/XML used from:
\href{http://stackoverflow.com/questions/753052/strip-html-from-strings-in-python}{http://stackoverflow.com/questions/753052/strip-html-from-strings-in-python}

\end{fulllineitems}

\index{sbml\_convertCOBRASBMLtoFBC() (in module cbmpy.CBXML)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBXML.sbml_convertCOBRASBMLtoFBC}\pysiglinewithargsret{\code{cbmpy.CBXML.}\bfcode{sbml\_convertCOBRASBMLtoFBC}}{\emph{fname}, \emph{outname=None}, \emph{work\_dir=None}, \emph{output\_dir=None}}{}
Read in a COBRA SBML Level 2 file and return the name of the created SBML Level 3 with FBC
file that is created in the output directory
\begin{itemize}
\item {} 
\emph{fname} is the filename

\item {} 
\emph{outname} the name of the output file. If not specified then \textless{}filename\textgreater{}.l3fbc.xml is used as default

\item {} 
\emph{work\_dir} {[}default=None{]} is the working directory

\item {} 
\emph{output\_dir} {[}default=None{]} is the output directory (default is work\_dir)

\end{itemize}

This method is based on code from libSBML (\href{http://sbml.org}{http://sbml.org}) in the file ``convertCobra.py''
written by Frank T. Bergmann.

\end{fulllineitems}

\index{sbml\_convertSBML3FBCToCOBRA() (in module cbmpy.CBXML)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBXML.sbml_convertSBML3FBCToCOBRA}\pysiglinewithargsret{\code{cbmpy.CBXML.}\bfcode{sbml\_convertSBML3FBCToCOBRA}}{\emph{fname}, \emph{outname=None}, \emph{work\_dir=None}, \emph{output\_dir=None}}{}
Read in a SBML Level 3 file and return the name of the created COBRA
file that is created in the output directory
\begin{itemize}
\item {} 
\emph{fname} is the filename

\item {} 
\emph{outname} the name of the output file. If not specified then \textless{}filename\textgreater{}.cobra.xml is used as default

\item {} 
\emph{work\_dir} {[}default=None{]} is the working directory

\item {} 
\emph{output\_dir} {[}default=None{]} is the output directory (default is work\_dir)

\end{itemize}

This method is based on code from libSBML (\href{http://sbml.org}{http://sbml.org}) in the file ``convertFbcToCobra.py''
written by Frank T. Bergmann.

\end{fulllineitems}

\index{sbml\_createAssociationFromAST() (in module cbmpy.CBXML)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBXML.sbml_createAssociationFromAST}\pysiglinewithargsret{\code{cbmpy.CBXML.}\bfcode{sbml\_createAssociationFromAST}}{\emph{node}, \emph{out}}{}
Converts a GPR string `((g1 and g2) or g3)' to an association via a Python AST.
In future I will get rid of all the string elements and work only with associations
and AST's.
\begin{itemize}
\item {} 
\emph{node} a Python AST note (e.g. body)

\item {} 
\emph{out} a new shiny FBC V2 GeneProductAssociation

\end{itemize}

\end{fulllineitems}

\index{sbml\_createModelL2() (in module cbmpy.CBXML)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBXML.sbml_createModelL2}\pysiglinewithargsret{\code{cbmpy.CBXML.}\bfcode{sbml\_createModelL2}}{\emph{fba}, \emph{level=2}, \emph{version=1}}{}
Create an SBML model and document:
\begin{itemize}
\item {} 
\emph{fba} a PySCeSCBM model instance

\item {} 
\emph{level} always 2

\item {} 
\emph{version} always 1

\end{itemize}

and returns:
\begin{itemize}
\item {} 
\emph{model} an SBML model

\end{itemize}

\end{fulllineitems}

\index{sbml\_exportSBML2FBAModel() (in module cbmpy.CBXML)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBXML.sbml_exportSBML2FBAModel}\pysiglinewithargsret{\code{cbmpy.CBXML.}\bfcode{sbml\_exportSBML2FBAModel}}{\emph{document}, \emph{filename}, \emph{directory=None}, \emph{return\_doc=False}, \emph{remove\_note\_body=False}}{}
Writes an SBML model object to file. Note this is an internal SBML method use \emph{sbml\_writeSBML2FBA()} to write an FBA model:
\begin{itemize}
\item {} 
\emph{model} a libSBML model instance

\item {} 
\emph{filename} the output filename

\item {} 
\emph{directory} {[}default=None{]} by default use filename otherwise join, \textless{}dir\textgreater{}\textless{}filename\textgreater{}

\item {} 
\emph{return\_doc} {[}default=False{]} return the SBML document used to write the XML

\end{itemize}

\end{fulllineitems}

\index{sbml\_getCVterms() (in module cbmpy.CBXML)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBXML.sbml_getCVterms}\pysiglinewithargsret{\code{cbmpy.CBXML.}\bfcode{sbml\_getCVterms}}{\emph{sb}, \emph{model=False}}{}
Get the MIRIAM compliant CV terms and return a MIRIAMAnnotation or None
\begin{itemize}
\item {} 
\emph{sb} a libSBML SBase derived object

\item {} 
\emph{model} is this a BQmodel term

\end{itemize}

\end{fulllineitems}

\index{sbml\_getGeneRefs() (in module cbmpy.CBXML)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBXML.sbml_getGeneRefs}\pysiglinewithargsret{\code{cbmpy.CBXML.}\bfcode{sbml\_getGeneRefs}}{\emph{association}, \emph{out}}{}
Walk through a gene association and extract GeneRefs inspired by Frank

\end{fulllineitems}

\index{sbml\_readCOBRANote() (in module cbmpy.CBXML)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBXML.sbml_readCOBRANote}\pysiglinewithargsret{\code{cbmpy.CBXML.}\bfcode{sbml\_readCOBRANote}}{\emph{s}}{}
Parses a COBRA style note from a XML string
\begin{itemize}
\item {} 
\emph{s} an XML string

\end{itemize}

\end{fulllineitems}

\index{sbml\_readCOBRASBML() (in module cbmpy.CBXML)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBXML.sbml_readCOBRASBML}\pysiglinewithargsret{\code{cbmpy.CBXML.}\bfcode{sbml\_readCOBRASBML}}{\emph{fname}, \emph{work\_dir=None}, \emph{return\_sbml\_model=False}, \emph{delete\_intermediate=False}, \emph{fake\_boundary\_species\_search=False}, \emph{output\_dir=None}, \emph{speciesAnnotationFix=True}}{}
Read in a COBRA format SBML Level 2 file with FBA annotation where and return either a CBM model object
or a (cbm\_mod, sbml\_mod) pair if return\_sbml\_model=True
\begin{itemize}
\item {} 
\emph{fname} is the filename

\item {} 
\emph{work\_dir} is the working directory

\item {} 
\emph{return\_sbml\_model} {[}default=False{]} return a a (cbm\_mod, sbml\_mod) pair

\item {} 
\emph{delete\_intermediate} {[}default=False{]} delete the intermediate SBML Level 3 FBC file

\item {} 
\emph{fake\_boundary\_species\_search} {[}default=False{]} after looking for the boundary\_condition of a species search for overloaded id's \textless{}id\textgreater{}\_b

\item {} 
\emph{output\_dir} {[}default=None{]} the directory to output the intermediate SBML L3 files (if generated) default to input directory

\item {} 
\emph{speciesAnnotationFix} {[}default=True{]}

\end{itemize}

\end{fulllineitems}

\index{sbml\_readKeyValueDataAnnotation() (in module cbmpy.CBXML)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBXML.sbml_readKeyValueDataAnnotation}\pysiglinewithargsret{\code{cbmpy.CBXML.}\bfcode{sbml\_readKeyValueDataAnnotation}}{\emph{annotations}}{}
Reads KeyValueData annotation (\href{http://pysces.sourceforge.net/KeyValueData}{http://pysces.sourceforge.net/KeyValueData}) and returns a dictionary of key:value pairs

\end{fulllineitems}

\index{sbml\_readSBML2FBA() (in module cbmpy.CBXML)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBXML.sbml_readSBML2FBA}\pysiglinewithargsret{\code{cbmpy.CBXML.}\bfcode{sbml\_readSBML2FBA}}{\emph{fname}, \emph{work\_dir=None}, \emph{return\_sbml\_model=False}, \emph{fake\_boundary\_species\_search=False}}{}
Read in an SBML Level 2 file with FBA annotation where and return either a CBM model object
or a (cbm\_mod, sbml\_mod) pair if return\_sbml\_model=True
\begin{itemize}
\item {} 
\emph{fname} is the filename

\item {} 
\emph{work\_dir} is the working directory (only used if not None)

\item {} 
\emph{return\_sbml\_model} {[}default=False{]} return a a (cbm\_mod, sbml\_mod) pair

\item {} 
\emph{fake\_boundary\_species\_search} {[}default=False{]} after looking for the boundary\_condition of a species search for overloaded id's \textless{}id\textgreater{}\_b

\end{itemize}

\end{fulllineitems}

\index{sbml\_readSBML3FBC() (in module cbmpy.CBXML)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBXML.sbml_readSBML3FBC}\pysiglinewithargsret{\code{cbmpy.CBXML.}\bfcode{sbml\_readSBML3FBC}}{\emph{fname}, \emph{work\_dir=None}, \emph{return\_sbml\_model=False}, \emph{xoptions=\{\}}}{}
Read in an SBML Level 3 file with FBC annotation where and return either a CBM model object
or a (cbm\_mod, sbml\_mod) pair if return\_sbml\_model=True
\begin{itemize}
\item {} 
\emph{fname} is the filename

\item {} 
\emph{work\_dir} is the working directory

\item {} 
\emph{return\_sbml\_model} {[}default=False{]} return a a (cbm\_mod, sbml\_mod) pair

\item {} 
\emph{xoptions} special load options enable with option = True
- \emph{nogenes} do not load/process genes
- \emph{noannot} do not load/process any annotations
- \emph{validate} validate model and display errors and warnings before loading

\end{itemize}

\end{fulllineitems}

\index{sbml\_setCVterms() (in module cbmpy.CBXML)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBXML.sbml_setCVterms}\pysiglinewithargsret{\code{cbmpy.CBXML.}\bfcode{sbml\_setCVterms}}{\emph{sb}, \emph{uridict}, \emph{model=False}}{}
Add MIRIAM compliant CV terms to a sbml object from a CBM object
\begin{itemize}
\item {} 
\emph{sb} a libSBML SBase derived object

\item {} 
\emph{uridict} a dictionary of uri's as produced by getAllMIRIAMUris()

\item {} 
\emph{model} is this a BQmodel term {[}deprecated attribute, ignored and autodetected{]}

\end{itemize}

\end{fulllineitems}

\index{sbml\_setCompartmentsL3() (in module cbmpy.CBXML)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBXML.sbml_setCompartmentsL3}\pysiglinewithargsret{\code{cbmpy.CBXML.}\bfcode{sbml\_setCompartmentsL3}}{\emph{model}, \emph{fba}}{}
Sets the model compartments.
\begin{itemize}
\item {} 
\emph{model} a libSBML model instance

\item {} 
\emph{fba} a PySCeSCBM model instance

\end{itemize}

\end{fulllineitems}

\index{sbml\_setDescription() (in module cbmpy.CBXML)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBXML.sbml_setDescription}\pysiglinewithargsret{\code{cbmpy.CBXML.}\bfcode{sbml\_setDescription}}{\emph{model}, \emph{fba}}{}
Sets the model description as a \textless{}note\textgreater{} containing \emph{txt} in an HTML paragraph on the model object.
\begin{itemize}
\item {} 
\emph{model} a libSBML model instance

\item {} 
\emph{fba} a PySCeSCBM model instance

\end{itemize}

\end{fulllineitems}

\index{sbml\_setGroupsL3() (in module cbmpy.CBXML)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBXML.sbml_setGroupsL3}\pysiglinewithargsret{\code{cbmpy.CBXML.}\bfcode{sbml\_setGroupsL3}}{\emph{cs}, \emph{fba}}{}
add groups to the SBML model
\begin{itemize}
\item {} 
\emph{cs} a CBMLtoSBML instance

\item {} 
\emph{fba} a CBMPy model instance

\end{itemize}

\end{fulllineitems}

\index{sbml\_setNotes3() (in module cbmpy.CBXML)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBXML.sbml_setNotes3}\pysiglinewithargsret{\code{cbmpy.CBXML.}\bfcode{sbml\_setNotes3}}{\emph{obj}, \emph{s}}{}
Formats the CBMPy notes as an SBML note and adds it to the SBMl object
\begin{itemize}
\item {} 
\emph{obj} an SBML object

\item {} 
\emph{s} a string that should be added as a note

\end{itemize}

\end{fulllineitems}

\index{sbml\_setReactionsL2() (in module cbmpy.CBXML)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBXML.sbml_setReactionsL2}\pysiglinewithargsret{\code{cbmpy.CBXML.}\bfcode{sbml\_setReactionsL2}}{\emph{model}, \emph{fba}, \emph{return\_dict=False}}{}
Add the FBA instance reactions to the SBML model
\begin{itemize}
\item {} 
\emph{model} an SBML model instance

\item {} 
\emph{fba} a PySCeSCBM model instance

\item {} 
\emph{return\_dict} {[}default=False{]} if True do not add reactions to SBML document instead return a dictionary description of the reactions

\end{itemize}

\end{fulllineitems}

\index{sbml\_setReactionsL3Fbc() (in module cbmpy.CBXML)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBXML.sbml_setReactionsL3Fbc}\pysiglinewithargsret{\code{cbmpy.CBXML.}\bfcode{sbml\_setReactionsL3Fbc}}{\emph{fbcmod}, \emph{fba}, \emph{return\_dict=False}, \emph{add\_cobra\_anno=False}, \emph{add\_cbmpy\_anno=True}, \emph{fbc\_version=1}}{}
Add the FBA instance reactions to the SBML model
\begin{itemize}
\item {} 
\emph{fbcmod} a CBM2SBML instance

\item {} 
\emph{fba} a PySCeSCBM model instance

\item {} 
\emph{return\_dict} {[}default=False{]} if True do not add reactions to SBML document instead return a dictionary description of the reactions

\item {} 
\emph{add\_cbmpy\_anno} {[}default=True{]} add CBMPy KeyValueData annotation. Replaces \textless{}notes\textgreater{}

\item {} 
\emph{add\_cobra\_anno} {[}default=False{]} add COBRA \textless{}notes\textgreater{} annotation

\item {} 
\emph{fbc\_version} {[}default=1{]} writes either FBC v1 (2013) or v2 (2015)

\end{itemize}

\end{fulllineitems}

\index{sbml\_setSpeciesL2() (in module cbmpy.CBXML)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBXML.sbml_setSpeciesL2}\pysiglinewithargsret{\code{cbmpy.CBXML.}\bfcode{sbml\_setSpeciesL2}}{\emph{model}, \emph{fba}, \emph{return\_dicts=False}}{}
Add the species definitions to the SBML object:
\begin{itemize}
\item {} 
\emph{model} {[}default='`{]} a libSBML model instance or can be None if \emph{return\_dicts} == True

\item {} 
\emph{fba} a PySCeSCBM model instance

\item {} 
\emph{return\_dicts} {[}default=False{]} only returns the compartment and species dictionaries without updated the SBML

\end{itemize}

returns:
\begin{itemize}
\item {} 
\emph{compartments} a dictionary of compartments (except when give \emph{return\_dicts} argument)

\end{itemize}

\end{fulllineitems}

\index{sbml\_setSpeciesL3() (in module cbmpy.CBXML)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBXML.sbml_setSpeciesL3}\pysiglinewithargsret{\code{cbmpy.CBXML.}\bfcode{sbml\_setSpeciesL3}}{\emph{model}, \emph{fba}, \emph{return\_dicts=False}, \emph{add\_cobra\_anno=False}, \emph{add\_cbmpy\_anno=True}, \emph{substance\_units=True}}{}
Add the species definitions to the SBML object:
\begin{itemize}
\item {} 
\emph{model} and SBML model instance or can be None if \emph{return\_dicts} == True

\item {} 
\emph{fba} a PySCeSCBM model instance

\item {} 
\emph{return\_dicts} {[}default=False{]} only returns the compartment and species dictionaries without updating the SBML

\item {} 
\emph{add\_cbmpy\_anno} {[}default=True{]} add CBMPy KeyValueData annotation. Replaces \textless{}notes\textgreater{}

\item {} 
\emph{add\_cobra\_anno} {[}default=False{]} add COBRA \textless{}notes\textgreater{} annotation

\item {} 
\emph{substance\_units} {[}default=True{]} defines the species in amounts rather than concentrations (necessary for default mmol/gdw.h)

\end{itemize}

returns:
\begin{itemize}
\item {} 
\emph{compartments} a dictionary of compartments (except when given \emph{return\_dicts} argument)

\end{itemize}

\end{fulllineitems}

\index{sbml\_setUnits() (in module cbmpy.CBXML)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBXML.sbml_setUnits}\pysiglinewithargsret{\code{cbmpy.CBXML.}\bfcode{sbml\_setUnits}}{\emph{model}, \emph{units=None}, \emph{give\_default=False}, \emph{L3=True}}{}
Adds units to the model:
\begin{itemize}
\item {} 
\emph{model} a libSBML model instance

\item {} 
\emph{units} {[}default=None{]} a dictionary of units, if None default units are used

\item {} 
\emph{give\_default} {[}default=False{]} if true method returns the default unit dictionary

\item {} 
\emph{L3} {[}default=True{]} use the L3 defaults

\end{itemize}

\end{fulllineitems}

\index{sbml\_setValidationOptions() (in module cbmpy.CBXML)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBXML.sbml_setValidationOptions}\pysiglinewithargsret{\code{cbmpy.CBXML.}\bfcode{sbml\_setValidationOptions}}{\emph{D}, \emph{level}}{}
set the validation level of an SBML document
\begin{quote}
\begin{itemize}
\item {} 
\emph{D} an SBML document

\item {} 
\emph{level} the level of consistency check can be either one of:

\end{itemize}
\begin{itemize}
\item {} 
`normal' basic id checking only

\item {} 
`full' all checks enabled

\end{itemize}
\end{quote}

\end{fulllineitems}

\index{sbml\_validateDocument() (in module cbmpy.CBXML)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBXML.sbml_validateDocument}\pysiglinewithargsret{\code{cbmpy.CBXML.}\bfcode{sbml\_validateDocument}}{\emph{D}, \emph{fullmsg=False}}{}
Validates and SBML document returns three dictionaries, errors, warnings, other and a boolean indicating an invalid document:
\begin{itemize}
\item {} 
\emph{D} and SBML document

\item {} 
\emph{fullmsg} {[}default=False{]} optionally display the full error message

\end{itemize}

\end{fulllineitems}

\index{sbml\_writeAnnotationsAsCOBRANote() (in module cbmpy.CBXML)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBXML.sbml_writeAnnotationsAsCOBRANote}\pysiglinewithargsret{\code{cbmpy.CBXML.}\bfcode{sbml\_writeAnnotationsAsCOBRANote}}{\emph{annotations}}{}
Writes the annotations dictionary as a COBRA compatible SBML \textless{}note\textgreater{}

\end{fulllineitems}

\index{sbml\_writeCOBRASBML() (in module cbmpy.CBXML)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBXML.sbml_writeCOBRASBML}\pysiglinewithargsret{\code{cbmpy.CBXML.}\bfcode{sbml\_writeCOBRASBML}}{\emph{fba}, \emph{fname}, \emph{directory=None}}{}
Takes an FBA model object and writes it to file as a COBRA compatible :
\begin{itemize}
\item {} 
\emph{fba} an fba model object

\item {} 
\emph{fname} the model will be written as XML to \emph{fname}

\item {} 
\emph{directory} {[}default=None{]} if defined it is prepended to fname

\end{itemize}

\end{fulllineitems}

\index{sbml\_writeKeyValueDataAnnotation() (in module cbmpy.CBXML)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBXML.sbml_writeKeyValueDataAnnotation}\pysiglinewithargsret{\code{cbmpy.CBXML.}\bfcode{sbml\_writeKeyValueDataAnnotation}}{\emph{annotations}}{}
Writes the key:value annotations as a KeyValueData annotation (http://pysces.sourceforge.net/KeyValueData)

\end{fulllineitems}

\index{sbml\_writeSBML2FBA() (in module cbmpy.CBXML)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBXML.sbml_writeSBML2FBA}\pysiglinewithargsret{\code{cbmpy.CBXML.}\bfcode{sbml\_writeSBML2FBA}}{\emph{fba}, \emph{fname}, \emph{directory=None}, \emph{sbml\_level\_version=None}}{}
Takes an FBA model object and writes it to file as SBML L3 FBA:
\begin{itemize}
\item {} 
\emph{fba} an fba model object

\item {} 
\emph{fname} the model will be written as XML to \emph{fname}

\item {} 
\emph{directory} {[}default=None{]} if defined it is prepended to fname

\item {} 
\emph{sbml\_level\_version} {[}default=None{]} a tuple containing the SBML level and version e.g. (2,4) (ignored)

\end{itemize}

\end{fulllineitems}

\index{sbml\_writeSBML3FBC() (in module cbmpy.CBXML)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBXML.sbml_writeSBML3FBC}\pysiglinewithargsret{\code{cbmpy.CBXML.}\bfcode{sbml\_writeSBML3FBC}}{\emph{fba}, \emph{fname}, \emph{directory=None}, \emph{sbml\_level\_version=(3}, \emph{1)}, \emph{autofix=True}, \emph{return\_fbc=False}, \emph{gpr\_from\_annot=False}, \emph{add\_groups=False}, \emph{add\_cbmpy\_annot=True}, \emph{add\_cobra\_annot=False}, \emph{xoptions=\{\}}}{}
Takes an FBA model object and writes it to file as SBML L3 FBC:
\begin{itemize}
\item {} 
\emph{fba} an fba model object

\item {} 
\emph{fname} the model will be written as XML to \emph{fname}

\item {} 
\emph{directory} {[}default=None{]} if defined it is prepended to fname

\item {} 
\emph{sbml\_level\_version} {[}default=(3,1){]} a tuple containing the SBML level and version e.g. (3,1)

\item {} 
\emph{autofix} convert \textless{}\textgreater{} to \textless{}=\textgreater{}=

\item {} 
\emph{return\_fbc} return the FBC converter instance

\item {} 
\emph{gpr\_from\_annot} {[}default=True{]} if enabled will attempt to add the gene protein associations from the annotations
if no gene protein association objects exist

\item {} 
\emph{add\_cbmpy\_annot} {[}default=True{]} add CBMPy KeyValueData annotation. Replaces \textless{}notes\textgreater{}

\item {} 
\emph{add\_cobra\_annot} {[}default=True{]} add COBRA \textless{}notes\textgreater{} annotation

\item {} 
\emph{xoptions} extended options
\begin{itemize}
\item {} 
\emph{fbc\_version} {[}default=1{]} write SBML3FBC using version 1 (2013) or version 2 (2015)

\item {} 
\emph{validate} {[}default=False{]} validate the output SBML file

\item {} 
\emph{compress\_bounds} {[}default=False{]} try compress output flux bound parameters

\end{itemize}

\end{itemize}

\end{fulllineitems}

\index{setCBSBOterm() (in module cbmpy.CBXML)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBXML.setCBSBOterm}\pysiglinewithargsret{\code{cbmpy.CBXML.}\bfcode{setCBSBOterm}}{\emph{sbo}, \emph{obj}}{}
Given an SBOterm from libSBML, add it to a CBMPy object
\begin{itemize}
\item {} 
\emph{sbo} the sbo term string

\item {} 
\emph{obj} the CBMPy Fbase derived object

\end{itemize}

\end{fulllineitems}

\index{xml\_addSBML2FBAFluxBound() (in module cbmpy.CBXML)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBXML.xml_addSBML2FBAFluxBound}\pysiglinewithargsret{\code{cbmpy.CBXML.}\bfcode{xml\_addSBML2FBAFluxBound}}{\emph{document}, \emph{rid}, \emph{operator}, \emph{value}, \emph{fbid=None}}{}
Adds an SBML3FBA flux bound to the document:
\begin{itemize}
\item {} 
\emph{document} a minidom XML document created by xml\_createSBML2FBADoc

\item {} 
\emph{rid} the reaction id

\item {} 
\emph{operator} one of {[}'greater','greaterEqual','less','lessEqual','equal','\textgreater{}','\textgreater{}=','\textless{}','\textless{}=','='{]}

\item {} 
\emph{value} a float which will be cast to a string using str(value)

\item {} 
\emph{fbid} the flux bound id, autogenerated by default

\end{itemize}

\end{fulllineitems}

\index{xml\_addSBML2FBAObjective() (in module cbmpy.CBXML)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBXML.xml_addSBML2FBAObjective}\pysiglinewithargsret{\code{cbmpy.CBXML.}\bfcode{xml\_addSBML2FBAObjective}}{\emph{document}, \emph{objective}, \emph{active=True}}{}
Adds an objective element to the documents listOfObjectives and sets the active attribute:
\begin{itemize}
\item {} 
\emph{document} a minidom XML document created by \emph{xml\_createSBML2FBADoc}

\item {} 
\emph{objective} a minidom XML objective element created with \emph{xml\_createSBML2FBAObjective}

\item {} 
\emph{active} {[}default=True{]} a boolean flag specifiying whether this objective is active

\end{itemize}

\end{fulllineitems}

\index{xml\_createListOfFluxObjectives() (in module cbmpy.CBXML)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBXML.xml_createListOfFluxObjectives}\pysiglinewithargsret{\code{cbmpy.CBXML.}\bfcode{xml\_createListOfFluxObjectives}}{\emph{document}, \emph{fluxObjectives}}{}
Create a list of fluxObjectives to add to an Objective:
\begin{itemize}
\item {} 
\emph{document} a minidom XML document created by xml\_createSBML2FBADoc

\item {} 
\emph{fluxobjs} a list of (rid, coefficient) tuples

\end{itemize}

\end{fulllineitems}

\index{xml\_createSBML2FBADoc() (in module cbmpy.CBXML)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBXML.xml_createSBML2FBADoc}\pysiglinewithargsret{\code{cbmpy.CBXML.}\bfcode{xml\_createSBML2FBADoc}}{}{}
Create a `document' to store the SBML2FBA annotation, returns:
\begin{itemize}
\item {} 
\emph{DOC} a minidom document

\end{itemize}

\end{fulllineitems}

\index{xml\_createSBML2FBAObjective() (in module cbmpy.CBXML)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBXML.xml_createSBML2FBAObjective}\pysiglinewithargsret{\code{cbmpy.CBXML.}\bfcode{xml\_createSBML2FBAObjective}}{\emph{document}, \emph{oid}, \emph{sense}, \emph{fluxObjectives}}{}
Create a list of fluxObjectives to add to an Objective:
\begin{itemize}
\item {} 
\emph{document} a minidom XML document created by xml\_createSBML2FBADoc

\item {} 
\emph{oid} the objective id

\item {} 
\emph{sense} a string containing the objective sense either: \textbf{maximize} or \textbf{minimize}

\item {} 
\emph{fluxObjectives} a list of (rid, coefficient) tuples

\end{itemize}

\end{fulllineitems}

\index{xml\_getSBML2FBAannotation() (in module cbmpy.CBXML)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBXML.xml_getSBML2FBAannotation}\pysiglinewithargsret{\code{cbmpy.CBXML.}\bfcode{xml\_getSBML2FBAannotation}}{\emph{fba}, \emph{fname=None}}{}
Takes an FBA model object and returns the SBML3FBA annotation as an XML string:
\begin{itemize}
\item {} 
\emph{fba} an fba model object

\item {} 
\emph{fname} {[}default=None{]} if supplied the XML will be written to file \emph{fname}

\end{itemize}

\end{fulllineitems}

\index{xml\_stripTags() (in module cbmpy.CBXML)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBXML.xml_stripTags}\pysiglinewithargsret{\code{cbmpy.CBXML.}\bfcode{xml\_stripTags}}{\emph{html}}{}
Strip a string of HTML/XML, returns a string
\begin{itemize}
\item {} 
\emph{html} the string containing html

\end{itemize}

\end{fulllineitems}

\index{xml\_viewSBML2FBAXML() (in module cbmpy.CBXML)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.CBXML.xml_viewSBML2FBAXML}\pysiglinewithargsret{\code{cbmpy.CBXML.}\bfcode{xml\_viewSBML2FBAXML}}{\emph{document}, \emph{fname=None}}{}
Print a minidom XML document to screen or file, arguments:
\begin{itemize}
\item {} 
\emph{document} a minidom XML document

\item {} 
\emph{fname} {[}default=None{]} by default print to screen or write to file fname

\end{itemize}

\end{fulllineitems}

\phantomsection\label{modules_doc:module-cbmpy.PyscesStoich}\index{cbmpy.PyscesStoich (module)}

\section{PyscesStoich}
\label{modules_doc:pyscesstoich}
PySCeS stoichiometric analysis classes.
\index{MathArrayFunc (class in cbmpy.PyscesStoich)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.PyscesStoich.MathArrayFunc}\pysigline{\strong{class }\code{cbmpy.PyscesStoich.}\bfcode{MathArrayFunc}}
PySCeS array functions - used by Stoich
\index{MatrixFloatFix() (cbmpy.PyscesStoich.MathArrayFunc method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.PyscesStoich.MathArrayFunc.MatrixFloatFix}\pysiglinewithargsret{\bfcode{MatrixFloatFix}}{\emph{mat}, \emph{val=1.e-15}}{}
Clean an array removing any floating point artifacts defined as being smaller than a specified value.
Processes an array inplace

Arguments:

mat: the input 2D array
val {[}default=1.e-15{]}: the threshold value (effective zero)

\end{fulllineitems}

\index{MatrixValueCompare() (cbmpy.PyscesStoich.MathArrayFunc method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.PyscesStoich.MathArrayFunc.MatrixValueCompare}\pysiglinewithargsret{\bfcode{MatrixValueCompare}}{\emph{matrix}}{}
Finds the largest/smallest abs(value) \textgreater{} 0.0 in a matrix.
Returns a tuple containing (smallest,largest) values

Arguments:

matrix: the input 2D array

\end{fulllineitems}

\index{SwapCol() (cbmpy.PyscesStoich.MathArrayFunc method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.PyscesStoich.MathArrayFunc.SwapCol}\pysiglinewithargsret{\bfcode{SwapCol}}{\emph{res\_a}, \emph{r1}, \emph{r2}}{}
Swap two columns using BLAS swap, arrays can be (or are upcast to) type double (d) or double complex (D).
Returns the colswapped array

Arguments:

res\_a: the input array
r1: the first column to be swapped
r2: the second column to be swapped

\end{fulllineitems}

\index{SwapCold() (cbmpy.PyscesStoich.MathArrayFunc method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.PyscesStoich.MathArrayFunc.SwapCold}\pysiglinewithargsret{\bfcode{SwapCold}}{\emph{res\_a}, \emph{c1}, \emph{c2}}{}
Swaps two double (d) columns in an array using BLAS DSWAP. Returns the colswapped array.

Arguments:

res\_a: input array
c1: column index 1
c2: column index 2

\end{fulllineitems}

\index{SwapColz() (cbmpy.PyscesStoich.MathArrayFunc method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.PyscesStoich.MathArrayFunc.SwapColz}\pysiglinewithargsret{\bfcode{SwapColz}}{\emph{res\_a}, \emph{c1}, \emph{c2}}{}
Swaps two double complex (D) columns in an array using BLAS ZSWAP. Returns the colswapped array.

Arguments:

res\_a: input array
c1: column index 1
c2: column index 2

\end{fulllineitems}

\index{SwapElem() (cbmpy.PyscesStoich.MathArrayFunc method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.PyscesStoich.MathArrayFunc.SwapElem}\pysiglinewithargsret{\bfcode{SwapElem}}{\emph{res\_a}, \emph{r1}, \emph{r2}}{}
Swaps two elements in a 1D vector

Arguments:

res\_a: the input vector
r1: index 1
r2: index 2

\end{fulllineitems}

\index{SwapRow() (cbmpy.PyscesStoich.MathArrayFunc method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.PyscesStoich.MathArrayFunc.SwapRow}\pysiglinewithargsret{\bfcode{SwapRow}}{\emph{res\_a}, \emph{r1}, \emph{r2}}{}
Swaps two rows using BLAS swap, arrays can be (or are upcast to) type double (d) or double complex (D).
Returns the rowswapped array.

Arguments:

res\_a: the input array
r1: the first row index to be swapped
r2:  the second row index to be swapped

\end{fulllineitems}

\index{SwapRowd() (cbmpy.PyscesStoich.MathArrayFunc method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.PyscesStoich.MathArrayFunc.SwapRowd}\pysiglinewithargsret{\bfcode{SwapRowd}}{\emph{res\_a}, \emph{c1}, \emph{c2}}{}
Swaps two double (d) rows in an array using BLAS DSWAP. Returns the rowswapped array.

Arguments:

res\_a: input array
c1: row index 1
c2: row index 2

\end{fulllineitems}

\index{SwapRowz() (cbmpy.PyscesStoich.MathArrayFunc method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.PyscesStoich.MathArrayFunc.SwapRowz}\pysiglinewithargsret{\bfcode{SwapRowz}}{\emph{res\_a}, \emph{c1}, \emph{c2}}{}
Swaps two double complex (D) rows in an array using BLAS ZSWAP. Returns the rowswapped array.

Arguments:

res\_a: input array
c1: row index 1
c2: row index 2

\end{fulllineitems}

\index{assertRank2() (cbmpy.PyscesStoich.MathArrayFunc method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.PyscesStoich.MathArrayFunc.assertRank2}\pysiglinewithargsret{\bfcode{assertRank2}}{\emph{*arrays}}{}
Check that we are using a 2D array

Arguments:

*arrays: input array(s)

\end{fulllineitems}

\index{castCopyAndTranspose() (cbmpy.PyscesStoich.MathArrayFunc method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.PyscesStoich.MathArrayFunc.castCopyAndTranspose}\pysiglinewithargsret{\bfcode{castCopyAndTranspose}}{\emph{type}, \emph{*arrays}}{}
Cast numeric arrays to required type and transpose

Arguments:

type: the required type to cast to
*arrays: the arrays to be processed

\end{fulllineitems}

\index{commonType() (cbmpy.PyscesStoich.MathArrayFunc method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.PyscesStoich.MathArrayFunc.commonType}\pysiglinewithargsret{\bfcode{commonType}}{\emph{*arrays}}{}
Numeric detect and set array precision (will be replaced with new scipy.core compatible code when ready)

Arguments:

*arrays: input arrays

\end{fulllineitems}


\end{fulllineitems}

\index{Stoich (class in cbmpy.PyscesStoich)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.PyscesStoich.Stoich}\pysiglinewithargsret{\strong{class }\code{cbmpy.PyscesStoich.}\bfcode{Stoich}}{\emph{input}}{}
PySCeS stoichiometric analysis class: initialized with a stoichiometric matrix N (input)
\index{AnalyseK() (cbmpy.PyscesStoich.Stoich method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.PyscesStoich.Stoich.AnalyseK}\pysiglinewithargsret{\bfcode{AnalyseK}}{}{}
Evaluate the stoichiometric matrix and calculate the nullspace using LU decomposition and backsubstitution .
Generates the MCA K and Ko arrays and associated row and column vectors

Arguments:
None

\end{fulllineitems}

\index{AnalyseL() (cbmpy.PyscesStoich.Stoich method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.PyscesStoich.Stoich.AnalyseL}\pysiglinewithargsret{\bfcode{AnalyseL}}{}{}
Evaluate the stoichiometric matrix and calculate the left nullspace using LU factorization and backsubstitution.
Generates the MCA L, Lo, Nr and Conservation matrix and associated row and column vectors

Arguments:
None

\end{fulllineitems}

\index{BackSubstitution() (cbmpy.PyscesStoich.Stoich method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.PyscesStoich.Stoich.BackSubstitution}\pysiglinewithargsret{\bfcode{BackSubstitution}}{\emph{res\_a}, \emph{row\_vector}, \emph{column\_vector}}{}
Jordan reduction of a scaled upper triangular matrix. The returned array is now in the form {[}I R{]} and can
be used for nullspace determination. Modified row and column tracking vetors are also returned.

Arguments:

res\_a: unitary pivot upper triangular matrix
row\_vector: row tracking vector
column\_vector: column tracking vector

\end{fulllineitems}

\index{GetUpperMatrix() (cbmpy.PyscesStoich.Stoich method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.PyscesStoich.Stoich.GetUpperMatrix}\pysiglinewithargsret{\bfcode{GetUpperMatrix}}{\emph{a}}{}
Core analysis algorithm; an input is preconditioned using PivotSort\_initial and then cycles of PLUfactorize and
PivotSort are run until the factorization is completed. During this process the matrix is reordered by
column swaps which emulates a full pivoting LU factorization. Returns the pivot matrix P, upper factorization U
as well as the row/col tracking vectors.

Arguments:

a: a stoichiometric matrix

\end{fulllineitems}

\index{GetUpperMatrixUsingQR() (cbmpy.PyscesStoich.Stoich method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.PyscesStoich.Stoich.GetUpperMatrixUsingQR}\pysiglinewithargsret{\bfcode{GetUpperMatrixUsingQR}}{\emph{a}}{}
GetUpperMatrix(a)

Core analysis algorithm; an input is preconditioned using PivotSort\_initial and then cycles of PLUfactorize and
PivotSort are run until the factorization is completed. During this process the matrix is reordered by
column swaps which emulates a full pivoting LU factorization. Returns the pivot matrix P, upper factorization U
as well as the row/col tracking vectors.

Arguments:

a: a stoichiometric matrix

\end{fulllineitems}

\index{K\_split\_R() (cbmpy.PyscesStoich.Stoich method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.PyscesStoich.Stoich.K_split_R}\pysiglinewithargsret{\bfcode{K\_split\_R}}{\emph{R\_a}, \emph{row\_vector}, \emph{column\_vector}}{}
Using the R factorized form of the stoichiometric matrix we now form the K and Ko matrices. Returns
the r\_ipart,Komatrix,Krow,Kcolumn,Kmatrix,Korow,info

Arguments:

R\_a: the Gauss-Jordan reduced stoichiometric matrix
row\_vector: row tracking vector
column\_vector: column tracking vector

\end{fulllineitems}

\index{L\_split\_R() (cbmpy.PyscesStoich.Stoich method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.PyscesStoich.Stoich.L_split_R}\pysiglinewithargsret{\bfcode{L\_split\_R}}{\emph{Nfull}, \emph{R\_a}, \emph{row\_vector}, \emph{column\_vector}}{}
Takes the Gauss-Jordan factorized N\textasciicircum{}T and extract the L, Lo, conservation (I -Lo) and reduced stoichiometric matrices. Returns: lmatrix\_col\_vector, lomatrix, lomatrix\_row, lomatrix\_co, nrmatrix, Nred\_vector\_row, Nred\_vector\_col, info

Arguments:

Nfull: the original stoichiometric matrix N
R\_a: gauss-jordan factorized form of N\textasciicircum{}T
row\_vector: row tracking vector
column\_vector: column tracking vector

\end{fulllineitems}

\index{PLUfactorize() (cbmpy.PyscesStoich.Stoich method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.PyscesStoich.Stoich.PLUfactorize}\pysiglinewithargsret{\bfcode{PLUfactorize}}{\emph{a\_in}}{}
Performs an LU factorization using LAPACK D/ZGetrf. Now optimized for FLAPACK interface.
Returns LU - combined factorization, IP - rowswap information and info - Getrf error control.

Arguments:

a\_in: the matrix to be factorized

\end{fulllineitems}

\index{PivotSort() (cbmpy.PyscesStoich.Stoich method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.PyscesStoich.Stoich.PivotSort}\pysiglinewithargsret{\bfcode{PivotSort}}{\emph{a}, \emph{row\_vector}, \emph{column\_vector}}{}
This is a sorting routine that accepts a matrix and row/colum vectors
and then sorts them so that: there are no zero rows (by swapping with first
non-zero row) The abs(largest) pivots are moved onto the diagonal to maintain
numerical stability. Row and column swaps are recorded in the tracking vectors.

Arguments:

a: the input array
row\_vector: row tracking vector
column\_vector: column tracking vector

\end{fulllineitems}

\index{PivotSort\_initial() (cbmpy.PyscesStoich.Stoich method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.PyscesStoich.Stoich.PivotSort_initial}\pysiglinewithargsret{\bfcode{PivotSort\_initial}}{\emph{a}, \emph{row\_vector}, \emph{column\_vector}}{}
This is a sorting routine that accepts a matrix and row/colum vectors
and then sorts them so that: the abs(largest) pivots are moved onto the diagonal to maintain
numerical stability i.e. the matrix diagonal is in descending max(abs(value)).
Row and column swaps are recorded in the tracking vectors.

Arguments:

a: the input array
row\_vector: row tracking vector
column\_vector: column tracking vector

\end{fulllineitems}

\index{SVD\_Rank\_Check() (cbmpy.PyscesStoich.Stoich method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.PyscesStoich.Stoich.SVD_Rank_Check}\pysiglinewithargsret{\bfcode{SVD\_Rank\_Check}}{\emph{matrix=None}, \emph{factor=1.0e4}, \emph{resultback=0}}{}
Calculates the dimensions of L/L0/K/K) by way of SVD and compares them to the Guass-Jordan results. Please note that for LARGE ill conditioned matrices the SVD can become numerically unstable when used for nullspace determinations

Arguments:

matrix {[}default=None{]}: the stoichiometric matrix default is self.Nmatrix
factor {[}default=1.0e4{]}: factor used to calculate the `zero pivot' mask = mach\_eps*factor
resultback {[}default=0{]}: return the SVD results, U, S, vh

\end{fulllineitems}

\index{ScalePivots() (cbmpy.PyscesStoich.Stoich method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.PyscesStoich.Stoich.ScalePivots}\pysiglinewithargsret{\bfcode{ScalePivots}}{\emph{a\_one}}{}
Given an upper triangular matrix U, this method scales the diagonal (pivot values) to one.

Arguments:

a\_one: an upper triangular matrix U

\end{fulllineitems}

\index{SplitLU() (cbmpy.PyscesStoich.Stoich method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.PyscesStoich.Stoich.SplitLU}\pysiglinewithargsret{\bfcode{SplitLU}}{\emph{plu}, \emph{row}, \emph{col}, \emph{t}}{}
PLU takes the combined LU factorization computed by PLUfactorize and extracts the upper matrix.
Returns U.

Arguments:

plu: LU factorization
row: row tracking vector
col: column tracking vector
t {[}default=None){]}: typecode argument (currently not used)

\end{fulllineitems}


\end{fulllineitems}

\index{StructMatrix (class in cbmpy.PyscesStoich)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.PyscesStoich.StructMatrix}\pysiglinewithargsret{\strong{class }\code{cbmpy.PyscesStoich.}\bfcode{StructMatrix}}{\emph{array}, \emph{ridx}, \emph{cidx}, \emph{row=None}, \emph{col=None}}{}
This class is specifically designed to store structural matrix information
give it an array and row/col index permutations it can generate its own
row/col labels given the label src.
\index{getColsByIdx() (cbmpy.PyscesStoich.StructMatrix method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.PyscesStoich.StructMatrix.getColsByIdx}\pysiglinewithargsret{\bfcode{getColsByIdx}}{\emph{*args}}{}
Return the columns referenced by index (1,3,5)

\end{fulllineitems}

\index{getColsByName() (cbmpy.PyscesStoich.StructMatrix method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.PyscesStoich.StructMatrix.getColsByName}\pysiglinewithargsret{\bfcode{getColsByName}}{\emph{*args}}{}
Return the columns referenced by label (`s','x','d')

\end{fulllineitems}

\index{getIndexes() (cbmpy.PyscesStoich.StructMatrix method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.PyscesStoich.StructMatrix.getIndexes}\pysiglinewithargsret{\bfcode{getIndexes}}{\emph{axis='all'}}{}
Return the matrix indexes ({[}rows{]},{[}cols{]}) where axis='row'/'col'/'all'

\end{fulllineitems}

\index{getLabels() (cbmpy.PyscesStoich.StructMatrix method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.PyscesStoich.StructMatrix.getLabels}\pysiglinewithargsret{\bfcode{getLabels}}{\emph{axis='all'}}{}
Return the matrix labels ({[}rows{]},{[}cols{]}) where axis='row'/'col'/'all'

\end{fulllineitems}

\index{getRowsByIdx() (cbmpy.PyscesStoich.StructMatrix method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.PyscesStoich.StructMatrix.getRowsByIdx}\pysiglinewithargsret{\bfcode{getRowsByIdx}}{\emph{*args}}{}
Return the rows referenced by index (1,3,5)

\end{fulllineitems}

\index{getRowsByName() (cbmpy.PyscesStoich.StructMatrix method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.PyscesStoich.StructMatrix.getRowsByName}\pysiglinewithargsret{\bfcode{getRowsByName}}{\emph{*args}}{}
Return the rows referenced by label (`s','x','d')

\end{fulllineitems}

\index{setCol() (cbmpy.PyscesStoich.StructMatrix method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.PyscesStoich.StructMatrix.setCol}\pysiglinewithargsret{\bfcode{setCol}}{\emph{src}}{}
Assuming that the col index array is a permutation (full/subset)
of a source label array by supplying that src to setCol
maps the row labels to cidx and creates self.col (col label list)

\end{fulllineitems}

\index{setRow() (cbmpy.PyscesStoich.StructMatrix method)}

\begin{fulllineitems}
\phantomsection\label{modules_doc:cbmpy.PyscesStoich.StructMatrix.setRow}\pysiglinewithargsret{\bfcode{setRow}}{\emph{src}}{}
Assuming that the row index array is a permutation (full/subset)
of a source label array by supplying that source to setRow it
maps the row labels to ridx and creates self.row (row label list)

\end{fulllineitems}


\end{fulllineitems}

\phantomsection\label{modules_doc:module-cbmpy._multicorefva}\index{cbmpy.\_multicorefva (module)}

\section{CBMPy: MultiCoreFVA module}
\label{modules_doc:cbmpy-multicorefva-module}
PySCeS Constraint Based Modelling (\href{http://cbmpy.sourceforge.net}{http://cbmpy.sourceforge.net})
Copyright (C) 2009-2015 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see \textless{}\href{http://www.gnu.org/licenses/}{http://www.gnu.org/licenses/}\textgreater{}

Author: Brett G. Olivier
Contact email: \href{mailto:bgoli@users.sourceforge.net}{bgoli@users.sourceforge.net}
Last edit: \$Author: bgoli \$ (\$Id: \_multicorefva.py 378 2015-09-14 16:32:38Z bgoli \$)
\phantomsection\label{modules_doc:module-cbmpy._multicoreenvfva}\index{cbmpy.\_multicoreenvfva (module)}

\section{CBMPy: MultiCoreEnvFVA module}
\label{modules_doc:cbmpy-multicoreenvfva-module}
PySCeS Constraint Based Modelling (\href{http://cbmpy.sourceforge.net}{http://cbmpy.sourceforge.net})
Copyright (C) 2009-2015 Brett G. Olivier, VU University Amsterdam, Amsterdam, The Netherlands

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see \textless{}\href{http://www.gnu.org/licenses/}{http://www.gnu.org/licenses/}\textgreater{}

Author: Brett G. Olivier
Contact email: \href{mailto:bgoli@users.sourceforge.net}{bgoli@users.sourceforge.net}
Last edit: \$Author: bgoli \$ (\$Id: \_multicoreenvfva.py 378 2015-09-14 16:32:38Z bgoli \$)
\phantomsection\label{modules_doc:module-cbmpy.miriamids}\index{cbmpy.miriamids (module)}

\chapter{Indices and tables}
\label{cbmpy:indices-and-tables}\begin{itemize}
\item {} 
\DUspan{xref,std,std-ref}{genindex}

\item {} 
\DUspan{xref,std,std-ref}{modindex}

\item {} 
\DUspan{xref,std,std-ref}{search}

\end{itemize}


\renewcommand{\indexname}{Python Module Index}
\begin{theindex}
\def\bigletter#1{{\Large\sffamily#1}\nopagebreak\vspace{1mm}}
\bigletter{c}
\item {\texttt{cbmpy.\_multicoreenvfva}}, \pageref{modules_doc:module-cbmpy._multicoreenvfva}
\item {\texttt{cbmpy.\_multicorefva}}, \pageref{modules_doc:module-cbmpy._multicorefva}
\item {\texttt{cbmpy.CBCommon}}, \pageref{modules_doc:module-cbmpy.CBCommon}
\item {\texttt{cbmpy.CBConfig}}, \pageref{modules_doc:module-cbmpy.CBConfig}
\item {\texttt{cbmpy.CBCPLEX}}, \pageref{modules_doc:module-cbmpy.CBCPLEX}
\item {\texttt{cbmpy.CBDataStruct}}, \pageref{modules_doc:module-cbmpy.CBDataStruct}
\item {\texttt{cbmpy.CBGUI}}, \pageref{modules_doc:module-cbmpy.CBGUI}
\item {\texttt{cbmpy.CBModel}}, \pageref{modules_doc:module-cbmpy.CBModel}
\item {\texttt{cbmpy.CBModelTools}}, \pageref{modules_doc:module-cbmpy.CBModelTools}
\item {\texttt{cbmpy.CBMultiCore}}, \pageref{modules_doc:module-cbmpy.CBMultiCore}
\item {\texttt{cbmpy.CBMultiEnv}}, \pageref{modules_doc:module-cbmpy.CBMultiEnv}
\item {\texttt{cbmpy.CBNetDB}}, \pageref{modules_doc:module-cbmpy.CBNetDB}
\item {\texttt{cbmpy.CBPlot}}, \pageref{modules_doc:module-cbmpy.CBPlot}
\item {\texttt{cbmpy.CBQt4}}, \pageref{modules_doc:module-cbmpy.CBQt4}
\item {\texttt{cbmpy.CBRead}}, \pageref{modules_doc:module-cbmpy.CBRead}
\item {\texttt{cbmpy.CBReadtxt}}, \pageref{modules_doc:module-cbmpy.CBReadtxt}
\item {\texttt{cbmpy.CBSolver}}, \pageref{modules_doc:module-cbmpy.CBSolver}
\item {\texttt{cbmpy.CBTools}}, \pageref{modules_doc:module-cbmpy.CBTools}
\item {\texttt{cbmpy.CBWrite}}, \pageref{modules_doc:module-cbmpy.CBWrite}
\item {\texttt{cbmpy.CBWx}}, \pageref{modules_doc:module-cbmpy.CBWx}
\item {\texttt{cbmpy.CBXML}}, \pageref{modules_doc:module-cbmpy.CBXML}
\item {\texttt{cbmpy.miriamids}}, \pageref{modules_doc:module-cbmpy.miriamids}
\item {\texttt{cbmpy.PyscesStoich}}, \pageref{modules_doc:module-cbmpy.PyscesStoich}
\end{theindex}

\renewcommand{\indexname}{Index}
\printindex
\end{document}
